<html>
<head>
<meta name="author" content="helpdesk" />
<meta name="navigation" content="Ensembl Core" />
<title>Ensembl Core</title>
</head>

<body>

<h2>Ensembl Core</h2>


<h3 class="boxed">Database Descriptions</h3>


<h4>Core database</h4>

<p>
For each organism supported by Ensembl, a species-specific Ensembl Core database stores the bulk of annotation information.
Some of the key features are:
</p>

<ul>

<li>
<p>
The unmasked genome sequence is stored on the sequence-level i.e. in form of BAC clones or whole genome shotgun scaffolds.
</p>
</li>

<li>
<p>
Assembly information the generate top-level sequences like chromosomes from sequence-level entities.
</p>
</li>

<li>
<p>
The non-redundant set of Ensembl gene, transcript and protein models annotated by the automated genome analysis and
<a href="../../data/docs/genome_annotation.html">annotation pipeline</a>.
</p>
</li>

<li>
<p>
Sequence tagged sites (STS) marker information on physical sequence and genetic map locations.
</p>
</li>

<li>
<p>
Cytogenetic banding patterns if a reliable banding pattern has been established for the organism.
</p>
</li>

<li>
<p>
Microarray probe set annotation.
</p>
</li>

<li>
<p>
External references to other databases of biological information.
</p>
</li>

</ul>

<h4><a id="est" name="est">cDNA database</a></h4>

<p>
Alignments of cDNAs and proteins to the genome sequence, which serve as source of biological evidence in the Ensembl annotation strategy.
</p>

<h4><a id="est" name="est">OtherFeatures database</a></h4>

<p>
For well-characterised species with a reasonable amount of biological sequence records as supporting evidence
Ensembl provides an independent EST gene set, which is solely based on EST information.
The key features stored are:
</p>

<ul>

<li>
<p>
The non-redundant Ensembl EST gene set.
</p>
</li>

<li>
<p>
A pre-filtered set of EST sequences aligned to the genome sequence.
This set seved as the biological evidence for EST gene annotation.
</p>
</li>

</ul>

<p>
Please note that Ensembl OtherFeatures databases do not contain genome sequences.
This saves disk space and by connecting to Core and the OtherFeatures databases simultaneously, it is still possible to retrieve sequence information.
</p>


<h3 class="boxed">Database Schema</h3>

<p>
The layout of Ensembl Core and EST Databases is identical so that schema descriptions and API access apply similarly.
The database table layout is explained in the following document:
</p>

<ul>

<li>
<p>
<a href="./schema/index.html">Core Schema Description</a>
</p>
</li>

</ul>


<h3 class="boxed">Perl API</h3>

<p>
A comprehensive Perl Application Programme Interface (API) provides efficient access to tables within the Ensembl Core and EST databases.
</p>

<ul>

<li>
<p>
<a href="../api_installation.html">Core Perl API Installation</a>: A step-by-step installation guide for all Ensembl Perl APIs.
</p>
</li>

<li>
<p>
<a href="../Pdoc/ensembl/index.html">Core Perl API Documentation</a>: A complete reference to the objects and methods used in the Core and EST database API.
</p>
</li>

<li>
<p>
<a href="./core_tutorial.html">Core Perl API Tutorial</a>: An introduction to the underlying concepts of the Core and EST database API.
</p>
</li>

</ul>

<!--
<h3 class="boxed">Java API</h3>

<p>
The Ensembl Java API (Ensj) is slightly less complete than the Perl APIs but implements access to Ensembl
<a href="../core/index.html">Core</a>,
<a href="../compara/index.html">Compara</a> and
<a href="../variation/index.html">Variation</a> databases simultaneously.
</p>

<ul>
<li>
<p>
<a href="../java/index.html">Java API Documentation</a>
</p>
</li>
</ul>
-->

<h3 class="boxed" id="support">Ensembl Software Support</h3>

<p>
Ensembl is an open project and we would like to encourage correspondence and discussions on any subject on any aspect of Ensembl.
Please see the
<a href="../../about/contact.html">Ensembl Contacts</a> page for suitable options getting in touch with us.
</p>

</body>
</html>
