<html>
<head>
<title>Regulation Database and API</title>
<meta name="order" content="13" />
</head>

<body>

<h1>Ensembl Funcgen Summary</h1> 


<h2>Database Description</h2>

<p>
The Ensembl regulation database and API are developed and maintained by the Ensembl Regulation team. The main focus of the 'funcgen' database is on experimental data which may elucidate transcriptional regulation.  The schema was originally designed to support ChIP on Chip tiling array experiments but soon expanded to support other sequence based technologies.  The original requirements supported use as a local data storage and analysis platform, as such the regulation data bases supports multiple coordinate systems. There is also support for micro array designs and externally curated data sets. See <a href='/info/docs/funcgen/index.html'>here</a> for a summary of the data base content.

<!-- The database content includes:

</p>

<ul>
<li>The <a href='http://www.ensembl.org/info/docs//index.html'>Ensembl Regulatory Build</a> - Regulatory features generated from a combined analysis of all of the above.</li>
<li>Experimental data and meta data from functional genomics experiments using the following technologies:
  <ul>
  <li>Chromatin immunoprecipitation using tiling arrays (ChIP-chip)</li>
  <li>Chromatin immunoprecipitation sequencing</li>
  <li>Other techniques identifying open chromatin e.g. DNase sequencing, FAIRE</li>
  </ul>
</li>
<li>Experimental data on epigenomic modifications and DNA binding proteins:
  <ul>
  <li>Histone modifications</li>
  <li>Transcription factors e.g. CCCTC binding factor (CTCF) etc.</li>
  <li>RNAPollII</li>
  <li><a href='http://genome.ucsc.edu/cgi-bin/hgFileUi?db=hg19&g=wgEncodeHaibMethylRrbs'>DNA Methylation data</a> from Reduced Representation Bisulfite Sequencing (RRBS). Data from the <a href='http://www.genome.gov/10005107'>ENCODE</a> project  </ul>
</li>
<li>Externally curated data with a regulatory focus:
  <ul>
  <li><a href='http://www.cisred.org/'>cisRED</a> - Regulatory motifs</li>
  <li><a href='http://www.ebi.ac.uk/enright-srv/microcosm/htdocs/targets/v5/consset.html'>miRanda</a> - miRNA targets</li>
  <li><a href='http://enhancer.lbl.gov/'>VISTA</a> - Enhancers</li>
  </ul>
</li>
<li>Array designs
  <ul>
  <li>For various array types e.g. Affy IVT/ST, Illumina Whole Genome/Infinium, Codelink, Agilent etc</li>
  <li>Ensembl transcript annotations for Expression arrays.
  </ul>
</li>
</ul>
 -->

<h2>Database Schema</h2>

<p>
The table layout of the database is explained in the following document:
</p>

<ul>
<li>
<a href="funcgen_schema.html">Funcgen Schema Description</a>
</li>
</ul>


<h2 id="api">Perl API</h2>

<p>
A comprehensive Perl Application Programme Interface (API) provides efficient access to the Ensembl Funcgen database.
</p>

<ul>

<li>
<a href="../api_installation.html">EnsEMBL Perl API Installation</a>: A step-by-step installation guide for all Ensembl Perl APIs.
</li>

<!-- <li>
<a href="efg_introduction.html">Introduction to eFG</a>: An introduction on how to set up eFG as a local data storage and analysis platform.
</li> -->

<li>
<a href="/info/docs/Doxygen/funcgen-api/index.html">Regulation Perl API Documentation</a>: A complete reference to the objects and methods used in the Funcgen database API.
</li>

<li>
<a href="regulation_tutorial.html">Regulation Perl API Tutorial</a>: An introduction to the underlying concepts of the Funcgen database API.
</li>


<li>
<a href="reg_features.html">API support for old core Regulatory Features</a>: Details of the migration of regulatory feature support from the core API to Funcgen API.
</li>

<!-- <li>Guidelines on how to set up the eFG analysis environment are available from the eFG perl package: ensembl-functgenomics/scripts/introduction_to_efg.txt</li> -->

</ul>


<h2 id="support">Ensembl Software Support</h2>

<p>Ensembl is an open project and we would like to encourage
correspondence and discussions on any subject on any aspect of Ensembl.
Please see the
<a href="http://www.ensembl.org/info/about/contact/">Ensembl Contacts</a> page for suitable options getting in touch with us.
</p>


</body>
</html>

