<html>
<head>
<title>ENCODE resources</title>
</head>
<body>

  
<h2>ENCODE resources at Ensembl</h2>

<p> The <a href="http://www.genome.gov/10005107">ENCODE</a> project aims to discover the bulk of functional elements in the human genome by intensively annotating 1% of the genome.</p>

<p>The ENCODE project has 44 separate regions across the genome.  These were either chosen by a stratified random selection or because they are
biologically inspired regions and  chosen mainly for their existing density of
experimental results (e.g. parts of chromosome 22 and the <a href="http://www.ensembl.org/Homo_sapiens/diseaseview?disease=Cystic%20fibrosis">CFTR</a> region on chromosome 7).</p>

<p>This page coordinates the resources available via the Ensembl browser for
the ENCODE project.  Through the links below, you can</p>
<ul>
    <li> browse the ENCODE regions in ContigView</li>
    <li> simultaneously view other species data in MultiContigView</li>
    <li> use the <a href="/Multi/martview">EnsMart</a> data mining tool to dump gene and gene-associated features in the ENCODE region</li>
</ul>

<p>As part of the Ensembl team's involvement in the ENCODE project we are
generating repeat libraries for the genomes newly sequenced by NISC. The
libraries and further information can be found on the EBI anonymous ftp site:</p>

<p><kbd>
ftp ftp.ebi.ac.uk<br />
ftp&gt; cd pub/databases/ensembl/encode/repeat_libraries
</kbd></p>


<p>The ENCODE project primary data management is at <a href="http://genome.ucsc.edu/">UCSC</a> and we are coordinating with them to deliver genome resources
 for ENCODE.</p>

<ul>
    <li><a href="http://genome.ucsc.edu/ENCODE/regions_build34.html">More information on ENCODE regions from UCSC</a>.</li>
</ul>

  
<h3>ENCODE regions</h3>
   
[[INCLUDE::Homo_sapiens/ssi/encode_table.html]]

</body>
</html>

