<html>
<head>
<meta name="navigation" content="Ensembl" />
<title>Ensembl Genome Browser</title>
</head>
<body>

<h2>Ensembl BlastView Configuration</h2>
<h3 class="boxed">Contents</h3>
<ol>
<li><a href="#Intro">Introduction</a></li>
 <ul>
   <li><a href="#EnsConfig">Ensembl Configuration Overview</a></li>
   <li><a href="#BlastViewConfig">BlastView Configuration</a></li>
 </ul>
<li><a href="#seq">Registering Sequence Databases</a></li>
<li><a href="#methods">Registering Search Methods</a></li>
<li><a href="#links">Registering Method vs. Database Links</a></li>
<li><a href="#db">Configuring the ENSEMBL BLAST Database</a></li> 
<li><a href="#conc">Conclusion</a></li>
</ol>

<h3 class="boxed"><a id="Intro" name="Intro">Introduction</a></h3>
This document provides details for configuring the Ensembl BlastView web interface.

<h4><a id="EnsConfig">Ensembl Configuration Overview</a></h4>
<p>The Ensembl distribution is usually stored within a filesystem 
directory (hereafter refered to as $ENSEMBL ROOT), e.g.:
<kbd>$ setenv ENSEMBL_ROOT /usr/local/ensembl</kbd>
</p>

<p>Ensembl web site configuration data are stored in text files 
within $ENSEMBL ROOT, in the following directory;
<kbd>$ENSEMBL_ROOT/conf</kbd>
</p>

<p>Species-independent configuration are stored in 
(hereafter refered to as MULTI.ini):
<kbd>$ENSEMBL_ROOT/conf/ini-files/MULTI.ini</kbd>
</p>

<p>Species-default configuration are stored in (hereafter refered to as <kbd>DEFAULTS.ini</kbd>):
<kbd>$ENSEMBL_ROOT/conf/ini-files/DEFAULTS.ini</kbd>
</p>

<p>Species-specific configuration are stored in (hereafter refered to as <SPECIES>.ini), for example:<br />
<kbd>$ENSEMBL_ROOT/conf/ini-files/Homo_sapiens.ini<br />
$ENSEMBL_ROOT/conf/ini-files/Danio_rerio.ini</kbd>
</p>

<p>The .ini files are seperated into sections containing key-value pairs using the format:
<br />
<pre>
[SECTION_HEADING]
key1 = value1
key2 = value2
</pre>

<h4><a id="BlastViewConfig">BlastView Configuration</a></h4>
<p>BlastView configuration consists of the following components:</p>
<ul>
<li>Register of sequence databases to search against</li>
<li>Register of methods (executables) that run the search</li>
<li>Register of search method vs. sequence database linkage</li>
<li>Location of MySQL database for search report storage (ENSEMBL BLAST)</li>
</ul>

<p>Databases and methods are generally species independent, and 
their configurations are therefore stored in the <kbd>DEFAULTS.ini</kbd> file. 
Method vs. sequence database linkage, however, is species specific, 
and stored in the <SPECIES>.ini files.</p>

<p>For interface responsiveness search reports are parsed once and the results stored in a database.  The location of this database is configured in the MULTI.ini file.</p>

<h3 class="boxed"><a id="seq" name="seq">Registering Sequence Databases</a></h3>
<p>Sequence database configuration requires the following data;</p>
<ul>
<li>A "type" string, used for grouping of databases across species and methods, (e.g. CDNA ALL, PEP ALL)</li>
<li>A "label" string, used for display in the interface, (e.g. "Ensembl cDNAs", "Ensembl Peptides").</li>
</ul>

<p>To register databases, type/label pairs must be entered in the 
<kbd>BLAST DATASOURCES</kbd>  configuration section of
<kbd>DEFAULTS.ini</kbd>. An additional DEFAULT key sets the type of the default database e.g.:
<pre>
[BLAST_DATASOURCES]
# Register of BlastView sequence databases. Use format;
# DB_TYPE = DB_LABEL
# DB_TYPE is the internal identifier (no whitespace)

# DB_LABEL is the human-readable label.
LATESTGP = Genomic sequence
CDNA_ALL = Ensembl cDNAs
CDNA_GENSCAN = Genscan cDNAs
PEP_ALL = Ensembl Peptides
PEP_GENSCAN = Genscan Peptides
# Default sequence database. Use format;
# DEFAULT = DB_TYPE
DEFAULT = LATESTGP
</pre>

<p>As for all Ensembl configuration sections, <kbd>DEFAULTS.ini</kbd> section headers must also be included in each relevent
<kbd>&lt;SPECIES&gt;.ini</kbd> file. For example, <kbd>Homo_sapiens.ini</kbd> contains the following:</p>
<pre>
[BLAST_DATASOURCES]
# Register of BlastView sequence databases.
# Accept defaults
</pre>

<p>It is also possible to register species-specific database types in the corresponding <kbd>&lt;SPECIES&gt;.ini</kbd> files.
However, as non-empty sections in <kbd>&lt;SPECIES&gt;</kbd> take precedence over those in <kbd>DEFAULTS.ini</kbd>, one must
copy all other types relevent to the species. For example, in <kbd>Homo_sapiens.ini</kbd>:
<pre>
[BLAST_DATASOURCES]
# Homo_sapiens-specific database type:
HAPLO = Haplotype blocks
# Standard types copied from <kbd>DEFAULTS.ini</kbd>
LATESTGP = Genomic sequence
CDNA_ALL = Ensembl cDNAs
...
</pre>

<p>Given the problems with registering species-specific database types, 
it is probably best to simply register all types in <kbd>DEFAULTS.ini</kbd> - there is an alternative way of per-species database type selection, which is covered later.</p>


<h3 class="boxed"><a id="methods" name="methods">Registering</a> Search Methods</h3>
<p>Within the context of BlastView, a search method is an algorithm/executable that takes a query sequence
of a particular type (DNA/peptide), and a sequence database of a particular type (DNA/peptide), and
computes some measure of sequence similarity between the two. This means that BLASTN, TBLASTX,
BLASTP, SSAHA etc. are entirely seperate at the configuration level. Before use with BlastView, code
must be written to "wrap" each executable in a perl module with a "runnable" interface. These wrapper
modules hide the differences in calling and report handling between methods from the BlastView interface
code. Coding of these wrappers for different methods/system configurations is covered in the BlastView tech
document.
The key attributes of BlastView methods are therefore:
</p>
<ul>
<li>A "type" string used for grouping methods across species, also used as human-readable label</li>
<li>A "module" string used to identify the appropriate wrapper class.</li>
</ul>

<p>To register methods, type/module pairs must be entered in the <kbd>ENSEMBL BLAST METHODS</kbd> configuration section of <kbd>DEFAULTS.ini</kbd> e.g.:
<pre>
[ENSEMBL_BLAST_METHODS]
# Register of BlastView methods. Use format:
# ME_TYPE = ME_WRAP
# ME_TYPE is the internal identifier (no whitespace)
# ME_WRAP is the Bio::Tools::Run::Search wrapper class
BLASTN = ensembl_wublastn
BLASTX = ensembl_wublastx
BLASTP = ensembl_wublastp
TBLASTN = ensembl_wutblastn
TBLASTX = ensembl_wutblastx
SSAHA = ensembl_ssaha
</pre>

<p>For example, the ensembl wublastn wrapper contains the logic to run a wu-blastn search using the Compaq
BSUB job submission system as used for the Ensembl Blast cluster. Conversly, the ensembl ssaha wrapper
runs SSAHA searches over TCP-IP uising a client-server model. Further wrappers are being developed to,
for example, run blast searches on the same machine as the web server.</p>

<p>Unfortunately, it is not currently possible to override the blast method wrappers on a per-species basis. This
is a known limitation of the system, and may be addressed in the future. However, you still need to add a
"stub" <kbd>ENSEMBL BLAST METHODS</kbd> section for each species, e.g. within <kbd>Homo_sapiens.ini</kbd>:
<pre>
[ENSEMBL_BLAST_METHODS]
# Register of BlastView sequence databases.
# Accept defaults
</pre></p>

<p>In addition to the ENSEMBL BLAST METHODS section, there are several attributes in the general section that also affect method configuration. These are:
<pre>
# Path to binaries on local machine
ENSEMBL_BINARIES_PATH = /usr/local/bin
# Path to binaries on remote machine
ENSEMBL_BLAST_BIN_PATH = /usr/remote/bin
# Path to blast databases
ENSEMBL_BLAST_DATA_PATH = /usr/remote/data
# Path to blast filter directory
ENSEMBL_BLAST_FILTER = /usr/remote/blast/filter
# Path to blast matrix directory
ENSEMBL_BLAST_MATRIX = /usr/remote/blast/matrix
# Path to RepeatMasker executable
ENSEMBL_REPEATMASKER = /usr/remote/RepeatMasker
# Names of method executables. Names correspond to method types in ENSEMBL_BLAST_METHODS
ENSEMBL_SSAHA_PROGRAM_NAME = ssahaClient3.1
ENSEMBL_BLASTN_PROGRAM_NAME = wublastn
ENSEMBL_BLASTX_PROGRAM_NAME = wublastx
ENSEMBL_BLASTP_PROGRAM_NAME = wublastp
ENSEMBL_TBLASTN_PROGRAM_NAME = wutblastn
ENSEMBL_TBLASTX_PROGRAM_NAME = wutblastx
</pre></p>

<h3 class="boxed"><a id="links" name="links">Registering </a>Method vs. Database Links</h3>
<p>Method/database links are species-specific, and therefore entered in the <kbd>&lt;SPECIES&gt;</kbd> files. Each method
must have a corresponding section named &lt;METHOD&gt; DATASOURCES, where &lt;METHOD&gt; refers to one of the
method types registered in <kbd>DEFAULTS.ini</kbd>, e.g. BLASTN DATASOURCES, SSAHA DATASOURCES.</p>

<p>The &lt;METHOD&gt; DATASOURCES section contains links to one or more sequence databases, where the key=value pairs correspond to; &lt;DATABASE&gt; (refering to one of the database types registered in <kbd>DEFAULTS.ini</kbd>), and
&lt;LOCATOR&gt; (refering to the filesystem or TCP-IP location of the database).</p>

<p>In addition to the &lt;DATABASE&gt; = &lt;LOCATOR&gt; entries in the &lt;METHOD&gt; DATASOURCES section, a
DATASOURCE TYPE key is used to specify the query type (dna/peptide) that the search method expects as input.
For example, <kbd>Homo_sapiens.ini</kbd> contains:</p>

<pre>[BLASTN_DATASOURCES]
DATASOURCE_TYPE = dna
LATESTGP = Homo_sapiens.NCBI33.contig.fa
CDNA_ALL = Homo_sapiens.NCBI33.cdna.fa
CDNA_GENSCAN = Homo_sapiens.NCBI33.cdna_genscan.fa
[TBLASTX_DATASOURCES]
DATASOURCE_TYPE = dna
LATESTGP = Homo_sapiens.NCBI33.contig.fa
CDNA_ALL = Homo_sapiens.NCBI33.cdna.fa
CDNA_GENSCAN = Homo_sapiens.NCBI33.cdna_genscan.fa
[TBLASTN_DATASOURCES]
DATASOURCE_TYPE = peptide
LATESTGP = Homo_sapiens.NCBI33.contig.fa
CDNA_ALL = Homo_sapiens.NCBI33.cdna.fa
CDNA_GENSCAN = Homo_sapiens.NCBI33.cdna_genscan.fa
[BLASTP_DATASOURCES]
DATASOURCE_TYPE = peptide
PEP_ALL = Homo_sapiens.NCBI33.pep.fa
PEP_GENSCAN = Homo_sapiens.NCBI33.pep_genscan.fa
[BLASTX_DATASOURCES]
DATASOURCE_TYPE = dna
PEP_ALL = Homo_sapiens.NCBI33.pep.fa
PEP_GENSCAN = Homo_sapiens.NCBI33.pep_genscan.fa
[SSAHA_DATASOURCES]
DATASOURCE_TYPE = dna
LATESTGP = ssaha01:50001
</pre>

<h3 class="boxed"><a id="db" name="db">Configuring</a> the ENSEMBL BLAST Database</h3>
<p>To improve interface responsiveness, search reports are parsed once, and the results stored in a database.
The location of this database is configured in the MULTI.ini file. Configuration is similar to that of all other
Ensembl databases, with the name of the database being ENSEMBL BLAST. The main difference between
ENSEMBL BLAST and other Ensembl databases, however, is that ENSEMBL BLAST is a read/write database, so
the configured database user must have write permission. For example:</p>

<pre>[databases]
ENSEMBL_BLAST = ensembl_blast
[ENSEMBL_BLAST]
HOST = localhost
PORT = 3306
USER = admin_user
PASS = secret
</pre>
<p>The database schema of the ENSEMBL BLAST database is distributed within the perl code, rather than available
by FTP download. Firstly, an empty database should be created. E.g.
<pre>$ mysql -u admin_user -p secret -e "create database ensembl_blast"</pre>

Next, a script can be run that creates the database automatically. E.g.
<pre>$ $ENSEMBL_ROOT/utils/utils/blast_database.pl</pre>
The correct execution of this script can be checked as follows:
<pre>$ mysqldump -u admin_user -p secret ensembl_blast</pre>
The above should result in output similar to the following. Note that the blast result, blast hit and blast hsp
tables are timestamped. Re-running the blast database.pl script will cause the blast result, blast hit and
blast hsp tables to be rotated, meaning that, whilst old results/hsps/hits will still be available, new searches
will be stored in the new tables. It is simple, therefore, to maintain the ENSEMBL BLAST database by dropping
old tables. See the $ENSEMBL ROOT/utils/blast cleaner.pl script for an example of how this is done.</p>

<pre>
-- -- Table structure for table "blast_table_log" --
CREATE TABLE blast_table_log (
table_id int(10) unsigned NOT NULL auto_increment,
table_name varchar(32) default NULL,
table_type enum("TICKET","RESULT","HIT","HSP") default NULL,
table_status enum("CURRENT","FILLED","DELETED") default NULL,
use_date date default NULL,
create_time datetime default NULL,
delete_time datetime default NULL,
num_objects int(10) default NULL,
PRIMARY KEY (table_id),
KEY table_name (table_name),
KEY table_type (table_type),
KEY use_date (use_date),
KEY table_status (table_status)
) TYPE=MyISAM;

-- -- Table structure for table "blast_ticket" --
CREATE TABLE blast_ticket (
ticket_id int(10) unsigned NOT NULL auto_increment,
create_time datetime NOT NULL default "0000-00-00 00:00:00",
update_time datetime NOT NULL default "0000-00-00 00:00:00",
ticket varchar(32) NOT NULL default "",
object longblob, PRIMARY KEY (ticket_id),
UNIQUE KEY ticket (ticket),
KEY create_time (create_time),
KEY update_time (update_time)
) TYPE=MyISAM;

-- -- Table structure for table "blast_result20030821" --
CREATE TABLE blast_result20030821 (
result_id int(10) unsigned NOT NULL auto_increment,
ticket varchar(32) default NULL, object longblob,
PRIMARY KEY (result_id), KEY ticket (ticket)
) TYPE=MyISAM;
CREATE TABLE blast_hit20030821 (
hit_id int(10) unsigned NOT NULL auto_increment,
ticket varchar(32) default NULL, object longblob,
PRIMARY KEY (hit_id),
KEY ticket (ticket)
) TYPE=MyISAM;

-- -- Table structure for table "blast_hsp20030821" --
CREATE TABLE blast_hsp20030821 (
hsp_id int(10) unsigned NOT NULL auto_increment,
ticket varchar(32) default NULL, object longblob,
chr_name varchar(32) default NULL,
chr_start int(10) unsigned default NULL,
chr_end int(10) unsigned default NULL,
PRIMARY KEY (hsp_id),
KEY ticket (ticket)
) TYPE=MyISAM MAX_ROWS=705032704 AVG_ROW_LENGTH=4000;
</pre>

<h3 class="boxed"><a id="conc" name="conc">Conclusion</a></h3>
<p>By following the above steps, the BlastView interface should be available for use on the next restart of the
Ensembl web server. Nowever, if the configuration .ini files have been changed, the following file should be
deleted before server restart (this is a filesystem cache of the config):
<pre>$ENSEMBL_ROOT/conf/config.packed</pre>

For further details about how BlastView works, please see the BlastView technical documentation:
<pre>In preperation. Contact <a href="mailto:whs@sanger.ac.uk">whs@sanger.ac.uk</a> for a draft.</pre>

Please address all comments/suggestions on this document, or on BlastView in general to: <a href="mailto:helpdesk@ensembl.org">helpdesk@ensembl.org</a>
 (cc <a href="maito:whs@sanger.ac.uk">whs@sanger.ac.uk</a>).</p>

</body>
</html>
