<html>
  <head>
    <meta name="navigation" content="Build" />
    <title>Building/Installing an Ensembl Website</title>
  </head>
<body> 
  <h2>Building/Installing an Ensembl Website</h2>

<p>The installation can be split into two sections, <a href="#non-ensembl">non-Ensembl</a> and <a href="#ensembl">Ensembl</a>. The first of these involves the
installation of applications/modules that are not part of the Ensembl project but are necessary for the
web site to work, such as Apache and MySQL. The second involves installing the Ensembl data,
modules and web site code.</p>

<h3 class="boxed"><a name="non-ensembl">Non-Ensembl Applications Build/Install</a></h3>
<p>Ensembl is built on the following applications:</p>
<ul>
  <li><a href="#perl">Perl</a> - Open source scripting language</li>
  <li><a href="#apache">Apache and mod_perl</a> - Open source web server</li>
  <li><a href="#mysql">MySQL</a> - Open source database server</li>
  <li><a href="#dotter">Dotter</a> - Graphical alignment package</li>
  <li><a href="#cvs">CVS</a> - versioning system to allow you to download the Ensembl code</a></li>
</ul> 
<p>These applications are not version-specific for Ensembl; that is, if you upgrade your Ensembl installation
to a newer version when one becomes available, you probably won't need to install new versions of
these applications.</p>

<p>All of this software, like all of Ensembl, is Open Source software and can be downloaded and used free
of charge. You should, however, check the documentation for each application to see what license it
has been released under, particularly if you are installing Ensembl in a commercial environment.</p>

<p>The following instructions assume you have root access to the installation machine. If you do not, get
your systems administrator to install this software for you.</p>

<p>You may have some or all of this software installed already. If you have any problems getting the site
running with pre-installed software (in particular Apache with mod_perl installed from RPMs), we
recommend simply installing the latest version using the following instructions.</p>

<h3 class="boxed"><a name="cvs">CVS</a></h3>
<p>CVS is a software version control system that we use for storing the source code to Ensembl. You will
need CVS installed if you want to download the Ensembl source code. It will also help you keep up to
date with any bug fixes.  We also have a  
<a href="http://cvsweb.sanger.ac.uk/cgi-bin/cvsweb.cgi/?cvsroot=Ensembl">Web-based CVS</a> repository.
</p>

</p>
<h4 class="boxed">To install CVS:</h4>
<ol>
<li>
  <p>Download the latest source from <a href="https://ccvs.cvshome.org/servlets/ProjectDocumentList" target="_blank">https://ccvs.cvshome.org/servlets/ProjectDocumentList</a> (you may be
able to download binaries from here instead if you prefer). At the time of writing this is version
1.11.19 and the file to download is cvs-1.11.19.tar.gz</p>
</li>
<li>
  <p>Unpack the source in a working directory with:</p>
  <kbd>
  gunzip < cvs-1.11.19.tar.gz | tar xvf -
  cd cvs-1.11.19<br />
  ./configure<br />
  make<br />
  make install</kbd>
</li>
</ol>

<h3 class="boxed"><a name="perl">Perl</a></h3>
<p>You will probably already have Perl installed. You need Perl5, version 5.8.0 or higher to run the website.
We run the site under Perl 5.8.3. To see if you have Perl installed, and/or to check it's version number,
type:</p>
<kbd>perl -v</kbd>
<p>
  If you don't have Perl installed, or need to upgrade, go to <a href="http://www.cpan.org/" target="_blank">http://www.cpan.org/</a> and
  choose the 'source code' install. Follow the installation instructions on the web site.
</p>

<h3 class="boxed"><a name="mysql">MySQL</a></h3>
<p>MySQL is a very popular Open Source relational database system. The easiest way to install MySQL is
to use the pre-compiled binaries from <a href="http://dev.mysql.com" target="_blank">http://dev.mysql.com</a>. You can also get source from
<a href="http://dev.mysql.com" target="_blank">http://dev.mysql.com</a> if you wish to compile MySQL yourself.</p>
<h4>To install MySQL:</h4>
<ol>
<li><p>Download the appropriate standard binaries from <a href="http://dev.mysql.com/downloads/mysql" target="_blank">http://dev.mysql.com/downloads/mysql</a>. Get the
current stable version - at the time of writing, this is 4.1.12.</p></li>
<li><p>Create a directory for MySQL to be installed into. A subdirectory of this will hold the databases, so
choose somewhere that has sufficient space free - at least 170 GB for the complete set. We will use
/data/ as an example. Again, when following these instructions, replace /data/ with whatever path
you choose.</p></li>
<li><p>Move the binary tarball to /data/</p></li>
<li><p>Unpack the tarball with:</p>
<kbd>$ gunzip < mysql-WHATEVER.tar.gz | tar xvf -</kbd>
</li>
<p>Follow the straightforward setup instructions in the INSTALL-BINARY file that comes with MySQL. It
should be located in the "mysql-WHATEVER" directory you just unpacked.</p>
</ol>

<h3 class="boxed"><a name="apache">Apache &amp; mod_perl</a></h3>
<p>Apache is the web server that the Ensembl site runs on. mod_perl is a module for Apache that enables
it to compile perl scripts once rather than each time they are requested, and so makes everything run a
lot faster.</p>

<p>Note that while the latest version of Apache is the version 2 series, mod_perl for Apache 2 is still under
development, and so we aren't currently using it. The installation instructions that follow may not,
therefore, work with version 2 of Apache and mod_perl.</p>

<p><b>Please follow these instructions precisely as often the default version of Apache or mod-perl does not work correctly for Ensembl.</b></p>

<h4>To Install Apache with mod_perl:</h4>
<ol>
<li><p>
  Download the Apache source tarball from <a href="http://httpd.apache.org/dist/httpd/" target="_blank">http://httpd.apache.org/dist/httpd/</a>. Get the current stable
version - at the time of writing, this is 1.3.33, and the file to download is apache_1.3.33.tar.gz.
</p></li>
<li>
  <p>Download the mod_perl source from <a href="http://www.cpan.org/modules/by-module/Apache/" target="_blank">http://www.cpan.org/modules/by-module/Apache/</a> . Again, get
the latest version, currently this is 1.29 and the file to download is mod_perl-1.29.tar.gz.
  </p>
</li>
<li>
  <p>Unpack all the sources in a working directory with:</p>
  <kbd>gunzip < apache_1.3.33.tar.gz | tar xvf -<br />
  gunzip < mod_perl-1.29.tar.gz | tar xvf -<br />
  cd apache_1.3.33<br />
  ./configure<br />
  cd ../mod_perl-1.29</kbd>
</li>
<li>
  <p>Build the perl makefile:</p>
  <kbd>perl Makefile.PL APACHE_SRC=../apache_1.3.33/src \<br />
&nbsp; DO_HTTPD=1 USE_APACI=1 EVERYTHING=1</kbd>
</li>
<li>
  <p>Run the 'make' utility:</p>
  <kbd>make</kbd>
</li>
<li>
  <p>...and install</p>
  <kbd>make install</kbd>
</li>
<li>
  <p>Return to the Apache install directory:</p>
  <kbd>cd ../apache_1.3.33</kbd>
</li>
<li>
  <p>...and run the Apache configure script:</p>
  <kbd>./configure --enable-module=include \<br />
&nbsp; --activate-module=src/modules/perl/libperl.a \<br />
&nbsp; --enable-module=perl</kbd>
</li>
<li>
  <p>Run 'make', and install:</p>
  <kbd>make<br />
  make install</kbd>
</li>
</ol>

<h3 class="boxed">Perl modules</h3>
<p>The Ensembl website needs quite a few Perl modules to be installed in order for it to run. These
modules can all be downloaded from www.cpan.org, and are all installed in much the same way:
Download the module tarball, unpack in a working directory, and install the module:</p>
<kbd>
gunzip < module.tar.gz | tar xvf -<br />
cd module<br />
perl Makefile.PL<br />
make<br />
make test<br />
make install<br />
</kbd>
<p>The modules that are required are listed below, along with their URLs. The file part of the URL is
current at the time of writing - you should install whatever is the latest version of the module.</p>

<table>
  <tr>
    <th>CGI</th>
    <td><p>Enables Perl scripts to easily parse cgi</p>
    <p><a href="http://www.cpan.org/modules/by-module/CGI/CGI.pm-2.99.tar.gz">http://www.cpan.org/modules/by-module/CGI/CGI.pm-2.99.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Compress::Zlib</th>
    <td><p>A compression module for DAS.</p>
      <p><a href="http://www.cpan.org/modules/by-module/Compress/Compress-Zlib-1.34.tar.gz">http://www.cpan.org/modules/by-module/Compress/Compress-Zlib-1.34.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>DBI</th>
    <td><p>A common database interface for Perl</p>
      <p><a href="http://www.cpan.org/modules/by-module/DBI/DBI-1.47.tar.gz">http://www.cpan.org/modules/by-module/DBI/DBI-1.47.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>DBD::Mysql</th>
    <td><p>The MySQL drivers for the DBI interface</p>
      <p><a href="http://www.cpan.org/modules/by-module/DBD/Msql-Mysql-modules-1.2219.tar.gz">http://www.cpan.org/modules/by-module/DBD/Msql-Mysql-modules-1.2219.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>GD</th>
    <td><p>A Graphics library</p>
      <p>Note: may require additional modules. Please read install docs.</p>
      <p><a href="http://www.cpan.org/modules/by-module/GD/GD-2.21.tar.gz">http://www.cpan.org/modules/by-module/GD/GD-2.21.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>ParallelUserAgent</th>
    <td><p>Allows for parallel requests</p>
      <p><a href="http://www.cpan.org/modules/by-module/LWP/ParallelUserAgent-2.57.tar.gz">http://www.cpan.org/modules/by-module/LWP/ParallelUserAgent-2.57.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Bio::DasLite</th>
    <td><p>Lightweight DAS fetcher</p>
      <p><a href="http://www.cpan.org/modules/by-module/Bio/Bio-DasLite-0.10.tar.gz">http://www.cpan.org/modules/by-module/Bio/Bio-DasLite-0.10.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Data::UUID</th>
    <td><p>Creates a unique ID</p>
      <p><a href="http://www.cpan.org/modules/by-module/Data/Data-UUID-0.11.tar.gz">http://www.cpan.org/modules/by-module/Data/Data-UUID-0.11.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Digest::MD5</th>
    <td><p>calculates a unique checksum for a file</p>
      <p><a href="http://www.cpan.org/modules/by-module/Digest/Digest-MD5-2.33.tar.gz">http://www.cpan.org/modules/by-module/Digest/Digest-MD5-2.33.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Storable</th>
    <td><p>used to store and restore data structures (part of standard Perl distribution)</p>
      <p><a href="http://www.cpan.org/modules/by-module/Storable/Storable-2.13.tar.gz   ">http://www.cpan.org/modules/by-module/Storable/Storable-2.13.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>LWP</th>
    <td><p>used by DAS to communicate with DAS sources</p>
      <p>Note: may require additional modules. Please read install docs.</p>
      <p><a href="http://www.cpan.org/modules/by-module/LWP/libwww-perl-5.803.tar.gz">http://www.cpan.org/modules/by-module/LWP/libwww-perl-5.803.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>SOAP::Lite</th>
    <td><p>used by DAS to communicate with DAS sources</p>
      <p>Note: may require additional modules. Please read install docs.</p>
      <p><a href="http://www.cpan.org/modules/by-module/SOAP/SOAP-Lite-0.60a.tar.gz">http://www.cpan.org/modules/by-module/SOAP/SOAP-Lite-0.60a.tar.gz</a></p>
  </tr>
  <tr>
    <th>XML::Parser</th>
    <td><p>used by DAS to parse DAS sources</p>
    <p>Note: module is a wrapper around the expat library, which should be installed
first:</p>
<p>Download the expat source from http://sourceforge.net/projects/expat/. Get the
latest version (currently expat-1.95.8.tar.gz). Run the following commands:</p>
  <kbd>
gunzip < expat-1.95.8.tar.gz | tar xvf -<br />
cd expat-1.95.8<br />
./configure<br />
make<br />
make install
</kbd>
    <p><a href="http://www.cpan.org/modules/by-module/XML/XML-Parser-2.34.tar.gz">http://www.cpan.org/modules/by-module/XML/XML-Parser-2.34.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>XML::Simple</th>
    <td><p>used by Mart to parse configuration</p>
      <p><a href="http://www.cpan.org/modules/by-module/XML/XML-Simple-2.14.tar.gz">http://www.cpan.org/modules/by-module/XML/XML-Simple-2.14.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Parse-RecDescent</th>
    <td><p>used by Excel exporter</p>
      <p><a href="http://www.cpan.org/modules/by-module/Parse/Parse-RecDescent-1.94.tar.gz">http://www.cpan.org/modules/by-module/Parse/Parse-RecDescent-1.94.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>PDF::API2</th>
    <td><p>used by Image exporter for exporting as PDF</p>
      <p><a href="http://www.cpan.org/modules/by-module/PDF/PDF-API2-0.40.86.tar.gz">http://www.cpan.org/modules/by-module/PDF/PDF-API2-0.40.86.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Spreadsheet::WriteExcel</th>
    <td><p>used for exporting Excel spreadsheets</p>
      <p><a href="http://www.cpan.org/modules/by-module/Spreadsheet/Spreadsheet-WriteExcel-2.12.tar.gz">http://www.cpan.org/modules/by-module/Spreadsheet/Spreadsheet-WriteExcel-2.12.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>OLE::Storage_Lite</th>
    <td><p>Used for exporting Excel spreadsheets</p>
      <p><a href="http://www.cpan.org/modules/by-module/OLE/OLE-Storage_Lite-0.11.tar.gz">http://www.cpan.org/modules/by-module/OLE/OLE-Storage_Lite-0.11.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Time::HiRes</th>
    <td><p>Used for code profiling</p>
      <p><a href="http://www.cpan.org/modules/by-module/Time/Time-HiRes-1.66.tar.gz">http://www.cpan.org/modules/by-module/Time/Time-HiRes-1.66.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>HTML::Template</th>
    <td><p>Used by MartView</p>
      <p><a href="http://www.cpan.org/modules/by-module/HTML/HTML-Template-2.7.tar.gz">http://www.cpan.org/modules/by-module/HTML/HTML-Template-2.7.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>File::Temp</th>
    <td><p>Used by MartView</p>
      <p><a href="http://www.cpan.org/modules/by-module/File/File-Temp-0.14.tar.gz">http://www.cpan.org/modules/by-module/File/File-Temp-0.14.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Mail::Mailer</th>
    <td><p>Used by MartView</p>
      <p><a href="http://www.cpan.org/modules/by-module/Mail/MailTools-1.66.tar.gz">http://www.cpan.org/modules/by-module/Mail/MailTools-1.66.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Math::Bezier</th>
    <td><p>Used by drawing code</p>
      <p><a href="http://www.cpan.org/modules/by-module/Math/Math-Bezier-0.01.tar.gz">http://www.cpan.org/modules/by-module/Math/Math-Bezier-0.01.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>IO::String</th>
    <td><p>Used for sequence handling</p>
      <p><a href="http://www.cpan.org/modules/by-module/IO/IO-String-1.06.tar.gz">http://www.cpan.org/modules/by-module/IO/IO-String-1.06.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>Image::Size</th>
    <td><p>Used for getting size of images</p>
      <p><a href="http://www.cpan.org/modules/by-module/Image/Image-Size-2.992.tar.gz">http://www.cpan.org/modules/by-module/Image/Image-Size-2.992.tar.gz</a></p></td>
  </tr>
  <tr>
    <th>PathTools</th>
    <td><p>Used for architecture independent file path manipulation</p>
      <p><a href="http://www.cpan.org/modules/by-module/File/PathTools-3.09.tar.gz">http://www.cpan.org/modules/by-module/File/PathTools-3.09.tar.gz</a></p></td>
  </tr>
</table>


<h3 class="boxed"><a name="dotter">Dotter</a></h3>
<p>The Ensembl website contains a page called DotterView which displays a graphical dotplot comparison
of two sequences, using the application Dotter. If you wish to use this page on a local installation you
will need a local Dotter binary:</p>
<p>Dotter is part of the AceDB package available at:</p>
<blockquote><a href="http://www.acedb.org/Software/Downloads/supported.shtml" target="_blank">http://www.acedb.org/Software/Downloads/supported.shtml</a></blockquote>
<p>Pre-built binaries are available for alpha/Tru64 unix and intel/linux. All other flavours will have to be built
from sources.</p>
<p>The acedb code makes use of the GTK graphics and GNU readline packages. If you don't have these
on your system you can either install them yourself or use the copies distributed with the acedb source.</p>
<p>To install them go to</p>
  <a target="_blank" href="http://www.gtk.org">www.gtk.org</a>
<p>for the GTK package and</p>
  <a href="ftp://ftp.cwru.edu/pub/bash/readline-4.3.tar.gz">ftp://ftp.cwru.edu/pub/bash/readline-4.3.tar.gz</a>
<p>for readline.</p>
<p>Follow the GTK and GNU instructions for installing these packages.</p>

<p>If you want to use the versions distributed with the acedb source then follow the instructions for making
them in the next section.</p>
<p>Next ungzip/untar the acedb package into an temporary directory:</p>
<kbd>mkdir acedb<br />
mv ACEDB-source.4_9l.tar.gz ./acedb/<br />
cd ./acedb<br />
gzip -d < ACEDB-source.4_9l.tar.gz | tar xvf -<br />
</kbd>
<p>Set an environment variable to tell the build process what platform you are building on - one of:</p>
<kbd>ALPHA_4 ALPHA_4_GCC ALPHA_4_GCC_OPT ALPHA_4_LINUX ALPHA_4_OLDSTYLE ALPHA_4_OPT ALPHA_5 ALPHA_CHRONO_4<br />
DARWIN_4<br />
FreeBSD<br />
FUJITSU_4<br />
HP_4 HP_4_GCC HP_4_OPT<br />
IBM_4_3 IBM_4<br />
IRIX4_4<br />
LINUX_4 LINUX_MAC_4<br />
MACOSX_4_DEF<br />
NEC_4 NEC_4_R10 NEC_4_R11<br />
NEXT_4<br />
POSIX_4 POSIX_4_GCC<br />
SGI_4 SGI_4_GCC SGI_4_IRIX5 SGI_4_PURE SGI_5_GCC SGI_65_GCC_DEF
SOLARIS_4 SOLARIS_4_OPT SOLARIS_4_RELEASE SUN_4_DEF
WIN32_4</kbd>
<p>For example (in csh):</p>
<kbd>setenv ACEDB_MACHINE SOLARIS_4</kbd>
<p>Consult the AceDB documentation if you need more help.</p>

<p>If you wish to use the GTK/readline packages distributed with acedb you should build them before
building any acedb code:</p>
<kbd>make gnulibs</kbd>
<p>To build the dotter binary issue the command:</p>
<kbd>make dotter</kbd>
<p>(If you just run "make" you will build the whole of AceDB - which is unnecessary).</p>
<p>The build will place the executable at:</p>
<kbd>./bin.${ACEDB_MACHINE}/dotter</kbd>
<p>Check that the binary runs properly and then copy it to your Ensembl shared binaries directory.
DotterView should now work.</p>

<h3 class="boxed"><a name="ensembl">Ensembl Build/Install</a></h3>
<p>This section explains how to install the <a href="#ensembldata">Ensembl data</a>, <a href="#ensemblcode">Perl modules, and web code</a>. It also covers the installation of BioPerl.</p>

<h4>Versioning</h4>

<p>Each Ensembl release has an integer version number associated with it, and that version number is
used to identify the correct versions of API, Web code and databases (see below) that make up that
release.</p>
<p>For the API & Web code, a CVS branch (essentially a named snapshot of the code) is made for each
release, named with the release version number. The current release is version 39, and the CVS tag for
identifying the API & Web code for this release is 'branch-ensembl-39'.</p>

<p>The Ensembl database names consist of the species, the database type, the release number, and the
data version. The current human 'core' database is named homo_sapiens_core_39_36a. i.e. a
human core database, release 39, data version 36a built on the NCBI 36 assembly. Subsequent data
releases on the same assembly are suffixed with a lower case letter (a, b, etc.).</p>

<p>The idea is that components with the same release version should work together - i.e. any web site built
with 'version 31' API and web code, and 'version 31' databases should work correctly.</p>

<h3 class="boxed"><a name="ensembldata">Installing the Ensembl Data</a></h3>

<p>The Ensembl data is provided on the Ensembl FTP site in the form of tab-delimited text files for
importing into MySQL. ftp://ftp.ensembl.org/pub contains a directory for each release for each species.
The latest versions are named current_species, e.g. current_human, current_mouse, etc. The directory
structure below this is as follows (using current_human as an example):</p>

<table>
<tr>
  <td colspan="4">ftp.ensembl.org/pub/current_human</td>
</tr>
<tr>
  <td style="width: 3em" rowspan="5">&nbsp; &nbsp;</td>
  <td colspan="2">|--&nbsp;data</td>
</tr>
<tr><td>|--&nbsp;fasta</td><td>cDNA, DNA (masked and unmasked chromosome sequence dumps), RNA and peptide dumps</td></tr>
<tr><td>|--&nbsp;flatfiles</td><td>EMBL and GenBank format dumps</td></tr>
<tr><td>|--&nbsp;mysql</td><td>Database dumps</td></tr>
</table>
<p>
The mysql directory contains a directory for each database.  This can be used to install your own copy of the Ensembl data, e.g.: 
</p>
<table>
<tr>
  <td colspan="3">ftp.ensembl.org/pub/current_human/data/mysql</td>
</tr>
<tr>
  <td style="width: 3em" rowspan="3">&nbsp; &nbsp;</td>
  <td>|--&nbsp;homo_sapiens_core_39_36a</td>
  <td>core ensembl database</td>
</tr>
<tr>
  <td>|--&nbsp;homo_sapiens_otherfeatures_39_36a</td>
  <td>other features database</td>
</tr>
<tr>
  <td>|--&nbsp;etc...</td>
</tr>
</table>
<p>Each database directory contains a data file for each table in that database an SQL file that contains the SQL commands necessary to build that database's table structure and a checksum file (using a UNIX "sum" utility) so you can verify that the
data has downloaded correctly.</p>

<h4>Multi-species data</h4>

<p>In addition to the individual species data, there is a directory containing multi-species databases - i.e.
databases which either affect the site as a whole (e.g. ensembl_web_user_db), or which describe metainformation
about multiple species (e.g. ensembl_compara).</p>
<table>
<tr>
  <td colspan="3">ftp.ensembl.org/pub/current_multispecies/data/mysql</td>
</tr>
<tr>
  <td style="width: 3em" rowspan="4">&nbsp; &nbsp;</td>
  <td>|--&nbsp;ensembl_compara_39</td>
  <td>Comparative genomics database</td>
</tr>
<tr>
  <td>|--&nbsp;ensembl_website_39</td>
  <td>Help database</td>
</tr>
<tr>
  <td>|--&nbsp;ensembl_web_user_db_39</td>
  <td>Web preference database</td>
</tr>
<tr>
  <td>|--&nbsp;etc...</td>
</tr>
</table>

<p>Regardless of which species you choose to install, for a full installation you will probably want to install
the multi-species databases as well. i.e. compara, help and web_user_db.</p>

<p>NB: The FTP site will ideally be laid out as described. If, however, for reasons of space or
maintainability, files are not located as described then check the ftp site for README files which should
explain where the data can be found.</p>

<h4>BioMart data</h4>
<p>The remaining multi-species databases can be downloaded from:</p>
<table>
<tr>
  <td colspan="3">ftp.ensembl.org/pub/current_mart/data/mysql</td>
</tr>
<tr>
  <td style="width: 3em" rowspan="4">&nbsp; &nbsp;</td>
  <td>|--&nbsp;ensembl_mart_39</td>
</tr>
<tr>
  <td>|--&nbsp;sequence_mart_39</td>
</tr>
<tr>
  <td>|--&nbsp;snp_mart_39</td>
</tr>
<tr>
  <td>|--&nbsp;vega_mart_39</td>
</tr>
</table>

<h4>To install the Ensembl Data:</h4>
<ol>
<li>
<p>Download the directories in ftp.ensembl.org/pub/current_organism/data/mysql for
whatever organism you want to install. Note that the ensembl directory contains several files for the
DNA and feature tables - these are very large tables, so the dump file is split into smaller chunks for
easier downloading.</p>
</li>
<li>
<p>Each table file is gzipped so unpack the data into working directories, keeping separate directories for
each database.</p>
<p>For each database you have downloaded, cd into the database directory and perform steps 3-5. For
illustration, we will use homo_sapiens_core_39_36a as the database - you need to change this
appropriately for each database you install. Remember, you also need to download and install the
multi-species databases.</p>
</li>
<li>
<p>Start a MySQL console session (see the Installing MySQL section above if necessary) and issue the
command:</p>
<kbd>create database homo_sapiens_core_39_36a</kbd>
</li>
<li>
<p>Exit the console session, and issue the following command to run the ensembl SQL file, which should
be in the directory where you unpacked the downloaded data. This creates the schema for the empty
database you created in step 3.</p>
<p>Note that we are using the example MySQL settings of /data/mysql as the install directory, and
mysqldba as the database user. Note that here mysqldba is a MySQL account with file access to the
database, which is not the same as a system user. See the MySQL documentation for instructions
on creating/administering users.</p>
<kbd>/data/mysql/bin/mysql -u mysqldba homo_sapiens_core_39_36a < homo_sapiens_core_39_36a.sql</kbd>
</li>
<li>
<p>Load the data into the database structure you have just built with the following command.</p>
<kbd>/data/mysql/bin/mysqlimport -u mysqldba homo_sapiens_core_39_36a -L *.txt.table</kbd>
</li>
</ol>

<p>You have now created and loaded the core Ensembl database for human.</p>

<p>Note that all the databases except the ensembl_web_user_db database only require read access for the
website to work. The ensembl_web_user_db requires a MySQL user with delete/insert/update
permissions. Also note that because its the only database that the website writes data into, the
ensembl_web_user_db has no .table (data) files to download.</p>

<p>NB MySQL needs quite a lot of temporary space to load the databases. It is quite possible that your /
tmp directory (which MySQL uses by default) is too small, in which case you might see an Error 28 (use
the MySQL tool perror to see what these error numbers mean). Fortunately, you can force MySQL to
write temporary files to another location. See the MySQL docs for details:

<a href="http://www.mysql.com/doc/T/e/Temporary_files.html" target="_blank">http://www.mysql.com/doc/T/e/Temporary_files.html</a>. The simplest solution is to start mysqld with the
argument --tmpdir my_spacious_tmp_location.</p>

<h4>GO data</h4>
<p>The Ensembl ftp site now includes a copy of the GO database as ensembl_go_39. Install this if you
want local GO information.</p>

<h3 class="boxed"><a name="ensemblcode">Installing the Ensembl, BioPerl, and BioMart modules</a></h3>

<p>If you review the Site Structure part of this document, you will recall that the site is based around a
single server-root directory. The Ensembl, BioPerl and BioMart modules are all installed into this
directory. Choose a suitable location, and create your server-root directory. For the purposes of
illustration, we will use /usr/local/ensembl. When following these instructions, replace /
usr/local/ensembl with your chosen server-root.</p>

<ol>
<li><p>Go to the server-root directory:</p>
<kbd>cd /usr/local/ensembl</kbd>
</li>
</ol>
<h3 class="boxed"><a id="cvs_co">To install the Ensembl modules</a></h3>
<ol start="2">
<li>
<p>Log into the Sanger CVS server (using a password of "CVSUSER"):</p>
<kbd>cvs -d :pserver:cvsuser@cvsro.sanger.ac.uk:/cvsroot/CVSmaster login</kbd>
</li>
<li>
<p>To check out the most recent stable version of the ensembl API and web code from CVS (i.e. to download the code from the Sanger CVS server onto your local machine) you need to use the latest branch of the code.  Please note the code on the CVS HEAD is under development and unstable.  Use the following command making sure you use the code that matches your databases:</p>
<kbd>cvs -d :pserver:cvsuser@cvsro.sanger.ac.uk:/cvsroot/CVSmaster \<br />
&nbsp; co -r branch-ensembl-39 ensembl-api ensembl-website</kbd>
<p>A listing of your server-root should now look something like:</p>
<table>
<tr>
  <td>
    conf/<br />
    ensembl/<br />
    ensembl-compara/
  </td>
  <td>
    ensembl-draw/<br />
    ensembl-external/<br />
    ensembl-variation/
  </td>
  <td>
    htdocs/<br />
    modules/<br />
    perl/
  </td>
  <td>
    public-plugins/<br />
    utils/
  </td>
</tr>
</table>
</li>
</ol>
<h3 class="boxed"><a id="cvs_cobiomart">To install the BioMart modules</a> (BioMart 0.4)</h3>
<ol start="4">
<li>
<p>Check the BioMart code out of CVS (downloads the code from the Sanger CVS server onto your local
machine):</p>
<kbd>cvs -d :pserver:cvsuser@cvsro.sanger.ac.uk:/cvsroot/CVSmaster \<br />
&nbsp; co -r release-0_4 biomart-plib biomart-web</kbd>
</li>
</ol>

<h3 class="boxed">To install the BioPerl modules (BioPerl 1.2.3)</h3>
<ol start="5">
<li>
  <p>Log into the BioPerl CVS server (using a password of: cvs):</p>
  <kbd>cvs -d :pserver:cvs@cvs.open-bio.org:/home/repository/bioperl login</kbd>
</li>
<li>
  <p>Check-out the BioPerl code:</p>
<kbd>cvs -d :pserver:cvs@cvs.open-bio.org:/home/repository/bioperl \<br />
&nbsp; co -r bioperl-release-1-2-3 bioperl-live</kbd>
<p>A listing of your server-root should now look something like:</p>
<table>
<tr>
  <td>
    bioperl-live/<br />
    biomart-plib/<br />
    biomart-web/<br />
    conf/
  </td>
  <td>
    ensembl/<br />
    ensembl-compara/<br />
    ensembl-draw/<br />
    ensembl-external/
  </td>
  <td>
    ensembl-variation/<br />
    htdocs/<br />
    modules/<br />
    perl/
  </td>
  <td>
    public-plugins/<br />
    utils/
  </td>
</tr>
</table>
</li>
</ol>
<p>You should now have all the Ensembl website code and data installed and ready to configure.</p>

</body>
</html>
  
