<html>
  <head>
    <meta name="navigation" content="Build" />
    <title>Building/Installing an Ensembl Website</title>
  </head>
<body> 
  <h2>Building/Installing an Ensembl Website</h2>

<p>
The build instructions are split into three sections:
</p>
<ul>
  <li><p><a href="non-ensembl-code.html">How to install other applications required by Ensembl</a><br />
    CVS, Perl, MySQL, Apache 2, Dotter</p></li>
  <li><p><a href="ensembl-code.html">How to install the Ensembl and BioMart software</a></p></li>
  <li><p><a href="ensembl-data.html">How to install the Ensembl data</a></p></li>
</ul>
<p>
  The first of these involves the installation of applications/modules that are not part of the Ensembl
  project but are necessary for the web site to work, such as Apache and MySQL. The second involves 
  installing the Ensembl data, modules and web site code.
</p>

<h4 class="boxed">Ensembl and MySQL</h4>
<p>
  PLEASE NOTE - even if you do not wish to download all the Ensembl data - for the website to function
  you will need:
</p>
<ul>
  <li>a local database server to manage display settings and user accounts (if these are enabled).</li>
  <li>the mysql client software installed on your machine.</li>
</ul>
<p>
  If you do not wish to run MySQL (or cannot run a MySQL server locally) <strong>AND</strong> you do not
  wish to enable user accounts then you can alternatively install SQLite as an alternative server.
</p>

<h4 class="boxed">Ensembl Version Numbers...</h4>
<p>
  Each Ensembl release has an integer version number associated with it, and that version number is
  used to identify the correct versions of API, Web code and databases (see below) that make up that
  release.
</p>
<p>
  For the API and Web code, a CVS branch (essentially a named snapshot of the code) is made for each
  release, named with the release version number. The current release is version [[SPECIESDEFS::ENSEMBL_VERSION]],
  and the CVS tag for identifying the API and Web code for this release is 'branch-ensembl-[[SPECIESDEFS::ENSEMBL_VERSION]]'.
</p>

<p>
  The Ensembl database names consist of the species, the database type, the release number, and the
  data version. The current human 'core' database is named homo_sapiens_core_[[SPECIESDEFS::ENSEMBL_VERSION]]_[[SPECIESINFO::Homo_sapiens:SPECIES_RELEASE_VERSION]]. i.e. a
  human core database, release [[SPECIESDEFS::ENSEMBL_VERSION]], data version [[SPECIESINFO::Homo_sapiens:SPECIES_RELEASE_VERSION]] built on the NCBI 36 assembly. Subsequent data
  releases on the same assembly are suffixed with a lower case letter (a, b, etc.).
</p>

<p>
  The idea is that components with the same release version should work together - i.e. any web site built
  with 'version 31' API and web code, and 'version 31' databases should work correctly.
</p>

<h3 class="boxed"><a name="#non-ensembl-code">Non-Ensembl Applications Build/Install</a></h3>
<p>Ensembl is built on the following applications:</p>
<ul>
  <li><a href="non-ensembl-code.html#perl">Perl</a> - Open source scripting language</li>
  <li><a href="non-ensembl-code.html#apache">Apache and mod_perl</a> - Open source web server</li>
  <li><a href="non-ensembl-code.html#mysql">MySQL</a> - Open source database server</li>
  <li><a href="non-ensembl-code.html#dotter">Dotter</a> - Graphical alignment package</li>
  <li><a href="non-ensembl-code.html#cvs">CVS</a> - versioning system to allow you to download the Ensembl code</li>
  <li><a href="non-ensembl-code.html#sqlite">SQLite</a> - alternative database server to store session data if not installing MySQL.</li>
</ul>
<p>These applications are not version-specific for Ensembl; that is, if you upgrade your Ensembl installation
to a newer version when one becomes available, you probably won't need to install new versions of
these applications.</p>

<p>All of this software, like all of Ensembl, is Open Source software and can be downloaded and used free
of charge. You should, however, check the documentation for each application to see what license it
has been released under, particularly if you are installing Ensembl in a commercial environment.</p>

<p><a href="non-ensembl-code.html">Detailed installation instructions</a> are available for each of these
packages</p>

<h3 class="boxed"><a name="ensembl-code">Installing Ensembl/BioMart Code and Data</a></h3>

<p>This section explains how to install the <a href="#ensembldata">Ensembl data</a>, <a href="#ensemblcode">Perl modules, and web code</a>. It also covers the installation of BioPerl.</p>

<h3 class="boxed"><a name="ensembldata">Installing the Ensembl Data</a></h3>

<p>
  The Ensembl data is provided on the Ensembl FTP site in the form of tab-delimited text files for
  importing into MySQL. ftp://ftp.ensembl.org/pub contains a directory for each release for each species.
  The latest versions are named current_species, e.g. current_human, current_mouse, etc. The directory
  structure below this is as follows (using current_human as an example):
</p>

<table>
<tr>
  <td colspan="4">ftp.ensembl.org/pub/current_human</td>
</tr>
<tr>
  <td style="width: 3em" rowspan="5">&nbsp; &nbsp;</td>
  <td colspan="2">|--&nbsp;data</td>
</tr>
<tr><td>|--&nbsp;fasta</td><td>cDNA, DNA (masked and unmasked chromosome sequence dumps), RNA and peptide dumps</td></tr>
<tr><td>|--&nbsp;flatfiles</td><td>EMBL and GenBank format dumps</td></tr>
<tr><td>|--&nbsp;mysql</td><td>Database dumps</td></tr>
</table>
<p>
The mysql directory contains a directory for each database.  This can be used to install your own copy of the Ensembl data, e.g.: 
</p>
<table>
<tr>
  <td colspan="3">ftp.ensembl.org/pub/current_human/data/mysql</td>
</tr>
<tr>
  <td style="width: 3em" rowspan="3">&nbsp; &nbsp;</td>
  <td>|--&nbsp;homo_sapiens_core_[[SPECIESDEFS::ENSEMBL_VERSION]]_[[SPECIESINFO::Homo_sapiens:SPECIES_RELEASE_VERSION]]</td>
  <td>core ensembl database</td>
</tr>
<tr>
  <td>|--&nbsp;homo_sapiens_otherfeatures_[[SPECIESDEFS::ENSEMBL_VERSION]]_[[SPECIESINFO::Homo_sapiens:SPECIES_RELEASE_VERSION]]</td>
  <td>other features database</td>
</tr>
<tr>
  <td>|--&nbsp;etc...</td>
</tr>
</table>
<p>Each database directory contains a data file for each table in that database an SQL file that contains the SQL commands necessary to build that database's table structure and a checksum file (using a UNIX "sum" utility) so you can verify that the
data has downloaded correctly.</p>

<h4>Multi-species data</h4>

<p>In addition to the individual species data, there is a directory containing multi-species databases - i.e.
databases which either affect the site as a whole (e.g. ensembl_web_user_db), or which describe metainformation
about multiple species (e.g. ensembl_compara).</p>
<table>
<tr>
  <td colspan="3">ftp.ensembl.org/pub/current_multispecies/data/mysql</td>
</tr>
<tr>
  <td style="width: 3em" rowspan="4">&nbsp; &nbsp;</td>
  <td>|--&nbsp;ensembl_compara_[[SPECIESDEFS::ENSEMBL_VERSION]]</td>
  <td>Comparative genomics database</td>
</tr>
<tr>
  <td>|--&nbsp;ensembl_website_[[SPECIESDEFS::ENSEMBL_VERSION]]</td>
  <td>Help database</td>
</tr>
<tr>
  <td>|--&nbsp;ensembl_web_user_db_[[SPECIESDEFS::ENSEMBL_VERSION]]</td>
  <td>Web preference database</td>
</tr>
<tr>
  <td>|--&nbsp;etc...</td>
</tr>
</table>

<p>Regardless of which species you choose to install, for a full installation you will probably want to install
the multi-species databases as well. i.e. compara, help and web_user_db.</p>

<p>NB: The FTP site will ideally be laid out as described. If, however, for reasons of space or
maintainability, files are not located as described then check the ftp site for README files which should
explain where the data can be found.</p>

<h4>BioMart data</h4>
<p>The remaining multi-species databases can be downloaded from:</p>
<table>
<tr>
  <td colspan="3">ftp.ensembl.org/pub/current_mart/data/mysql</td>
</tr>
<tr>
  <td style="width: 3em" rowspan="4">&nbsp; &nbsp;</td>
  <td>|--&nbsp;ensembl_mart_[[SPECIESDEFS::ENSEMBL_VERSION]]</td>
</tr>
<tr>
  <td>|--&nbsp;sequence_mart_[[SPECIESDEFS::ENSEMBL_VERSION]]</td>
</tr>
<tr>
  <td>|--&nbsp;snp_mart_[[SPECIESDEFS::ENSEMBL_VERSION]]</td>
</tr>
<tr>
  <td>|--&nbsp;vega_mart_[[SPECIESDEFS::ENSEMBL_VERSION]]</td>
</tr>
</table>

<h4>To install the Ensembl Data:</h4>
<ol>
<li>
<p>Download the directories in ftp.ensembl.org/pub/current_organism/data/mysql for
whatever organism you want to install. Note that the ensembl directory contains several files for the
DNA and feature tables - these are very large tables, so the dump file is split into smaller chunks for
easier downloading.</p>
</li>
<li>
<p>Each table file is gzipped so unpack the data into working directories, keeping separate directories for
each database.</p>
<p>For each database you have downloaded, cd into the database directory and perform steps 3-5. For
illustration, we will use homo_sapiens_core_[[SPECIESDEFS::ENSEMBL_VERSION]]_[[SPECIESINFO::Homo_sapiens:SPECIES_RELEASE_VERSION]] as the database - you need to change this
appropriately for each database you install. Remember, you also need to download and install the
multi-species databases.</p>
</li>
<li>
<p>Start a MySQL console session (see the Installing MySQL section above if necessary) and issue the
command:</p>
<kbd>create database homo_sapiens_core_[[SPECIESDEFS::ENSEMBL_VERSION]]_[[SPECIESINFO::Homo_sapiens:SPECIES_RELEASE_VERSION]]</kbd>
</li>
<li>
<p>Exit the console session, and issue the following command to run the ensembl SQL file, which should
be in the directory where you unpacked the downloaded data. This creates the schema for the empty
database you created in step 3.</p>
<p>Note that we are using the example MySQL settings of /data/mysql as the install directory, and
mysqldba as the database user. Note that here mysqldba is a MySQL account with file access to the
database, which is not the same as a system user. See the MySQL documentation for instructions
on creating/administering users.</p>
<kbd>/data/mysql/bin/mysql -u mysqldba homo_sapiens_core_[[SPECIESDEFS::ENSEMBL_VERSION]]_[[SPECIESINFO::Homo_sapiens:SPECIES_RELEASE_VERSION]] < homo_sapiens_core_[[SPECIESDEFS::ENSEMBL_VERSION]]_[[SPECIESINFO::Homo_sapiens:SPECIES_RELEASE_VERSION]].sql</kbd>
</li>
<li>
<p>Load the data into the database structure you have just built with the following command.</p>
<kbd>/data/mysql/bin/mysqlimport -u mysqldba homo_sapiens_core_[[SPECIESDEFS::ENSEMBL_VERSION]]_[[SPECIESINFO::Homo_sapiens:SPECIES_RELEASE_VERSION]] -L *.txt.table</kbd>
</li>
</ol>

<p>You have now created and loaded the core Ensembl database for human.</p>

<p>Note that all the databases except the ensembl_web_user_db database only require read access for the
website to work. The ensembl_web_user_db requires a MySQL user with delete/insert/update
permissions. Also note that because its the only database that the website writes data into, the
ensembl_web_user_db has no .table (data) files to download.</p>

<p>NB MySQL needs quite a lot of temporary space to load the databases. It is quite possible that your /
tmp directory (which MySQL uses by default) is too small, in which case you might see an Error 28 (use
the MySQL tool perror to see what these error numbers mean). Fortunately, you can force MySQL to
write temporary files to another location. See the MySQL docs for details:

<a href="http://www.mysql.com/doc/T/e/Temporary_files.html" target="_blank">http://www.mysql.com/doc/T/e/Temporary_files.html</a>. The simplest solution is to start mysqld with the
argument --tmpdir my_spacious_tmp_location.</p>

<h4>GO data</h4>
<p>The Ensembl ftp site now includes a copy of the GO database as ensembl_go_[[SPECIESDEFS::ENSEMBL_VERSION]]. Install this if you
want local GO information.</p>

<h3 class="boxed"><a name="ensemblcode">Installing the Ensembl, BioPerl, and BioMart modules</a></h3>

<p>If you review the Site Structure part of this document, you will recall that the site is based around a
single server-root directory. The Ensembl, BioPerl and BioMart modules are all installed into this
directory. Choose a suitable location, and create your server-root directory. For the purposes of
illustration, we will use /usr/local/ensembl. When following these instructions, replace /
usr/local/ensembl with your chosen server-root.</p>

<ol>
<li><p>Go to the server-root directory:</p>
<kbd>cd /usr/local/ensembl</kbd>
</li>
</ol>
<h3 class="boxed"><a id="cvs_co">To install the Ensembl modules</a></h3>
<ol start="2">
<li>
<p>Log into the Sanger CVS server (using a password of "CVSUSER"):</p>
<pre class="code">
$ cvs -d :pserver:cvsuser@cvs.sanger.ac.uk:/cvsroot/CVSmaster  login
Logging in to :pserver:cvsuser@cvs.sanger.ac.uk:2401/cvsroot/CVSmaster
CVS password: CVSUSER 
</pre>
</li>
<li>
<p>To check out the most recent stable version of the ensembl API and web code from CVS (i.e. to download the code from the Sanger CVS server onto your local machine) you need to use the latest branch of the code.  Please note the code on the CVS HEAD is under development and unstable.  Use the following command making sure you use the code that matches your databases:</p>
<pre class="code">
$ cvs -d :pserver:cvsuser@cvs.sanger.ac.uk:/cvsroot/CVSmaster co -r branch-ensembl-[[SPECIESDEFS::ENSEMBL_VERSION]] ensembl-api ensembl-website 
</pre>

<p>A listing of your server-root should now look something like:</p>
<table>
<tr>
  <td>
    conf/<br />
    ensembl/<br />
    ensembl-compara/
  </td>
  <td>
    ensembl-draw/<br />
    ensembl-external/<br />
    ensembl-variation/
  </td>
  <td>
    htdocs/<br />
    modules/<br />
    perl/
  </td>
  <td>
    public-plugins/<br />
    utils/
  </td>
</tr>
</table>
</li>
</ol>
<h3 class="boxed"><a id="cvs_cobiomart">To install the BioMart modules</a> (BioMart 0.5)</h3>
<ol start="4">
<li>
<p>Check the BioMart code out of CVS (downloads the code from the Sanger CVS server onto your local
machine):</p>
<pre class="code">
$ cvs -d :pserver:cvsuser@cvs.sanger.ac.uk:/cvsroot/CVSmaster co -r release-0_5 biomart-perl</pre>
</li>
</ol>

<p>More information on  <a href="biomart.html">tweaking and configuring BioMart to work within the Ensembl website</a> is available</p>

<h3 class="boxed">To install the BioPerl modules (BioPerl 1.2.3)</h3>
<ol start="5">
<li>
  <p>Log into the BioPerl CVS server (using a password of: cvs):</p>
  <kbd>cvs -d :pserver:cvs@cvs.open-bio.org:/home/repository/bioperl login</kbd>
</li>
<li>
  <p>Check-out the BioPerl code:</p>
<kbd>cvs -d :pserver:cvs@cvs.open-bio.org:/home/repository/bioperl \<br />
&nbsp; co -r bioperl-release-1-2-3 bioperl-live</kbd>
<p>A listing of your server-root should now look something like:</p>
<table>
<tr>
  <td>
    bioperl-live/<br />
    biomart-plib/<br />
    biomart-web/<br />
    conf/
  </td>
  <td>
    ensembl/<br />
    ensembl-compara/<br />
    ensembl-draw/<br />
    ensembl-external/
  </td>
  <td>
    ensembl-variation/<br />
    htdocs/<br />
    modules/<br />
    perl/
  </td>
  <td>
    public-plugins/<br />
    utils/
  </td>
</tr>
</table>
</li>
</ol>
<p>You should now have all the Ensembl website code and data installed and ready to configure.</p>

</body>
</html>
  
