<html>
<head>
<title>Variant Effect Predictor</title>
</head>
<body>

<div class="column-wrapper">
<h1>Variant Effect Predictor</h1>
</div>
<div class="column-wrapper">
  <div class="column-two">
    <div class="column-left">
      <div class="plain-box">
      <a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadVariations" class="modal_link">
      <img src="/i/vep_logo_sm.png" title="Launch the online VEP!" alt="Variant Effect Predictor" /></a>

<p>The VEP determines the effect of your polymorphisms on genes, transcripts, 
and protein sequence.  Simply input the genomic coordinates of SNPs, insertions 
or deletions, along with any nucleotides at those positions, and export:</p>
<ul>
<li>genes affected by the changes</li>
<li>where the variants are located with respect to cDNAs (e.g. upstream, 
downstream, in the coding sequence)</li>
<li>if the subsitutions, insertions or deletions have changed the protein sequence</li>
<li>SIFT and PolyPhen scores for the altered sequences (if they have affected the 
<li>protein sequence</li>
</ul>
      <p class="center"><a
href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadVariations"
class="modal_link"><strong><i>Launch the online VEP tool!</i></strong></a></p>
<h2>Sample output</h2>
<p><img src="/i/help/vep_output.jpg" title="Sample VEP results table" alt="screenshot" /></p>

      </div>
    </div>
  </div>
  <div class="column-two">
    <div class="column-right" style="padding-left:2em">
<p>
Ensembl provides the facility to predict the functional consequences of known and unknown variants
using the Variant Effect Predictor (VEP). There are three primary ways to use
the functionality of the VEP:
</p>

<ul>
  <li> through the web interface | <a href="#web">Documentation</a> | <a
href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadVariations"
class="modal_link">Launch web interface</a></li>
  <li> using the standalone perl script | <a href="vep_script.html">Documentation</a> | <a href="http://cvs.sanger.ac.uk/cgi-bin/viewvc.cgi/ensembl-tools/scripts/variant_effect_predictor.tar.gz?view=tar&root=ensembl&pathrev=branch-ensembl-[[SPECIESDEFS::ENSEMBL_VERSION]]">Download latest version</a></li>
  <li> using Ensembl's perl API | <a href="/info/docs/api/variation/index.html">Documentation</a></li>
</ul>

<p>
The web version is suitable for users with small volumes of data or those who
prefer not to use command-line utilities. The script version is the most
flexible of the VEP, and allows users to process large volumes of data using
their own compute resources. The API is suitable for perl programmers looking to
incorporate features of the VEP into their own code.
</p>
    </div>
  </div>
</div>
