<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>

  <meta name="author" content="web">

  <meta name="navigation" content="Comparative Genomics">
  <title>Comparative Genomics</title>


</head>

<body>

<h2>Comparative Genomics</h2>

<h3>Protein Homology Analysis</h3>

<p>Three types of homology relationships are provided.</p>

<h4>Protein trees</h4>

<p>Protein trees are calculated using the longest peptide of all
the Ensembl protein coding genes. Proteins are clustered based on
Best-Reciprocal Hits and Blast Score Ratios. Each cluster of protein is
aligned using Muscle. NJTREE is used to get a gene tree from each
multiple alignment. The gene tree is reconciliated with the species
tree to call duplication events.</p>

<h4>Homologues</h4>

<p>Homologues (orthologues and paralogues) are deduced from the
protein trees. These are built using the longest transcript of each gene only.
A detailed description of the method is provided <a href="homology_method.html">here</a>.</p>

<h4>Families</h4>

<p>Families are groups of related genes. The full set of Ensembl
proteins is extended using all the metazoan proteins from SwissProt and
SPTrEMBL.</p>

<h3>BlastZ-net Analysis</h3>

<p>BlastZ-net (<a href="http://www.genome.org/cgi/content/abstract/13/1/103">Schwartz
S et al., Genome Res.;13(1):103-7</a>, <a href="http://www.pnas.org/cgi/content/full/100/20/11484">Kent
WJ et al., Proc Natl Acad Sci U S A., 2003;100(20):11484-9</a>)
alignments are provided for closely related pairs of species. The
alignments are the results of post-processing the raw BlastZ results.
In the first step, original blocks are chained according to their
location in both genomes. The netting process chooses for the reference
species the best sub-chain in each region. The reference species in the
BlastZ-net alignments is defined in this table:</p>

[[INCLUDE::/info/data/compara/blastz_net.inc]]
<h3>Translated Blat Analysis</h3>

<p>Translated blat (<a href="http://www.genome.org/cgi/content/abstract/12/4/656">Kent
W, Genome Res., 2002;12(4):656-64</a>) is used to look for
homologous regions between more distantly related pairs of species. We
expect to find homologies mainly in coding regions.</p>

[[INCLUDE::/info/data/compara/tblat.inc]]
<h3>PECAN Analysis</h3>

<p>Pecan is used to provide global multiple genomic alignments. First,
Mercator is used to build a synteny map between the genomes and then
Pecan builds alignments in these syntenic regions.</p>

<p>Pecan is a global multiple sequence alignment program that makes
practical the probabilistic consistency methodology for significant
numbers of sequences of practically arbitrary length. As input it
takes a set of sequences and a phylogenetic tree. The parameters and
heuristics it employs are highly user configurable, it is written
entirely in Java and also requires the installation of Exonerate.
<a href="http://www.ebi.ac.uk/~bjp/pecan/">Read more about Pecan</a>.</p>

<p>Additionally we use Gerp
(<a href="http://www.genome.org/cgi/content/abstract/15/7/901">Cooper
GM et al., Genome Res., 2005; 15:901-913</a>) to calculate conservation
scores and call constrained elements on the 10 way multiple alignments.
Conservation scores are estimated on a column-by-column basis.
Constrained elements are stretches of the multiple alignment where the
sequences are highly conserved according to the previous score.</p>

[[INCLUDE::/info/data/compara/pecan.inc]]
<h3>Synteny Analysis</h3>

<p>We calculate syntenic regions using blastz-net alignments. We look for
stretches where the alignment blocks are in synteny. The search is run in
two phases. In the first one, syntenic alignments that are closer than 200
kbp are grouped. In the second phase, the groups that are in synteny are
linked provided that no more than 2 non-syntenic groups are found between
them and they are less than 3Mbp apart.</p>

[[INCLUDE::/info/data/compara/synteny.inc]]
</body>
</html>
