<html>
<head>
<meta name="author" content="helpdesk" />
<meta name="navigation" content="Documentation" />
<title>BAC Clone Ordering</title>
</head>

<body>


<h2>BAC Clone Ordering</h2>


<h3 class="boxed">Background</h3>

<p>
Only genome sequences determined by systematic bacterial artificial chromosome (BAC) clone sequencing will have clones available.
Whole Genome Shotgun (WGS) clones are generally not available for ordering, but if the genome sequences are the result of a composite assembly,
BAC clones may be available in certain regions.

The human genome sequence however has been solely determined by systematic BAC clone sequencing.
</p>

<p>
BAC clone numbers can be obtained looking up accession numbers in the
<a href="http://www.ebi.ac.uk/embl/" target="external">EMBL</a> database.
The alternating shades of blue '<b>DNA (contigs)</b>' track on Ensembl
'<a href="javascript:void(window.open('/perl/helpview?se=1;kw=contigview','helpview','width=700,height=550,resizable,scrollbars'))">ContigView</a>' and
'<a href="javascript:void(window.open('/perl/helpview?se=1;kw=cytoview','helpview','width=700,height=550,resizable,scrollbars'))">CytoView</a>' pages display
<a href="http://www.ebi.ac.uk/embl/" target="external">EMBL</a>,
<a href="http://www.ncbi.nlm.nih.gov/Genbank/" target="external">GenBank</a> or 
<a href="http://www.ddbj.nig.ac.jp/" target="external">DDBJ</a> 
accession numbers if the genome sequence has been submitted to one of these databases.  These accession numbers generally have two formats:
the original one with one letter and five numbers (U12345) and the current extension with two letters and six numbers (AB123456).
Ensembl normally uses the sequence version in the '<b>DNA (contigs)</b>' tracks.
The sequence version is based on the accession number and appends a dot and an number, which gets incremented each time the sequence in the EMBL source record is updated (AB123456.1).
</p>

<p>
If a particular contig sequence has been submitted to the public nucleotide databases, the corresponding pop-up window will display a menu entry '<b>Export EMBL source file</b>'.
Following this link leads to an EMBL flat file, which should contain the relevant international BAC clone number in its '<b>Description (DE)</b>' line.
</p>

<p>
More information on the international BAC clone 
<a href="http://www.ncbi.nlm.nih.gov/genome/clone/nomen.html" target="external">nomenclature</a>, as well as individual BAC 
<a href="http://www.ncbi.nlm.nih.gov/genome/clone/clbrowse.cgi" target="external">libraries</a> and clones is available from the NCBI 
<a href="http://www.ncbi.nlm.nih.gov/genome/clone/" target="external">Clone Registry</a>. 

Once it is known from which library a particular clone originates, Clone Registry provides additional information about the BAC library source, as well as its supplier. 
</p>

<p>
Other species like mouse and zebrafish have been sequenced in a composite approach. 
Initial whole genome shotgun sequencing was followed by systematic BAC clone sequencing. 
The assembly of the genome sequence is thus based on whole genome shotgun reads with additional BAC sequences stitched onto it. 
BAC clones in general have an
<a href="http://www.ebi.ac.uk/embl/" target="external">EMBL</a>, 
<a href="http://www.ncbi.nlm.nih.gov/Genbank/" target="external">GenBank</a> or 
<a href="http://www.ddbj.nig.ac.jp/" target="external">DDBJ</a> accession number assigned, as they are deposited in the High Throughput Genomic (HTG) sections of these nucleotide databases. 
</p>

<p>
The remaining species have been sequenced in a whole genome shotgun (WGS) approach only. 
WGS clones are generally not available, although it might be worth checking, whether a species-specific BAC library is available. 
Even if this library might not have served as the basis for the genome-sequencing project, substantial numbers of BAC clone sequences could well be available in
<a href="http://www.ebi.ac.uk/embl/" target="external">EMBL</a>,
<a href="http://www.ncbi.nlm.nih.gov/Genbank/" target="external">GenBank</a> or
<a href="http://www.ddbj.nig.ac.jp/" target="external">DDBJ</a>.
Sequence similarity searches could help to locate such clones.
</p>

<p>
If available, Ensembl will display a <b>BAC end pairs</b> track on
'<a href="javascript:void(window.open('/perl/helpview?se=1;kw=contigview','helpview','width=700,height=550,resizable,scrollbars'))">ContigView</a>' and
'<a href="javascript:void(window.open('/perl/helpview?se=1;kw=cytoview','helpview','width=700,height=550,resizable,scrollbars'))">CytoView</a>' pages.
This track can be switched on from the '<b>Decorations</b>' or the
'<b>DAS Sources</b> menu in the yellow menu bar in '<b>Detailed View</b>' if it is not on by default.

BAC end sequences are determined in an early sequencing stage as single reads from the library vector into the insert.
Individual BAC end sequences are matched onto the genome sequence and Ensembl will join corresponding BAC end pairs if both ends match within the region displayed.
Zooming out might thus reveal BAC clones suitable for the study of particular genes or regions.
</p>

<!-- List of suppliers is missing -->

</body>
</html>
