<!-- $Id$ -->

<html>
<head>
<meta name="author" content="DAS:AndreasKahari" />
<meta name="navigation" content="DAS documentation" />
<title>DAS in Ensembl</title>
</head>
<body>

<h2>DAS in Ensembl</h2>


<h3 class="boxed">The Ensembl DAS reference server</h3>

<p>Ensembl provides a DAS reference server which
gives access to genomic sequences, 
the latest Ensembl gene predictions and repeats.  The
Ensembl DAS reference server with a table of available sources may be found at:</p>

<blockquote>
<a href="http://servlet.sanger.ac.uk:8080/das/">http://servlet.sanger.ac.uk:8080/das/</a>
</blockquote>

<p>Here is a properly formatted XML document with the same information:</p>

<blockquote>
<a href="http://servlet.sanger.ac.uk:8080/das/dsn">http://servlet.sanger.ac.uk:8080/das/dsn</a>
</blockquote>


<h4>Example requests</h4>

<p>DAS request URLs have a specific format:</p>

<blockquote>
<var>protocol</var>://<var>site-prefix</var>/das/<var>data-source</var>/<var>command</var>?<var>arguments</var>
</blockquote>

<p>For example:</p>

<dl>
<dt>
<a href="http://servlet.sanger.ac.uk:8080/das/ensembl_Homo_sapiens_core_28_35a/features?segment=13:31787617,31871805">http://servlet.sanger.ac.uk:8080/das/ensembl_Homo_sapiens_core_28_35a/features?segment=13:31787617,31871805</a>
</dt>
<dd>
Request all features in the region containing BRCA2 [31787617,31871805] on 
human chromosome 13.
</dd>
<dt>
<a href="http://servlet.sanger.ac.uk:8080/das/ensembl_Gallus_gallus_core_28_1d/sequence?segment=1:1,1000">http://servlet.sanger.ac.uk:8080/das/ensembl_Gallus_gallus_core_28_1d/sequence?segment=1:1,1000</a>
</dt>
<dd>
Request the first 1000&nbsp;bp of the first chicken chromosome.
</dd>
</dl>

<h3 class="boxed">The Ensembl DAS Stylesheet support</h3>
<p>Ensembl supports use of `stylesheet` DAS request which provides a facility to make the display of the features nicer. The full document in PDF format on what is supported may be found at:</p>
<blockquote>
<a href="CSS_support.pdf">Ensembl Stylesheet Support</a>
</blockquote>


<h3 class="boxed">DAS clients in Ensembl and semantic extensions to the DAS
specification</h3>

<p>DAS clients are built into several Ensembl  displays to enable 
incorporation of third-party annotation data by "attaching a
DAS source".  Ensembl uses DAS for showing
external third-party data and data from in-house DAS sources.
The 
<a href="/Homo_sapiens/helpview?se=1;kw=dasconfview" target="_top">help pages for DASConfView</a> explain how to attach a DAS data source to an Ensembl display.</p>


<p>Some Ensembl displays introduce semantic extensions to
the DAS specification.  This is explained below.</p>


<h4>ContigView and CytoView</h4>


<p>A genomic DAS source serves data in the format specified in the <a
href="http://biodas.org/documents/spec.html">DAS/1.5</a>
specification.  

In Ensembl, genomic DAS sources can be attached to 
<a href="http://www.ensembl.org/Homo_sapiens/contigview?region=AC112492.3.1.116568;start=2015;end=116568">ContigView</a>
and
<a href="http://www.ensembl.org/Homo_sapiens/cytoview?region=AL592145.4.1.5885;start=2001;end=5885">CytoView</a> (see examples in human)
if the annotation is in chromosomal coordinates, 
on contigs or on any basic assembly structure known to Ensembl 
for that species.  For genomic DAS sources the assembly version is 
important as otherwise features from a source annotation (e.g. 
NCBI 34 human assembly) may be offset in ContigView.</p>

<p>Please note that for most species, Ensembl chromosomes are named like
"1", "2", "X", "Y" rather than like "chr1", "chr2", "chrX", "chrY".
The exceptions are the Drosophila melanogaster, Anopheles gambiae,
Caenorhabditis elegans, and Saccharomyces cerevisiae genomes which uses
other naming conventions.</p>


<h4>ProtView</h4>

<p><em>Proteomic DAS sources</em> (or <em>ProteinDAS sources</em>) 
can be attached to ProtView which displays protein information 
(see <a href="/Homo_sapiens/protview?peptide=ENSP00000338013">example in human</a>).</p>


<p>A proteomic DAS client and server exchange protein annotations
 using the DAS protocol.  Genomic DAS sources annotate DNA sequence 
from a genomic assembly in contrast to proteomic DAS sources
which annotate proteins.  The extension of the
DAS protocol to proteins is only semantic and the requests
and replies between client and server uses the exact same
specification as in the genomic case.  Sections of the XML reply 
that are not relevant to proteomic data (e.g. "phase" and "strand"), 
are simply ignored by the proteomic DAS client.</p>

<p>In most cases, the UniProt peptide sequences are used as the
common reference.</p>

<p>Ensembl currently provides two proof-of-concept ProteinDAS
annotation servers:</p>

<ul>
<li>
<a href="http://das.ensembl.org/das/swissprot">http://das.ensembl.org/das/swissprot</a>
<br />
serving SwissProt annotations, and
</li>
<li>
<a href="http://das.ensembl.org/das/interpro">http://das.ensembl.org/das/interpro</a>
<br />
serving InterPro annotations.
</li>
</ul>

<p>There is also an official UniProt reference server.  This
server is updated every second to serve data from
the UniProt group at the EBI and also serves peptide
sequences:</p>

<ul>
<li>
<a href="http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle">http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle</a>
</li>
</ul>

<p>Example requests would be:</p>

<ul>
<li>
<a href="http://das.ensembl.org/das/swissprot/features?segment=MIP_MOUSE">http://das.ensembl.org/das/swissprot/features?segment=MIP_MOUSE</a>
</li>
<li>
<a href="http://das.ensembl.org/das/interpro/features?segment=CLAT_HUMAN">http://das.ensembl.org/das/interpro/features?segment=CLAT_HUMAN</a>
</li>
<li>
<a href="http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/features?segment=P01591">http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/features?segment=P01591</a>
</li>
<li>
<a href="http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/sequence?segment=P01591">http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/sequence?segment=P01591</a>
</li>
</ul>

<p>The UniProt DAS reference server is able to respond
to any of the types of identifiers listed on the <a
href="http://www.ebi.ac.uk/uniprot-das/">UniProt DAS information
page</a>.

<p>ProteinDAS annotations can be browsed from Ensembl ProtView
pages, e.g.</p>

<ul>
<li>
<a href="http://www.ensembl.org/Mus_musculus/protview?peptide=MIP_MOUSE;db=core;ENSEMBL_GENE_DAS_SOURCES!!SwissProt!!active=1">http://www.ensembl.org/Mus_musculus/protview?peptide=MIP_MOUSE</a>
</li>
<li>
<a href="http://www.ensembl.org/Homo_sapiens/protview?peptide=CLAT_HUMAN;db=core;ENSEMBL_GENE_DAS_SOURCES!!SwissProt!!active=0;ENSEMBL_GENE_DAS_SOURCES!!InterPro!!active=1">http://www.ensembl.org/Homo_sapiens/protview?peptide=CLAT_HUMAN</a>
</li>
<li>
<a href="http://www.ensembl.org/Homo_sapiens/protview?peptide=P01591;db=core;ENSEMBL_GENE_DAS_SOURCES!!UniProt!!active=1">http://www.ensembl.org/Homo_sapiens/protview?peptide=P01591</a>
</li>
</ul>

<p>The ProtView display of Ensembl is able to display
non-positional features in the same way as GeneView (see
below).</p>

<h4>GeneView</h4>

<p>DAS sources serving non-positional features
(or GeneDAS sources) can be attached to GeneView
<a href="/Homo_sapiens/geneview?gene=ENSG00000139618">(see example for human)</a> which displays gene information.

Non-positional feature DAS is a semantic extension to
the DAS protocol.  It allows exchange of annotations
tied to identifiers, such as a HUGO gene name or an Ensembl
gene ID, rather than to a segment of reference sequence.
Positional data is irrelevant and the annotation applies to the
entire entity referenced by the request.  GeneDAS annotations
are defined as having 'start' and 'end' attributes set to 0
(zero) and the annotation itself is carried in the 'note'
attribute.</p>

<p>Among others, Ensembl currently uses the UniProt ProteinDAS
reference server (see above) in a GeneDAS capacity.  An example
DAS request with both GeneDAS and ProteinDAS annotations is:</p>

<ul>
<li>
<a href="http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/features?segment=P51587">http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/features?segment=P51587</a>
</li>
</ul>

<p>The corresponding Ensembl GeneView page is:</p>

<ul>
<li>
<a href="http://www.ensembl.org/Homo_sapiens/geneview?DASselect_UniProt=1;%3ADASselect_UniProt=0;gene=P51587;db=core">http://www.ensembl.org/Homo_sapiens/geneview?gene=P51587</a>
</li>
</ul>


<h3 class="boxed">URL Auto-Configuration</h3>

<p>
DAS sources can be automatically attached and configured while following links to Ensembl data displays.
</p>


<h4>Mandatory Parameters</h4>

<p>
URLs for automatic DAS source attachment must contain the following parameters:
</p>


<ul>

<li>
<p>
<strong>url</strong> -
The URL to the DAS server ending with '/das'.
</p>
</li>

<li>
<p>
<strong>dsn</strong> -
The DAS source name as defined in the DAS server.
</p>
</li>

<li>
<p>
<strong>type</strong> -
The mapping type, which defines the coordinate sytem to use for fetching features.
</p>

<p>
Possible values are:
</p>

<dl>

<dt>hugo</dt>
<dd>
The gene symbol as assigned by the Human Genome Nomencalture Committee
(<a href="http://www.gene.ucl.ac.uk/nomenclature/" target="external">HGNC</a>) <i>e.g. BRCA2</i>
</dd>

<dt>markersymbol</dt>
<dd>The gene symbol as assigned by Mouse Genome Informatics
(<a href="http://www.informatics.jax.org/" target="external">MGI</a>) <i>e.g. Brca2</i>
</dd>

<dt>mgi</dt>
<dd>MGI IDs corresponding to the Mouse Genome Informatics database <i>e.g. MGI:109337</i></dd>

<dt>ensembl_gene</dt>
<dd>Ensembl Gene ID <i>e.g. ENSG00000139618</i></dd>

<dt>ensembl_location</dt>
<dd>Ensembl Location <i>e.g. 13:31787617-31871805</i></dd>

<dt>ensembl_peptide</dt>
<dd>Ensembl Peptide ID <i>e.g. ENSP00000267071</i></dd>

<dt>ensembl_transcript</dt>
<dd>Ensembl Transcript ID <i>e.g. ENST00000267071</i></dd>

<dt>uniprot/swissprot</dt>
<dd><a href="http://www.ebi.ac.uk/swissprot/" target="external">Uniprot/Swiss-Prot</a> Name <i>e.g. BRCA2_HUMAN</i></dd>

<dt>uniprot/swissprot_acc</dt>
<dd><a href="http://www.ebi.ac.uk/swissprot/" target="external">Uniprot/Swiss-Prot</a> Accession number <i>P51587</i></dd>

<dt>entrezgene</dt>
<dd><a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=gene" target="external">Entrez Gene</a> ID <i>e.g. 675</i></dd>

<dt>ipi_acc</dt>
<dd>
International Protein Index
(<a href="http://www.ebi.ac.uk/ipi/" target="external">IPI</a>) accession number <i>e.g. IPI00015171</i>
</dd>

<dt>ipi_id</dt>
<dd>
International Protein Index
(<a href="http://www.ebi.ac.uk/ipi/" target="external">IPI</a>) name <i>e.g. IPI00015171.4</i>
</dd>

</dl>
</li>

<li>
<p>
<strong>name</strong> -
The DAS source name as it will appear on Ensembl data displays.
</p>
</li>

</ul>


<h4>Optional Parameters</h4>

<ul>

<li>
<p>
<strong>active</strong> -
Set active=1 to not only attach the DAS source but also automatically switch it on for display.
</p>
</li>

<li>
<p>
<strong>stylesheet</strong> -
Set stylesheet=y to use a source stylesheet to display features.
</p>
</li>

<li>
<p>
<strong>score</strong> -
Set score=h to display data as a histogram.
</p>
</li>

<li>
<p>
<strong>label</strong> -
Set label=TEXT to define a name for the DAS track in ContigView and CytoView.
</p>
</li>

<li>
<p>
<strong>caption</strong> -
Set caption=TEXT to define a name of the DAS source as it will appear in the DAS menu.
</p>
</li>

<li>
<p>
<strong>group</strong> -
Set group=y to group features.
</p>
</li>

<li>
<p>
<strong>color</strong> -
Set color=COLOUR to use this colour for feature display.
</p>
</li>


<li>
<p>
<strong>strand</strong> -
Set strand=b|f|r to display features on both, forward or reverse strand, respectively.
</p>
</li>

<li>
<p>
<strong>depth</strong> -
Set depth=INT for the number of rows to be used by this DAS track in ContigView.
Set to 10000 for unlimited.
</p>
</li>

<li>
<p>
<strong>labelflag</strong> -
Set labelflag=n|o|u to display no feature labels at all or to display them over or under the feature, respectively.
</p>
</li>

</ul>


<h4>Example URL</h4>

<p>
The following example URL activates a DAS track in Ensembl ContigView, which annotates the results of Ensembl Pepetide ID to PDB protein structure mapping.
</p>

<pre>
<a href="http://www.ensembl.org/Homo_sapiens/contigview?peptide=ENSP00000354398&amp;add_das_source=(url=http://das.sanger.ac.uk/das+dsn=ensppdbmapping+type=ensembl_peptide+name=ENSP_PDB_mapping+active=1)" target="external">http://www.ensembl.org/Homo_sapiens/contigview?peptide=ENSP00000354398&amp;add_das_source=(url=http://das.sanger.ac.uk/das+dsn=ensppdbmapping+type=ensembl_peptide+name=ENSP_PDB_mapping+active=1)</a>
</pre>

</body>
</html>
