<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta name="generator" content=
    "HTML Tidy for OpenBSD (vers 1 September 2005), see www.w3.org" />
    <meta name="author" content="DAS:AndreasKahari" />
    <meta name="navigation" content="DAS documentation" />
    <title>
      DAS in Ensembl
    </title>
  </head>
  <body>
    <!-- $Id$ -->
    <h2>
      DAS in Ensembl
    </h2>
    <h3 class="boxed">
      The Ensembl DAS reference server
    </h3>
    <p>
      Ensembl provides a DAS reference server which gives access to
      genomic sequences, the latest Ensembl gene predictions, and for
      some species, karyotypes and ditags. A list of the sources
      currently served from the Ensembl DAS reference server may be
      found as XML documents at
    </p>
    <blockquote>
      <a href=
      "http://www.ensembl.org/das/dsn">http://www.ensembl.org/das/dsn</a>
      (listing generated from the standard DAS data source name (DSN)
      request)
      <br />
      <a href=
      "http://www.ensembl.org/das/sources">http://www.ensembl.org/das/sources</a>
      (a listing with extended information)
    </blockquote>
    <h4>
      Example requests
    </h4>
    <p>
      DAS request URLs have a specific format:
    </p>
    <blockquote>
      <var>protocol</var>://<var>site-prefix</var>/das/<var>data-source</var>/<var>command</var>?<var>arguments</var>
    </blockquote>
    <p>
      For example:
    </p>
    <dl>
      <dt>
        <a href=
        "http://www.ensembl.org/das/Homo_sapiens.NCBI36.transcript/features?segment=13:31787617,31871806">
        http://www.ensembl.org/das/Homo_sapiens.NCBI36.transcript/features?segment=13:31787617,31871806</a>
      </dt>
      <dd>
        Request all transcripts (exons really) in the region
        [31787617,31871806] on human chromosome 13 (this is where the
        gene BRCA2 is located in the NCBI&nbsp;36 assembly).
      </dd>
      <dt>
        <a href=
        "http://www.ensembl.org/das/Gallus_gallus.WASHUC1.reference/sequence?segment=1:1,1000">
        http://www.ensembl.org/das/Gallus_gallus.WASHUC1.reference/sequence?segment=1:1,1000</a>
      </dt>
      <dd>
        Request the first 1000&nbsp;bp of the first chicken chromosome.
      </dd>
    </dl>
    <h3 class="boxed">
      DAS stylesheet support in Ensembl
    </h3>
    <p>
      Ensembl supports use of DAS "stylesheet" request which provides a
      facility to make the display of the features nicer. A PDF
      document that describes the stylesheet support is avaliable here:
    </p>
    <blockquote>
      <a href="CSS_support.pdf">Ensembl Stylesheet Support</a>
    </blockquote>
    <h3 class="boxed">
      DAS clients in Ensembl and semantic extensions to the DAS
      specification
    </h3>
    <p>
      DAS clients are built into several Ensembl displays to enable
      incorporation of third-party annotation data by "attaching a DAS
      source". Ensembl uses DAS for showing external third-party data
      and data from in-house DAS sources. The <a href=
      "/Homo_sapiens/helpview?se=1;kw=dasconfview" target="_top">help
      pages for DASConfView</a> explain how to attach a DAS data source
      to an Ensembl display.
    </p>
    <p>
      Some Ensembl displays introduce semantic extensions to the DAS
      specification. This is explained below.
    </p>
    <h4>
      ContigView and CytoView
    </h4>
    <p>
      A genomic DAS source serves data in the format specified in the
      <a href="http://biodas.org/documents/spec.html">DAS/1.5</a>
      specification. In Ensembl, genomic DAS sources can be attached to
      <a href=
      "http://www.ensembl.org/Homo_sapiens/contigview?region=AC112492.3.1.116568;start=2015;end=116568">
      ContigView</a> and to <a href=
      "http://www.ensembl.org/Homo_sapiens/cytoview?region=AL592145.4.1.5885;start=2001;end=5885">
      CytoView</a> (see examples in human) if the annotation is in
      chromosomal coordinates, on contigs or on any other top-level
      assembly structure known to Ensembl for that species. For genomic
      DAS sources the assembly version is important as otherwise
      features from a source annotation (e.g. from the NCBI&nbsp;36
      human assembly) may be offset in ContigView.
    </p>
    <p>
      Please note that for most species, Ensembl chromosomes are named
      like "1", "2", "X", "Y" rather than like "chr1", "chr2", "chrX",
      "chrY". The exceptions are the Drosophila melanogaster, Anopheles
      gambiae, Caenorhabditis elegans, and Saccharomyces cerevisiae
      genomes which uses other naming conventions.
    </p>
    <h4>
      ProtView
    </h4>
    <p>
      <em>Proteomic DAS sources</em> (or <em>ProteinDAS sources</em>)
      may be attached to ProtView which displays protein information
      (see <a href=
      "/Homo_sapiens/protview?peptide=ENSP00000338013">example in
      human</a>).
    </p>
    <p>
      A proteomic DAS client and server exchange protein annotations
      using the DAS protocol. Genomic DAS sources annotate DNA sequence
      from a genomic assembly in contrast to proteomic DAS sources
      which annotate proteins. The extension of the DAS protocol to
      proteins is only semantic and the requests and replies between
      client and server uses the exact same specification as in the
      genomic case. Sections of the XML reply that are not relevant to
      proteomic data (e.g. "phase" and "strand"), are simply ignored by
      the proteomic DAS client.
    </p>
    <p>
      In most cases, the UniProt peptide sequences are used as the
      common reference.
    </p>
    <p>
      There is an official UniProt reference server which is not
      maintained by the Ensembl group. This server is serving the
      latest UniProt data from the UniProt group at the EBI and it also
      serves peptide sequences:
    </p>
    <ul>
      <li>
        <a href=
        "http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle">http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle</a>
      </li>
    </ul>
    <p>
      Example requests would be:
    </p>
    <ul>
      <li>
        <a href=
        "http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/features?segment=P01591">
        http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/features?segment=P01591</a>
      </li>
      <li>
        <a href=
        "http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/sequence?segment=P01591">
        http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/sequence?segment=P01591</a>
      </li>
    </ul>
    <p>
      The UniProt DAS reference server is able to respond to any of the
      types of identifiers listed on the <a href=
      "http://www.ebi.ac.uk/uniprot-das/">UniProt DAS information
      page</a>.
    </p>
    <p>
      Proteomic DAS annotations can be browsed from Ensembl ProtView
      pages, e.g.
    </p>
    <ul>
      <li>
        <a href=
        "http://www.ensembl.org/Mus_musculus/protview?peptide=MIP_MOUSE;db=core;das_sources=Uniprot">
        http://www.ensembl.org/Mus_musculus/protview?peptide=MIP_MOUSE</a>
      </li>
      <li>
        <a href=
        "http://www.ensembl.org/Homo_sapiens/protview?peptide=CLAT_HUMAN;db=core;das_sources=UniProt">
        http://www.ensembl.org/Homo_sapiens/protview?peptide=CLAT_HUMAN</a>
      </li>
      <li>
        <a href=
        "http://www.ensembl.org/Homo_sapiens/protview?peptide=P01591;db=core;das_sources=UniProt">
        http://www.ensembl.org/Homo_sapiens/protview?peptide=P01591</a>
      </li>
    </ul>
    <p>
      The ProtView display of Ensembl is able to display non-positional
      features in the same way as GeneView (see below).
    </p>
    <h4>
      GeneView
    </h4>
    <p>
      DAS sources serving <em>non-positional</em> features (sometimes
      also called <em>GeneDAS sources</em>) can be attached to GeneView
      <a href="/Homo_sapiens/geneview?gene=ENSG00000139618">(see
      example for human)</a> which displays gene information.
      Non-positional feature DAS is a semantic extension to the DAS
      protocol. It allows exchange of annotations tied to identifiers,
      such as a HUGO gene name or an Ensembl gene ID, rather than to a
      segment of reference sequence. Positional data is irrelevant and
      the annotation applies to the entire entity referenced by the
      request. Non-positional annotations are defined as having 'start'
      and 'end' attributes set to 0 (zero) and the annotation itself is
      carried in the 'note' attribute.
    </p>
    <p>
      Among others, Ensembl currently uses the UniProt proteomic DAS
      reference server (see above) in a GeneDAS capacity. An example
      DAS request with both non-positional and proteomic DAS
      annotations is:
    </p>
    <ul>
      <li>
        <a href=
        "http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/features?segment=P51587">
        http://www.ebi.ac.uk/das-srv/uniprot/das/aristotle/features?segment=P51587</a>
      </li>
    </ul>
    <p>
      The corresponding Ensembl GeneView page is:
    </p>
    <ul>
      <li>
        <a href=
        "http://www.ensembl.org/Homo_sapiens/geneview?gene=P51587;db=core;das_sources=UniProt">
        http://www.ensembl.org/Homo_sapiens/geneview?gene=P51587</a>
      </li>
    </ul>
    <h3 class="boxed">
      URL Auto-Configuration
    </h3>
    <p>
      DAS sources can be automatically attached and configured while
      following links to Ensembl data displays.
    </p>
    <h4>
      Mandatory Parameters
    </h4>
    <p>
      URLs for automatic DAS source attachment must contain the
      following parameters:
    </p>
    <dl>
      <dt>
        url
      </dt>
      <dd>
        The URL to the DAS server ending with '/das'.
      </dd>
      <dt>
        dsn
      </dt>
      <dd>
        The DAS source name as defined in the DAS server.
      </dd>
      <dt>
        name
      </dt>
      <dd>
        The DAS source name as it will appear on Ensembl data displays.
      </dd>
      <dt>
        type
      </dt>
      <dd>
        The mapping type, which defines the coordinate sytem to use for
        fetching features.
        <p>
          Possible values for the type are:
        </p>
        <dl>
          <dt>
            hugo
          </dt>
          <dd>
            The gene symbol as assigned by the Human Genome
            Nomencalture Committee (<a href=
            "http://www.gene.ucl.ac.uk/nomenclature/" target=
            "external">HGNC</a>) <i>e.g. BRCA2</i>.
          </dd>
          <dt>
            markersymbol
          </dt>
          <dd>
            The gene symbol as assigned by Mouse Genome Informatics
            (<a href="http://www.informatics.jax.org/" target=
            "external">MGI</a>) <i>e.g. Brca2</i>.
          </dd>
          <dt>
            mgi
          </dt>
          <dd>
            MGI IDs corresponding to the Mouse Genome Informatics
            database <i>e.g. MGI109337</i> (note that the colon which
            is usually part of the MGI identifier is removed).
          </dd>
          <dt>
            ensembl_gene
          </dt>
          <dd>
            Ensembl Gene ID <i>e.g. ENSG00000139618</i>.
          </dd>
          <dt>
            ensembl_location_chromosome
          </dt>
          <dd>
            Ensembl Chromosome Location <i>e.g.
            13:31787617,31871805</i> (analogous to this there are also
            ensembl_location_contig, ensembl_location_scaffold, and
            ensembl_location_clone).
          </dd>
          <dt>
            ensembl_peptide
          </dt>
          <dd>
            Ensembl Peptide ID <i>e.g. ENSP00000267071</i>.
          </dd>
          <dt>
            ensembl_transcript
          </dt>
          <dd>
            Ensembl Transcript ID <i>e.g. ENST00000267071</i>.
          </dd>
          <dt>
            uniprot/swissprot
          </dt>
          <dd>
            <a href="http://www.ebi.ac.uk/swissprot/" target=
            "external">Uniprot/Swiss-Prot</a> Name <i>e.g.
            BRCA2_HUMAN</i>.
          </dd>
          <dt>
            uniprot/swissprot_acc
          </dt>
          <dd>
            <a href="http://www.ebi.ac.uk/swissprot/" target=
            "external">Uniprot/Swiss-Prot</a> Accession number
            <i>P51587</i>.
          </dd>
          <dt>
            entrezgene
          </dt>
          <dd>
            <a href=
            "http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?db=gene"
            target="external">Entrez Gene</a> ID <i>e.g. 675</i>.
          </dd>
          <dt>
            ipi_acc
          </dt>
          <dd>
            International Protein Index (<a href=
            "http://www.ebi.ac.uk/ipi/" target="external">IPI</a>)
            accession number <i>e.g. IPI00015171</i>.
          </dd>
          <dt>
            ipi_id
          </dt>
          <dd>
            International Protein Index (<a href=
            "http://www.ebi.ac.uk/ipi/" target="external">IPI</a>) name
            <i>e.g. IPI00015171.4</i>.
          </dd>
        </dl>
      </dd>
    </dl>
    <h4>
      Optional Parameters
    </h4>
    <dl>
      <dt>
        active
      </dt>
      <dd>
        Set active=1 to not only attach the DAS source but also
        automatically switch it on for display.
      </dd>
      <dt>
        stylesheet
      </dt>
      <dd>
        Set stylesheet=y to use a source stylesheet to display
        features.
      </dd>
      <dt>
        score
      </dt>
      <dd>
        Set score=h to display data as a histogram.
      </dd>
      <dt>
        label
      </dt>
      <dd>
        Set label=TEXT to define a name for the DAS track in ContigView
        and CytoView.
      </dd>
      <dt>
        caption
      </dt>
      <dd>
        Set caption=TEXT to define a name of the DAS source as it will
        appear in the DAS menu.
      </dd>
      <dt>
        group
      </dt>
      <dd>
        Set group=y to group features.
      </dd>
      <dt>
        color
      </dt>
      <dd>
        Set color=COLOUR to use this colour for feature display.
      </dd>
      <dt>
        strand
      </dt>
      <dd>
        Set strand=b|f|r to display features on both, forward or
        reverse strand, respectively.
      </dd>
      <dt>
        depth
      </dt>
      <dd>
        Set depth=INT for the number of rows to be used by this DAS
        track in ContigView. Set to 10000 for unlimited.
      </dd>
      <dt>
        labelflag
      </dt>
      <dd>
        Set labelflag=n|o|u to display no feature labels at all or to
        display them over or under the feature, respectively.
      </dd>
    </dl>
    <h4>
      Example URL
    </h4>
    <p>
      The following example URL activates a DAS track in Ensembl
      ContigView, which annotates the results of Ensembl Peptide ID to
      PDB protein structure mapping.
    </p>
    <pre>
<a href=
"http://www.ensembl.org/Homo_sapiens/contigview?peptide=ENSP00000354398&amp;add_das_source=(url=http://das.sanger.ac.uk/das+dsn=ensppdbmapping+type=ensembl_peptide+name=ENSP_PDB_mapping+active=1)"
target=
"external">http://www.ensembl.org/Homo_sapiens/contigview?peptide=ENSP00000354398&amp;add_das_source=(url=http://das.sanger.ac.uk/das+dsn=ensppdbmapping+type=ensembl_peptide+name=ENSP_PDB_mapping+active=1)</a>
</pre>
  </body>
</html>
