<html>
<head>
<title>Variation databases and API</title>
</head>

<body>

<h1>Ensembl Variation</h1>

<h2>Database Description</h2>

<p>
The Ensembl Variation database stores two types of variation data, depending it comes from an external source or data it is calculated on site.
</p>

<ul>
<li>Data imported from external sources (dbSNP, Sanger,...):
  <ul>
  <li>SNPS, in-dels</li>
  <li>Locations for variations</li>
  <li>Alleles</li>
  <li>Populations</li>
  <li>Genotypes</li>
  </ul>
</li>

<li>Calculated data:
  <ul>
  <li>Consequence of variations in transcripts (synonymous, non-synonymous,...)</li>
  <li>Linkage disequilibrium information</li>
  <li>Tagged variations</li>
  <li>Read coverage data</li>
  </ul>
</li>

</ul>

<h2>Database Schema</h2>

<p>
The table layout of the database, with all the tables, fields and relationships,  is found in the following pdf:
</p>

<ul>
<li>
<a href="./variation_schema.html">Variation Database Schema</a>
</li>
</ul>


<h2>Perl API</h2>

<p>
A comprehensive Perl Application Programme Interface (API) provides efficient access to the Ensembl Variation database.
</p>

<ul>

<li>
<a href="/info/docs/api/api_installation.html">Variation API Installation</a>: A step-by-step installation guide for all Perl Ensembl APIs.
</li>

<li>
<a href="/info/docs/Pdoc/ensembl-variation/index.html">Variation API Documentation</a>: A complete reference to the objects and methods used in the Variation database API.
</li>

<li>
<a href="./variation_tutorial.html">Variation Perl API Tutorial</a>: An introduction to the underlying concepts of the Variation database API.
</li>

</ul>

<h2 id="support">Ensembl Software Support</h2>

<p>
Ensembl is an open project and we would like to encourage correspondence and discussions on any subject on any aspect of Ensembl.
Please see the
<a href="http://www.ensembl.org/info/about/contact/">Ensembl Contacts</a> page for suitable options getting in touch with us.
</p>

</body>
</html>
