<html>
    <head>
        <title>
            Debugging Ensembl API Connections
        </title>
        <meta name="order" content="5"/>
        <meta name="navigation" content="Debugging Ensembl Installations">
    </head>
    <body>
        <h1>
            Debugging Ensembl API Connections
        </h1>
        <p>
            For most users installing Ensembl is a painless exercise. However computer setups are varied and sometimes issues will arise. This guide exists to help you debug and identify the root cause of a connection problem. To summarise, most issues are caused by
        </p>
        <ul>
            <li>Not installing the Ensembl required packages e.g. BioPerl, DBI and DBD::mysql
            </li>
            <li>Incorrect API usage
            </li>
            <li>Unable to access port 3306 due to local firewall restrictions
            </li>
            <li>Scheduled Ensembl database maintenance
            </li>
        </ul>
        <p>
            The following steps will not solve every issue but they will help diagnose these issues. This guide assumes familiarity with the *nix command line (such as bash).
        </p>
        <h2>
            1). Use ensembl/misc-scripts/ping_ensembl.pl
        </h2>
        <p>
            ping_ensembl.pl is our first port of call when debugging Ensembl connections and is distributed with the core API. It attempts to connect to one of the public MySQL databases (of your choosing) and tries to retrieve a human core DBAdaptor. The script will also try to diagnose common mistakes in setting up Ensembl. You should run the command line most appropriate to your situation:
        </p>
        <pre>
# To ping our UK server use
&gt; perl ensembl/misc-scripts/ping_ensembl.pl

# To ping our USEast server use
&gt; perl ensembl/misc-scripts/ping_ensembl.pl -ue

# To ping Ensembl Genomes use
&gt; perl ensembl/misc-scripts/ping_ensembl.pl -eg

# To ping our UK server but with a different species
&gt; perl ensembl/misc-scripts/ping_ensembl.pl -species pig
</pre>
        <p>
            Any issues with missing Perl library dependencies will be flagged here. Common missing modules include:
        </p>
        <ul>
            <li>DBI
            </li>
            <li>DBD::mysql
            </li>
        </ul>
        <p>
            Both of these libraries must be installed using your Perl distribution's installer. If you have built your own Perl then use CPAN/cpanminus. Linux users running system Perl can use a Linux distribution installer such as yum (Fedora, RHEL, Centos) and apt (Ubuntu). Other common issues include installing our unreleased API and not adding Ensembl libraries to your PERL5LIB. For both ping_ensembl.pl will advise on the best solution.
        </p>
        <h2>
            2). Are you Unable to Access Comparative, Variation or Regulation Data?
        </h2>
        <p>
            Ensembl is composed of multiple code projects based upon their data responsibility. This roughly equates to
        </p>
        <ul>
            <li>ensembl = genes, transcripts, translations, assembly, sequence
            </li>
            <li>ensembl-variation = SNVs, CNVs, somatic variations, phenotypes
            </li>
            <li>ensembl-compara = gene trees, protein families, multiple and pairwise genomic alignments
            </li>
            <li>ensembl-functgenomics = regulation, motifs, array probes
            </li>
        </ul>
        <p>
            You cannot access a set of data unless you have the correct libraries on your PERL5LIB. The previous step ping_ensembl will have warned you about the lack of Perl libraries on your PERL5LIB. If you require this data then install the library and put it onto your PERL5LIB.
        </p>
        <h2>
            3). Are you loading databases from the Registry?
        </h2>
        <p>
            The Registry has the option to print more information about what it is loading. Try using the following command and see if your species/data set has been loaded
        </p><pre>
Bio::EnsEMBL::Registry-&gt;load_registry_from_db( 
  -HOST =&gt; 'ensembldb.ensembl.org', 
  -USER =&gt; 'anonymous', 
  -VERBOSE =&gt; 1 
);
</pre>
        <p>
            Scan the output and try to find your required species.
        </p>
        <h2>
            4). Try Connecting to Ensembl Using a MySQL Client
        </h2>
        <p>
            If you can connect to our MySQL server using a client then the issue should be in your Perl or Ensembl setup.
        </p>
        <pre>
mysql --host=ensembldb.ensembl.org --port=3306 --user=anonymous

  Welcome to the MySQL monitor.  Commands end with ; or \g.
  Your MySQL connection id is 4292641
  Server version: 5.1.34-log Source distribution

  Copyright (c) 2000, 2013, Oracle and/or its affiliates. All rights reserved.

  Oracle is a registered trademark of Oracle Corporation and/or its
  affiliates. Other names may be trademarks of their respective
  owners.

  Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

  mysql&gt;
</pre>
        <p>
            You can also try using GUI alternatives such as <a href="http://www.sequelpro.com/" rel="external">Sequel Pro</a> (OSX), <a href="http://www.heidisql.com/" rel="external">HeidiSQL</a> (Windows) or <a href="http://squirrel-sql.sourceforge.net/" rel="external">SquirrelSQL</a> (multi-platform Java application).
        </p>
        <h2>
            5). Try Connecting to Another MySQL Server
        </h2>
        <p>
            If you cannot connect to our MySQL server then try another. Ensembl Genomes is available at mysql.ebi.ac.uk (port 4157 and user anonymous) and UCSC at genome-mysql.cse.ucsc.edu (port 3306 and user genome). Failure to connect to these servers indicates a systemic problem and not one unique to Ensembl.
        </p>
        <h2>
            6). Check for Scheduled Downtime
        </h2>
        <p>
            Sometimes downtime is unavoidable. When this occurs we will inform users which services will be unavailable via our <a rel="external" href="http://www.ensembl.info">blog</a> and <a href="/info/about/contact/mailing.html">mailing lists</a>.
        </p>
        <h2>
            7). Pinging the MySQL Server
        </h2>
        <p>
            The command ping (do not confuse with ping_ensembl.pl from step 1) issues packets to the DB servers. It assess if the address resolves to an IP address and that machine is available. USEast DB does not respond to pings.
        </p>
        <pre>
&gt; ping -c 5 ensembldb.ensembl.org

  PING ensembldb.sanger.ac.uk (XXX.XXX.XXX.XXX): 56 data bytes
  64 bytes from XXX.XXX.XXX.XXX: icmp_seq=0 ttl=59 time=1.600 ms
  64 bytes from XXX.XXX.XXX.XXX: icmp_seq=1 ttl=59 time=1.646 ms
  64 bytes from XXX.XXX.XXX.XXX: icmp_seq=2 ttl=59 time=1.724 ms
  64 bytes from XXX.XXX.XXX.XXX: icmp_seq=3 ttl=59 time=1.924 ms
  64 bytes from XXX.XXX.XXX.XXX: icmp_seq=4 ttl=59 time=1.374 ms

  --- ensembldb.sanger.ac.uk ping statistics ---
  5 packets transmitted, 5 packets received, 0.0% packet loss
  round-trip min/avg/max/stddev = 1.374/1.654/1.924/0.178 ms
</pre>
        <p>
            A healthy ping with no packet loss indicates a good connection to our SQL server.
        </p>
        <h2>
            8). Check with your Network Administrator if Port 3306 is Open
        </h2>
        <p>
            If you can ping Ensembl but still cannot connect check your local port settings. Some networks do not allow connection to MySQL servers by default. Check with your server network administrator that 3306 is open. If you are connecting to Ensembl Genomes check that port 4157 is open.
        </p>
        <h2>
            9). Search for your Error
        </h2>
        <p>
            It is sometimes possible that the error you are seeing is due to a component other than Ensembl. Try your favourite search engine (such as Google or Bing) and see what the internet can tell you. Sites like <a href="http://stackoverflow.com/" rel="external">Stack Overflow</a> and <a href="http://bugs.mysql.com/" rel="external">MySQL's bug tracking system</a> are invaluable resources. Be sure to use pertinent portions of error messages e.g. assuming the following error message
        </p>
        <pre>
Installation is good. Connection to Ensembl works and you can query the human core database
Error in my_thread_global_end(): 1 threads didn't exit
</pre>
        <p>
            Search for "my_thread_global_end" and "threads" rather than the entire script output. This particular issue is due to a MySQL library bug on Windows machines and not an issue directly related to Ensembl.
        </p>
        <h2>
            10). Try Re-installing your Setup or Talk to your Local System Administrator
        </h2>
        <p>
            Sometimes this works; sometimes it does not. However setups can become corrupted or out of date. A re-installation of later versions of software can help including re-installing <a href="/info/docs/api/api_installation.html">Ensembl</a>. If you are on a platform with systems support or a business contract then contact them for help. Some problems cannot be solved unless you talk to these people.
        </p>
        <p>
            Windows users may have more success using <a href="http://strawberryperl.com/" rel="external">Strawberry Perl</a> over <a href="http://www.activestate.com/activeperl" rel="external">ActiveState's Active Perl</a>.
        </p>
        <h2>
            11). Email Helpdesk
        </h2>
        <p>
            <a href="mailto:helpdesk@ensembl.org">Helpdesk</a> and the ensembl team are always available to help you to access our datasets and APIs. You should send the output of the above commands and steps with your problem as it will speed up the support process. Please be aware that our only supported platform for deployment is Linux (we also have some OSX experience) though we will try our best to help out on other platforms. Please give us as much information about your OS, architecture (e.g. 64bit) and all command line output from the following commands plus the previous steps (especially steps 1, 3, 4 and 5):
        </p>
        <pre>
&gt; perl -V
&gt; perl -MDBI -e 'warn $DBI::VERSION'
&gt; perl -MDBD::mysql -e 'warn $DBD::mysql::VERSION'
</pre>
    </body>
</html>
