<html>
<head>
<meta name="order" content="3" />
<title>API Documentation</title>
</head>

<body>

<h1>Perl API Documentation</h1>
<a href="http://www.perl.com/" title="www.perl.com"><img src="/img/info/powered_by_perl.gif" alt="Powered by Perl" class="float-right" style="width:122px;height:55px" /></a>

<p>Ensembl uses <a href="http://www.mysql.com/"
rel="external">MySQL</a> relational databases to store its
information. A comprehensive set of Application Programme Interfaces
(APIs) serve as a middle-layer between underlying database schemes and
more specific application programmes. The APIs aim to encapsulate the
database layout by providing efficient high-level access to data
tables and isolate applications from data layout changes.</p>

<p>Ensembl's API is written in Perl:</p>
<ul>
<li><a href="./api_installation.html">installation instructions</a></li>
<li><a href="http://www.ensembl.org/info/docs/Pdoc/">full documentation</a> of all modules</li>
</ul>

<table class="blobs" cellspacing="10">
<tr>

<td class="fourcol">
<h2 class="first">Core</h2>
<p>Sequence, genes and other <a href="/info/docs/genebuild/genome_annotation.html">automated annotation</a></p>
<div class="centered"><a href="/info/docs/api/core/"><img src="/i/blank.gif" style="width:200px;height:100px" /></a></div> 
<ul>
<li><a href="/info/docs/api/core/index.html#api">Perl API</a></li>
<li><a href="/info/docs/api/core/core_schema.html">Database schema</a></li>
</ul>
</td>

<td class="fourcol">
<h2 class="first">Comparative genomics</h2>
<p>Homologues, paralogues and protein families</p>
<div class="centered"><a href="/info/docs/api/compara/"><img src="/img/api_compara.gif" /></a></div> 
<ul>
<li><a href="/info/docs/api/compara/index.html#api">Perl API</a></li>
<li><a href="/info/docs/api/compara/compara_schema.html">Database schema</a></li>
</ul>
</td>

<td class="fourcol">
<h2 class="first">Variation</h2>
<p>SNPs, somatic mutations and structural variants</p>
<div class="centered"><a href="/info/docs/api/variation/"><img src="/img/api_variation.gif" /></a></div> 
<ul>
<li><a href="/info/docs/api/variation/index.html#api">Perl API</a></li>
<li><a href="/info/docs/api/variation/variation_schema.html">Database schema</a></li>
</ul>
</td>

<td class="fourcol">
<h2 class="first">Functional genomics</h2>
<p>Regulatory features, motifs and oligoprobes</p>
<div class="centered"><a href="/info/docs/api/funcgen/"><img src="/i/blank.gif" style="width:200px;height:100px" /></a></div> 
<ul>
<li><a href="/info/docs/api/funcgen/index.html#api">Perl API</a></li>
<li><a href="/info/docs/api/funcgen/funcgen_schema.html">Database schema</a></li>
</ul>
</td>

</tr>
</table>

<h2>Ensembl Registry</h2>

<p>
The Registry system allows to tell your programs where to find the
EnsEMBL databases and how to connect to them. It has been implemented
for the Ensembl Core and Compara APIs. 
</p>

<p><a href="./registry.html">More about the Registry...</a></p>


<h2 id="support">Ensembl Software Support</h2>

<p>
Ensembl is an open project and we would like to encourage
correspondence and discussions on any subject on any aspect of
Ensembl. Please see the <a href="/info/about/contact/">Ensembl
Contacts</a> page for suitable options for getting in touch with us.
</p>

<p>
If you are interested in undertaking a short-term collaborative
project, our &quot;<a href="/info/about/outreach/geek4aweek.html">Geek
for a Week</a>&quot; scheme allows developers and researchers to work
alongside Ensembl team members. 
</p>


</body>
</html>
