<html>
<head>
<title>Variant Effect Predictor</title>
</head>
<body>

<div class="column-wrapper">
<h1>Variant Effect Predictor</h1>
</div>
<div class="column-wrapper">
  <div class="column-two">
    <div class="column-left">
      <div class="plain-box">
      <a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadVariations" class="modal_link">
      <img src="/i/vep_logo_sm.png" title="Launch the online VEP!" alt="Variant Effect Predictor" /></a>

<p>
The VEP determines the effect of your variants (SNPs, insertions,
deletions, CNVs or structural variants) on genes, transcripts, and
protein sequence, as well as regulatory regions. Simply input the
coordinates of your variants and the nucleotide changes to find out the:</p>
<ul>
<li>genes affected by the variants</li>
<li>location of the variants (e.g. upstream of a transcript, in coding sequence, in non-coding RNA, in regulatory regions)</li>
<li>consequence your variants on the protein sequence (e.g. stop gained, missense, stop lost, frameshift)</li>
<li>known variants that match yours, and the associated minor allele frequencies</li>
<li>SIFT and PolyPhen scores for changes protein sequence</li>
<li>... And more!</li>
</ul>
      <p class="center"><a
href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadVariations"
class="modal_link"><strong><i>Launch the online VEP tool!</i></strong></a></p>
<h2>Sample output</h2>
<p><img src="/i/help/vep_output.jpg" title="Sample VEP results table" alt="screenshot" /></p>

      </div>
    </div>
  </div>
  <div class="column-two">
    <div class="column-right" style="padding-left:2em">
<h2>How to access the Variant Effect Predictor</h2>
<p>
There are two ways to use the Variant Effect Predictor (VEP) for
predicting functional consequences of known and unknown variants:
</p>

<ul>
  <li>web interface | <a href="/info/docs/tools/vep/vep_online.html">Documentation</a> | <a
href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadVariations"
class="modal_link">Launch web interface</a></li>
  <li>standalone perl script | <a href="/info/docs/tools/vep/vep_script.html">Documentation</a> | <a href="http://cvs.sanger.ac.uk/cgi-bin/viewvc.cgi/ensembl-tools/scripts/variant_effect_predictor.tar.gz?view=tar&root=ensembl&pathrev=branch-ensembl-[[SPECIESDEFS::ENSEMBL_VERSION]]">Download latest version</a> <!--</li>
  <li> using Ensembl's perl API | <a href="/info/docs/api/variation/index.html">Documentation</a></li>-->
</ul>

<p>
Use the web version if you have relatively small volumes of data or
prefer to avoid command-line scripts. The script version is more
flexible and allows processing of large volumes of data (it is written
using the Ensembl perl API, and can be incorporated into your own code
and extended).
</p>

<h2>Ensembl Variation</h2>
<p>For a comprehensive list of the consequence types used in Ensembl and the VEP, look at our <a href="/info/genome/variation/predicted_data.html#consequences">documentation</a>.  
It also provides descriptions of the <a href="/info/genome/variation/data_description.html">data and analyses in Ensembl Variation</a>.</p>

<p>
The VEP was formerly known as the SNP Effect Predictor, and was published under
this name. Please reference the following publication:
</p>

<p>McLaren W, Pritchard B, Rios D, Chen Y, Flicek P, Cunningham F.<br />
<strong>Deriving the consequences of genomic variants with the Ensembl API and SNP Effect Predictor.</strong><br />
<em>BMC Bioinformatics</em>26(16):2069-70(2010)<br />
<a href="http://dx.doi.org/10.1093/bioinformatics/btq330"
rel="external">doi:10.1093/bioinformatics/btq330 </a>
</p>


    </div>
  </div>
</div>
