<!--#set var="decor" value="none"--><!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.4.2_06) on Wed Oct 05 14:27:33 BST 2005 -->
<TITLE>
org.ensembl.datamodel (Ensj API)
</TITLE>

<META NAME="keywords" CONTENT="org.ensembl.datamodel package">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="org.ensembl.datamodel (Ensj API)";
}
</SCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Package</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-use.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../org/ensembl/compara/driver/package-summary.html"><B>PREV PACKAGE</B></A>&nbsp;
&nbsp;<A HREF="../../../org/ensembl/driver/package-summary.html"><B>NEXT PACKAGE</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-summary.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<H2>
Package org.ensembl.datamodel
</H2>
Data types that appear in the core ensembl databases such as
Genes, Exons and Karyotypes and various support classes such as
Location.
<P>
<B>See:</B>
<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<A HREF="#package_description"><B>Description</B></A>
<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Interface Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Accessioned.html" title="interface in org.ensembl.datamodel">Accessioned</A></B></TD>
<TD>Object is capable of having an accession (stable id) which is stable across
 Ensembl data releases and a version.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/AffyArray.html" title="interface in org.ensembl.datamodel">AffyArray</A></B></TD>
<TD>An Affymetrix oligo microarray chip which contains probes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/AffyFeature.html" title="interface in org.ensembl.datamodel">AffyFeature</A></B></TD>
<TD>The location where an Affymetrix Probe hits the genome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/AffyProbe.html" title="interface in org.ensembl.datamodel">AffyProbe</A></B></TD>
<TD>An oligo spot from one or more Affymetric micro arrays.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Analysis.html" title="interface in org.ensembl.datamodel">Analysis</A></B></TD>
<TD>Represents a process used to create features such as genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/ArchiveStableID.html" title="interface in org.ensembl.datamodel">ArchiveStableID</A></B></TD>
<TD>Represents a single immutable archive stable ID.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/AssemblyElement.html" title="interface in org.ensembl.datamodel">AssemblyElement</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/AssemblyMapper.html" title="interface in org.ensembl.datamodel">AssemblyMapper</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Attribute.html" title="interface in org.ensembl.datamodel">Attribute</A></B></TD>
<TD>A generic attribute that can be associatted with a features such as
 MiscFeature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Chromosome.html" title="interface in org.ensembl.datamodel">Chromosome</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/ChromosomeBand.html" title="interface in org.ensembl.datamodel">ChromosomeBand</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/DnaDnaAlignFeature.html" title="interface in org.ensembl.datamodel">DnaDnaAlignFeature</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/DnaDnaAlignment.html" title="interface in org.ensembl.datamodel">DnaDnaAlignment</A></B></TD>
<TD>An alignment between part of a genome and a dna hit</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/DnaProteinAlignment.html" title="interface in org.ensembl.datamodel">DnaProteinAlignment</A></B></TD>
<TD>An alignment between part of a genome and a protein.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Exon.html" title="interface in org.ensembl.datamodel">Exon</A></B></TD>
<TD>Exon.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/ExternalDatabase.html" title="interface in org.ensembl.datamodel">ExternalDatabase</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/ExternalRef.html" title="interface in org.ensembl.datamodel">ExternalRef</A></B></TD>
<TD>Reference to object outside ensembl database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Feature.html" title="interface in org.ensembl.datamodel">Feature</A></B></TD>
<TD>Simple genomic feature.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/FeaturePair.html" title="interface in org.ensembl.datamodel">FeaturePair</A></B></TD>
<TD>Feature hit.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Gene.html" title="interface in org.ensembl.datamodel">Gene</A></B></TD>
<TD>Gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/GeneSnapShot.html" title="interface in org.ensembl.datamodel">GeneSnapShot</A></B></TD>
<TD>Represents part of the state of a gene at some point in time.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/KaryotypeBand.html" title="interface in org.ensembl.datamodel">KaryotypeBand</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Locatable.html" title="interface in org.ensembl.datamodel">Locatable</A></B></TD>
<TD>Genome feature with a location on genome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/MappingSession.html" title="interface in org.ensembl.datamodel">MappingSession</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Marker.html" title="interface in org.ensembl.datamodel">Marker</A></B></TD>
<TD>Marker.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/MarkerFeature.html" title="interface in org.ensembl.datamodel">MarkerFeature</A></B></TD>
<TD>Represents a marker feature in the EnsEMBL database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/MiscFeature.html" title="interface in org.ensembl.datamodel">MiscFeature</A></B></TD>
<TD>MiscFeatures are extremely general features with a location and an
 arbitrary group of attributes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/MiscSet.html" title="interface in org.ensembl.datamodel">MiscSet</A></B></TD>
<TD>MiscSets represent classsifications or groupings of MiscFeatures.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Persistent.html" title="interface in org.ensembl.datamodel">Persistent</A></B></TD>
<TD>Capable of storing in persistent storage.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Population.html" title="interface in org.ensembl.datamodel">Population</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/PredictionExon.html" title="interface in org.ensembl.datamodel">PredictionExon</A></B></TD>
<TD>Predicted exon which lacks a gene and UTRs.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/PredictionTranscript.html" title="interface in org.ensembl.datamodel">PredictionTranscript</A></B></TD>
<TD>Transcript consisting of PredictionExons.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/ProteinFeature.html" title="interface in org.ensembl.datamodel">ProteinFeature</A></B></TD>
<TD>An alignment between a part of the genome and a peptide.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Qtl.html" title="interface in org.ensembl.datamodel">Qtl</A></B></TD>
<TD>Represents a quantitative trait locus (Qtl) in the EnsEMBL database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/QtlFeature.html" title="interface in org.ensembl.datamodel">QtlFeature</A></B></TD>
<TD>A QtlFeature represent a region where the Qtl appears in the Ensembl
 database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/QtlSynonym.html" title="interface in org.ensembl.datamodel">QtlSynonym</A></B></TD>
<TD>Synonym for a Qtl.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/RepeatConsensus.html" title="interface in org.ensembl.datamodel">RepeatConsensus</A></B></TD>
<TD>Repeat sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/RepeatFeature.html" title="interface in org.ensembl.datamodel">RepeatFeature</A></B></TD>
<TD>Repetitive sequence on genome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Sequence.html" title="interface in org.ensembl.datamodel">Sequence</A></B></TD>
<TD>Genomic DNA sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/SequenceEdit.html" title="interface in org.ensembl.datamodel">SequenceEdit</A></B></TD>
<TD>A sequence modification action.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/SequenceRegion.html" title="interface in org.ensembl.datamodel">SequenceRegion</A></B></TD>
<TD>Represents a region of sequence, which could be from any co-ordinate system.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/SimpleFeature.html" title="interface in org.ensembl.datamodel">SimpleFeature</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/SnapShot.html" title="interface in org.ensembl.datamodel">SnapShot</A></B></TD>
<TD>A partial representation of a data element's state at some point in time.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/StableIDEvent.html" title="interface in org.ensembl.datamodel">StableIDEvent</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/SubmittedVariation.html" title="interface in org.ensembl.datamodel">SubmittedVariation</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Transcript.html" title="interface in org.ensembl.datamodel">Transcript</A></B></TD>
<TD>Transcript.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/TranscriptSnapShot.html" title="interface in org.ensembl.datamodel">TranscriptSnapShot</A></B></TD>
<TD>Represents part of the state of a transcript at some point in time.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Translation.html" title="interface in org.ensembl.datamodel">Translation</A></B></TD>
<TD>Translation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/TranslationSnapShot.html" title="interface in org.ensembl.datamodel">TranslationSnapShot</A></B></TD>
<TD>Represents part of the state of a translation at some point in time.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/UngappedAlignment.html" title="interface in org.ensembl.datamodel">UngappedAlignment</A></B></TD>
<TD>Alignment to external reference.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Variation.html" title="interface in org.ensembl.datamodel">Variation</A></B></TD>
<TD>A variation such as a SNP.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/VariationFrequency.html" title="interface in org.ensembl.datamodel">VariationFrequency</A></B></TD>
<TD>&nbsp;</TD>
</TR>
</TABLE>
&nbsp;

<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Class Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/AssemblyException.html" title="class in org.ensembl.datamodel">AssemblyException</A></B></TD>
<TD>Defines how some semi-virtual sequence regions (PAR/Haplotypes) are
 partially constructed from other sequence regions.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/AssemblyLocation.html" title="class in org.ensembl.datamodel">AssemblyLocation</A></B></TD>
<TD><B>Deprecated.</B>&nbsp;<I>Create instances of Location directly.</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/ChainedAssemblyMapper.html" title="class in org.ensembl.datamodel">ChainedAssemblyMapper</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/CloneFragmentLocation.html" title="class in org.ensembl.datamodel">CloneFragmentLocation</A></B></TD>
<TD>Location on a Contig (Clone Fragment).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/CoordinateSystem.html" title="class in org.ensembl.datamodel">CoordinateSystem</A></B></TD>
<TD>A genomic co-ordinate system.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/CoordinateSystemMapping.html" title="class in org.ensembl.datamodel">CoordinateSystemMapping</A></B></TD>
<TD>Store information about the mapping between a pair of co-ordinate systems.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A></B></TD>
<TD>A Location is an 'address' of regions on sequences.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/Query.html" title="class in org.ensembl.datamodel">Query</A></B></TD>
<TD><B>Deprecated.</B>&nbsp;<I>since version 27.0</I></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/SimpleAssemblyMapper.html" title="class in org.ensembl.datamodel">SimpleAssemblyMapper</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/TranscriptAlleleHelper.html" title="class in org.ensembl.datamodel">TranscriptAlleleHelper</A></B></TD>
<TD>Denormalised representation of the transcript optimised for answering
 questions about variations.</TD>
</TR>
</TABLE>
&nbsp;

<P>

<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Exception Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/AttributeException.html" title="class in org.ensembl.datamodel">AttributeException</A></B></TD>
<TD>&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD WIDTH="15%"><B><A HREF="../../../org/ensembl/datamodel/InvalidLocationException.html" title="class in org.ensembl.datamodel">InvalidLocationException</A></B></TD>
<TD>Represents the fact that an attempt was made to create or use an
 invalid location.</TD>
</TR>
</TABLE>
&nbsp;

<P>
<A NAME="package_description"><!-- --></A><H2>
Package org.ensembl.datamodel Description
</H2>

<P>
Data types that appear in the core ensembl databases such as
Genes, Exons and Karyotypes and various support classes such as
Location.

<p>
Most of the types are defined as interfaces in this package with
implementing classes in org.ensembl.datamodel.impl.
</p>

<p>
Biological objects such as Exons are Locatable because they have
locations on sequences. 
</p>

<p>
Ensj supports one type of Location which can specify part or all of a
region in a coordinate system. This replaces the AssemblyLocation and
CloneFragmentLocation classes which are deprecated.
</p>
<P>

<P>
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Package</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <FONT CLASS="NavBarFont1">Class</FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-use.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../org/ensembl/compara/driver/package-summary.html"><B>PREV PACKAGE</B></A>&nbsp;
&nbsp;<A HREF="../../../org/ensembl/driver/package-summary.html"><B>NEXT PACKAGE</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="package-summary.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
