<!--#set var="decor" value="none"--><!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.4.2_06) on Wed Oct 05 14:27:37 BST 2005 -->
<TITLE>
AlleleConsequence (Ensj API)
</TITLE>

<META NAME="keywords" CONTENT="org.ensembl.variation.datamodel.AlleleConsequence class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="AlleleConsequence (Ensj API)";
}
</SCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/AlleleConsequence.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../../org/ensembl/variation/datamodel/ConsequenceType.html" title="class in org.ensembl.variation.datamodel"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AlleleConsequence.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.ensembl.variation.datamodel</FONT>
<BR>
Class AlleleConsequence</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../resources/inherit.gif" ALT="extended by">org.ensembl.datamodel.impl.PersistentImpl
      <IMG SRC="../../../../resources/inherit.gif" ALT="extended by">org.ensembl.datamodel.impl.LocatableImpl
          <IMG SRC="../../../../resources/inherit.gif" ALT="extended by"><B>org.ensembl.variation.datamodel.AlleleConsequence</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.lang.Cloneable, java.lang.Comparable, <A HREF="../../../../org/ensembl/datamodel/Locatable.html" title="interface in org.ensembl.datamodel">Locatable</A>, <A HREF="../../../../org/ensembl/datamodel/Persistent.html" title="interface in org.ensembl.datamodel">Persistent</A>, java.io.Serializable</DD>
</DL>
<HR>
<DL>
<DT>public class <B>AlleleConsequence</B><DT>extends org.ensembl.datamodel.impl.LocatableImpl</DL>

<P>
Represents part or all of the effect of a single allele on a transcript.
<P>

<P>
<DL>
<DT><B>Author:</B></DT>
  <DD><a href="mailto:craig@ebi.ac.uk">Craig Melsopp </a></DD>
<DT><B>See Also:</B><DD><A HREF="../../../../serialized-form.html#org.ensembl.variation.datamodel.AlleleConsequence">Serialized Form</A></DL>
<HR>

<P>
<!-- ======== NESTED CLASS SUMMARY ======== -->


<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Field Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#DOWNSTREAM">DOWNSTREAM</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele appears upstream of the transcript.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#ESSENTIAL_SPLICE_SITE">ESSENTIAL_SPLICE_SITE</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele hits the 1st or 2nd bases at the beginning or end of an
 intron.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#FIVE_PRIME_UTR">FIVE_PRIME_UTR</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele affects the 5prime UTR.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#FRAMESHIFT_CODING">FRAMESHIFT_CODING</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele causes a frameshift.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#INTRONIC">INTRONIC</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele appears within an intron.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#NON_SYNONYMOUS_CODING">NON_SYNONYMOUS_CODING</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele affects the peptide.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#SPLICE_SITE">SPLICE_SITE</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele hits the 3rd, 4th, 5th, 6th, 7th or 8th bases at the start or end of an 
 intron OR the 1st or 2nd bases of an exon (adjacent to an intron).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#STOP_GAINED">STOP_GAINED</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele causes a new stop codon.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#STOP_LOST">STOP_LOST</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele causes the loss of a stop codon.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#SYNONYMOUS_CODING">SYNONYMOUS_CODING</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele does not affect the peptide.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#THREE_PRIME_UTR">THREE_PRIME_UTR</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele affects the 3prim UTR.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#UPSTREAM">UPSTREAM</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Allele appears upstream of the transcript.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.ensembl.datamodel.impl.LocatableImpl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Fields inherited from class org.ensembl.datamodel.impl.LocatableImpl</B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>sequence</CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.ensembl.datamodel.impl.PersistentImpl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Fields inherited from class org.ensembl.datamodel.impl.PersistentImpl</B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>driver, internalID</CODE></TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#AlleleConsequence(org.ensembl.datamodel.Location, java.lang.String, java.lang.String)">AlleleConsequence</A></B>(<A HREF="../../../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A>&nbsp;location,
                  java.lang.String&nbsp;type,
                  java.lang.String&nbsp;spliceSite)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Method Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#getSpliceSite()">getSpliceSite</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Type of effect on splice site, if any.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#getType()">getType</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Type of consequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.ensembl.datamodel.impl.LocatableImpl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from class org.ensembl.datamodel.impl.LocatableImpl</B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, compareTo, getLocation, getSequence, setCoordinateSystem, setLocation, setSequence</CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.ensembl.datamodel.impl.PersistentImpl"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from class org.ensembl.datamodel.impl.PersistentImpl</B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>getDriver, getInternalID, sameInternalID, setDriver, setInternalID</CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from class java.lang.Object</B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, finalize, getClass, hashCode, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.ensembl.datamodel.Persistent"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from interface org.ensembl.datamodel.<A HREF="../../../../org/ensembl/datamodel/Persistent.html" title="interface in org.ensembl.datamodel">Persistent</A></B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../org/ensembl/datamodel/Persistent.html#getDriver()">getDriver</A>, <A HREF="../../../../org/ensembl/datamodel/Persistent.html#getInternalID()">getInternalID</A>, <A HREF="../../../../org/ensembl/datamodel/Persistent.html#sameInternalID(org.ensembl.datamodel.Persistent)">sameInternalID</A>, <A HREF="../../../../org/ensembl/datamodel/Persistent.html#setDriver(org.ensembl.driver.CoreDriver)">setDriver</A>, <A HREF="../../../../org/ensembl/datamodel/Persistent.html#setInternalID(long)">setInternalID</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Field Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="UPSTREAM"><!-- --></A><H3>
UPSTREAM</H3>
<PRE>
public static final java.lang.String <B>UPSTREAM</B></PRE>
<DL>
<DD>Allele appears upstream of the transcript.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.UPSTREAM">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="DOWNSTREAM"><!-- --></A><H3>
DOWNSTREAM</H3>
<PRE>
public static final java.lang.String <B>DOWNSTREAM</B></PRE>
<DL>
<DD>Allele appears upstream of the transcript.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.DOWNSTREAM">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="INTRONIC"><!-- --></A><H3>
INTRONIC</H3>
<PRE>
public static final java.lang.String <B>INTRONIC</B></PRE>
<DL>
<DD>Allele appears within an intron.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.INTRONIC">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="STOP_GAINED"><!-- --></A><H3>
STOP_GAINED</H3>
<PRE>
public static final java.lang.String <B>STOP_GAINED</B></PRE>
<DL>
<DD>Allele causes a new stop codon.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.STOP_GAINED">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="STOP_LOST"><!-- --></A><H3>
STOP_LOST</H3>
<PRE>
public static final java.lang.String <B>STOP_LOST</B></PRE>
<DL>
<DD>Allele causes the loss of a stop codon.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.STOP_LOST">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="SYNONYMOUS_CODING"><!-- --></A><H3>
SYNONYMOUS_CODING</H3>
<PRE>
public static final java.lang.String <B>SYNONYMOUS_CODING</B></PRE>
<DL>
<DD>Allele does not affect the peptide.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.SYNONYMOUS_CODING">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="NON_SYNONYMOUS_CODING"><!-- --></A><H3>
NON_SYNONYMOUS_CODING</H3>
<PRE>
public static final java.lang.String <B>NON_SYNONYMOUS_CODING</B></PRE>
<DL>
<DD>Allele affects the peptide.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.NON_SYNONYMOUS_CODING">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="FRAMESHIFT_CODING"><!-- --></A><H3>
FRAMESHIFT_CODING</H3>
<PRE>
public static final java.lang.String <B>FRAMESHIFT_CODING</B></PRE>
<DL>
<DD>Allele causes a frameshift.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.FRAMESHIFT_CODING">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="FIVE_PRIME_UTR"><!-- --></A><H3>
FIVE_PRIME_UTR</H3>
<PRE>
public static final java.lang.String <B>FIVE_PRIME_UTR</B></PRE>
<DL>
<DD>Allele affects the 5prime UTR.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.FIVE_PRIME_UTR">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="THREE_PRIME_UTR"><!-- --></A><H3>
THREE_PRIME_UTR</H3>
<PRE>
public static final java.lang.String <B>THREE_PRIME_UTR</B></PRE>
<DL>
<DD>Allele affects the 3prim UTR.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.THREE_PRIME_UTR">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="SPLICE_SITE"><!-- --></A><H3>
SPLICE_SITE</H3>
<PRE>
public static final java.lang.String <B>SPLICE_SITE</B></PRE>
<DL>
<DD>Allele hits the 3rd, 4th, 5th, 6th, 7th or 8th bases at the start or end of an 
 intron OR the 1st or 2nd bases of an exon (adjacent to an intron).
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.SPLICE_SITE">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="ESSENTIAL_SPLICE_SITE"><!-- --></A><H3>
ESSENTIAL_SPLICE_SITE</H3>
<PRE>
public static final java.lang.String <B>ESSENTIAL_SPLICE_SITE</B></PRE>
<DL>
<DD>Allele hits the 1st or 2nd bases at the beginning or end of an
 intron.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../constant-values.html#org.ensembl.variation.datamodel.AlleleConsequence.ESSENTIAL_SPLICE_SITE">Constant Field Values</A></DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="AlleleConsequence(org.ensembl.datamodel.Location, java.lang.String, java.lang.String)"><!-- --></A><H3>
AlleleConsequence</H3>
<PRE>
public <B>AlleleConsequence</B>(<A HREF="../../../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A>&nbsp;location,
                         java.lang.String&nbsp;type,
                         java.lang.String&nbsp;spliceSite)</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Method Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="getType()"><!-- --></A><H3>
getType</H3>
<PRE>
public java.lang.String <B>getType</B>()</PRE>
<DL>
<DD>Type of consequence.
 
 Legal values listed below.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>type of this consequence.<DT><B>See Also:</B><DD><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#UPSTREAM"><CODE>UPSTREAM</CODE></A>, 
<A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#DOWNSTREAM"><CODE>DOWNSTREAM</CODE></A>, 
<A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#FRAMESHIFT_CODING"><CODE>FRAMESHIFT_CODING</CODE></A>, 
<A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#INTRONIC"><CODE>INTRONIC</CODE></A>, 
<A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#NON_SYNONYMOUS_CODING"><CODE>NON_SYNONYMOUS_CODING</CODE></A>, 
<A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#SYNONYMOUS_CODING"><CODE>SYNONYMOUS_CODING</CODE></A>, 
<A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#STOP_GAINED"><CODE>STOP_GAINED</CODE></A>, 
<A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#STOP_LOST"><CODE>STOP_LOST</CODE></A>, 
<A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#FIVE_PRIME_UTR"><CODE>FIVE_PRIME_UTR</CODE></A>, 
<A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#THREE_PRIME_UTR"><CODE>THREE_PRIME_UTR</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="getSpliceSite()"><!-- --></A><H3>
getSpliceSite</H3>
<PRE>
public java.lang.String <B>getSpliceSite</B>()</PRE>
<DL>
<DD>Type of effect on splice site, if any.
 
 See below for legal values.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>splice site, null if splice site unaffected by allele.<DT><B>See Also:</B><DD><A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#SPLICE_SITE"><CODE>SPLICE_SITE</CODE></A>, 
<A HREF="../../../../org/ensembl/variation/datamodel/AlleleConsequence.html#ESSENTIAL_SPLICE_SITE"><CODE>ESSENTIAL_SPLICE_SITE</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/AlleleConsequence.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../../org/ensembl/variation/datamodel/ConsequenceType.html" title="class in org.ensembl.variation.datamodel"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AlleleConsequence.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
