<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>ensembl-ena: Bio/EnsEMBL/ENA/ENARegistryHelper.pm Source File</title>
<link href="/info/docs/Doxygen/doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('ENARegistryHelper_8pm.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">ENARegistryHelper.pm</div>  </div>
</div>
<div class="contents">
<a href="ENARegistryHelper_8pm.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 
<a name="l00002"></a>00002 =pod
<a name="l00003"></a>00003 =head1 LICENSE
<a name="l00004"></a>00004 
<a name="l00005"></a>00005   Copyright (c) 1999-2011 The European Bioinformatics Institute and
<a name="l00006"></a>00006   Genome Research Limited.  All rights reserved.
<a name="l00007"></a>00007 
<a name="l00008"></a>00008   This software is distributed under a modified Apache license.
<a name="l00009"></a>00009   For license details, please see
<a name="l00010"></a>00010 
<a name="l00011"></a>00011     http:<span class="comment">//www.ensembl.org/info/about/code_licence.html</span>
<a name="l00012"></a>00012 
<a name="l00013"></a>00013 =head1 CONTACT
<a name="l00014"></a>00014 
<a name="l00015"></a>00015   Please email comments or questions to the <span class="keyword">public</span> Ensembl
<a name="l00016"></a>00016   developers list at &lt;dev@ensembl.org&gt;.
<a name="l00017"></a>00017 
<a name="l00018"></a>00018   Questions may also be sent to the Ensembl help desk at
<a name="l00019"></a>00019   &lt;helpdesk@ensembl.org&gt;.
<a name="l00020"></a>00020  
<a name="l00021"></a>00021 =cut
<a name="l00022"></a>00022 
<a name="l00023"></a>00023 <span class="keyword">package </span>Bio::EnsEMBL::ENA::ENARegistryHelper;
<a name="l00024"></a>00024 
<a name="l00025"></a>00025 use warnings;
<a name="l00026"></a>00026 use strict;
<a name="l00027"></a>00027 use Bio::EnsEMBL::ApiVersion;
<a name="l00028"></a>00028 use Bio::EnsEMBL::Utils::ConfigRegistry;
<a name="l00029"></a>00029 use Bio::EnsEMBL::DBSQL::DBAdaptor;
<a name="l00030"></a>00030 use Bio::EnsEMBL::Utils::Argument qw(rearrange);
<a name="l00031"></a>00031 use Bio::EnsEMBL::Utils::Scalar qw(assert_ref check_ref);
<a name="l00032"></a>00032 use Bio::EnsEMBL::Utils::Exception qw(<span class="keywordflow">throw</span> warning);
<a name="l00033"></a>00033 use Scalar::Util qw(weaken);    #Used to not hold a strong ref to DBConnection
<a name="l00034"></a>00034 use DBI;
<a name="l00035"></a>00035 use JSON;
<a name="l00036"></a>00036 use LWP::Simple;
<a name="l00037"></a>00037 use Carp;
<a name="l00038"></a>00038 use Data::Dumper;
<a name="l00039"></a>00039 my $default_cache_file = qw/.ena_registry_cache.json/;
<a name="l00040"></a>00040 
<a name="l00041"></a>00041 sub <span class="keyword">new</span> {
<a name="l00042"></a>00042     my ( $class, @args ) = @_;
<a name="l00043"></a>00043     my ( $reg, $url, $file, $nocache, $cache_file, $clear_cache ) =
<a name="l00044"></a>00044       rearrange( [qw(registry url file no_cache cache_file clear_cache)],
<a name="l00045"></a>00045                  @args );
<a name="l00046"></a>00046     <span class="keywordflow">if</span> ( !defined $reg ) { $reg = <span class="stringliteral">&#39;Bio::EnsEMBL::Registry&#39;</span> }
<a name="l00047"></a>00047     my $self = bless( {}, ref($class) || $class );
<a name="l00048"></a>00048     $self-&gt;{registry}      = $reg;
<a name="l00049"></a>00049     $self-&gt;{dbas_by_taxid} = {};
<a name="l00050"></a>00050     $self-&gt;{dbas_by_name}  = {};
<a name="l00051"></a>00051     $self-&gt;{dbas_by_acc}   = {};
<a name="l00052"></a>00052     $self-&gt;{dbas_by_vacc}  = {};
<a name="l00053"></a>00053     $self-&gt;{dbas_by_gc}    = {};
<a name="l00054"></a>00054     $self-&gt;{dbas_by_vgc}   = {};
<a name="l00055"></a>00055     $self-&gt;{dbas}          = [];
<a name="l00056"></a>00056     $cache_file ||= $default_cache_file;
<a name="l00057"></a>00057     $self-&gt;cache_file($cache_file);
<a name="l00058"></a>00058 
<a name="l00059"></a>00059     <span class="keywordflow">if</span> ( defined $clear_cache &amp;&amp; $clear_cache == 1 ) {
<a name="l00060"></a>00060         $self-&gt;clear_cache();
<a name="l00061"></a>00061     }
<a name="l00062"></a>00062     <span class="keywordflow">if</span> ( !$nocache &amp;&amp; !defined $file &amp;&amp; -e $cache_file ) {
<a name="l00063"></a>00063         $file = $cache_file;
<a name="l00064"></a>00064     }
<a name="l00065"></a>00065     <span class="keywordflow">if</span> ( defined $file ) {
<a name="l00066"></a>00066         $self-&gt;load_registry_from_file($file);
<a name="l00067"></a>00067     } elsif ( defined $url ) {
<a name="l00068"></a>00068         $self-&gt;load_registry_from_url($url);
<a name="l00069"></a>00069     } <span class="keywordflow">else</span> {
<a name="l00070"></a>00070         $self-&gt;update_from_registry();
<a name="l00071"></a>00071     }
<a name="l00072"></a>00072     <span class="keywordflow">if</span> ( !$nocache &amp;&amp; !-e $cache_file ) {
<a name="l00073"></a>00073         $self-&gt;write_registry_to_file($cache_file);
<a name="l00074"></a>00074     }
<a name="l00075"></a>00075     <span class="keywordflow">return</span> $self;
<a name="l00076"></a>00076 } ## end sub <span class="keyword">new</span>
<a name="l00077"></a>00077 
<a name="l00078"></a>00078 sub registry {
<a name="l00079"></a>00079     my ( $self, $registry ) = @_;
<a name="l00080"></a>00080     <span class="keywordflow">if</span> ( defined $registry ) {
<a name="l00081"></a>00081         <span class="keywordflow">if</span> ( exists $self-&gt;{registry} ) {
<a name="l00082"></a>00082             <span class="keywordflow">throw</span>(<span class="stringliteral">&#39;Cannot reset the Registry object; already defined &#39;</span>);
<a name="l00083"></a>00083         }
<a name="l00084"></a>00084         $self-&gt;{registry} = $registry;
<a name="l00085"></a>00085         $self-&gt;update_from_registry();
<a name="l00086"></a>00086     }
<a name="l00087"></a>00087     <span class="keywordflow">return</span> $self-&gt;{registry};
<a name="l00088"></a>00088 }
<a name="l00089"></a>00089 
<a name="l00090"></a>00090 sub cache_file {
<a name="l00091"></a>00091     my ( $self, $cache_file ) = @_;
<a name="l00092"></a>00092     <span class="keywordflow">if</span> ( defined $cache_file ) {
<a name="l00093"></a>00093         $self-&gt;{cache_file} = $cache_file;
<a name="l00094"></a>00094     }
<a name="l00095"></a>00095     <span class="keywordflow">return</span> $self-&gt;{cache_file};
<a name="l00096"></a>00096 }
<a name="l00097"></a>00097 
<a name="l00098"></a>00098 sub update_from_registry {
<a name="l00099"></a>00099     my ($self) = @_;
<a name="l00100"></a>00100     $self-&gt;_intern_db_connections();
<a name="l00101"></a>00101     <span class="keywordflow">for</span>
<a name="l00102"></a>00102       my $dba ( @{ $self-&gt;registry()-&gt;get_all_DBAdaptors( -group =&gt; <span class="stringliteral">&#39;core&#39;</span> ) } )
<a name="l00103"></a>00103     {
<a name="l00104"></a>00104         $self-&gt;_hash_dba($dba);
<a name="l00105"></a>00105     }
<a name="l00106"></a>00106     <span class="keywordflow">return</span>;
<a name="l00107"></a>00107 }
<a name="l00108"></a>00108 
<a name="l00109"></a>00109 sub _register_dba {
<a name="l00110"></a>00110     my ( $self, $species, $dba ) = @_;
<a name="l00111"></a>00111     $self-&gt;registry()-&gt;add_DBAdaptor( $species, <span class="stringliteral">&quot;core&quot;</span>, $dba );
<a name="l00112"></a>00112     $self-&gt;_hash_dba($dba);
<a name="l00113"></a>00113     <span class="keywordflow">return</span>;
<a name="l00114"></a>00114 }
<a name="l00115"></a>00115 
<a name="l00116"></a>00116 sub _hash_dba {
<a name="l00117"></a>00117     my ( $self, $dba ) = @_;
<a name="l00118"></a>00118     my $meta       = $dba-&gt;get_MetaContainer();
<a name="l00119"></a>00119     my $taxid      = $meta-&gt;get_taxonomy_id();
<a name="l00120"></a>00120     my $aliases    = $meta-&gt;list_value_by_key(<span class="stringliteral">&#39;species.alias&#39;</span>);
<a name="l00121"></a>00121     my $vgc        = $meta-&gt;single_value_by_key(<span class="stringliteral">&#39;assembly.accession&#39;</span>);
<a name="l00122"></a>00122     my $sa         = $dba-&gt;get_SliceAdaptor();
<a name="l00123"></a>00123     my $accessions = [];
<a name="l00124"></a>00124     <span class="keywordflow">for</span> my $slice ( @{ $sa-&gt;fetch_all(<span class="stringliteral">&quot;seqlevel&quot;</span>) } ) {
<a name="l00125"></a>00125         my $acc = $slice-&gt;seq_region_name();
<a name="l00126"></a>00126         push( @{$accessions}, $acc );
<a name="l00127"></a>00127     }
<a name="l00128"></a>00128     $self-&gt;_hash_dba_from_values( $dba, [$taxid], $aliases, $accessions, $vgc );
<a name="l00129"></a>00129     <span class="keywordflow">return</span>;
<a name="l00130"></a>00130 }
<a name="l00131"></a>00131 
<a name="l00132"></a>00132 sub _hash_dba_from_values {
<a name="l00133"></a>00133     my ( $self, $dba, $taxids, $aliases, $accessions, $vgc ) = @_;
<a name="l00134"></a>00134 
<a name="l00135"></a>00135     <span class="keywordflow">for</span> my $taxid ( @{$taxids} ) {
<a name="l00136"></a>00136         push @{ $self-&gt;{dbas_by_taxid}{$taxid} }, $dba;
<a name="l00137"></a>00137     }
<a name="l00138"></a>00138     <span class="keywordflow">for</span> my $name ( @{$aliases} ) {
<a name="l00139"></a>00139         push @{ $self-&gt;{dbas_by_name}{$name} }, $dba;
<a name="l00140"></a>00140     }
<a name="l00141"></a>00141     <span class="keywordflow">for</span> my $acc ( @{$accessions} ) {
<a name="l00142"></a>00142         <span class="keywordflow">if</span> ( defined $self-&gt;{dbas_by_vacc}{$acc} ) {
<a name="l00143"></a>00143             croak <span class="stringliteral">&quot;DBA with INSDC accession $acc already registered&quot;</span>;
<a name="l00144"></a>00144         }
<a name="l00145"></a>00145         $self-&gt;{dbas_by_vacc}{$acc} = $dba;
<a name="l00146"></a>00146         $acc =~ s/\.[0-9]+$<span class="comment">//x;</span>
<a name="l00147"></a>00147         <span class="keywordflow">if</span> ( defined $self-&gt;{dbas_by_acc}{$acc} ) {
<a name="l00148"></a>00148             croak <span class="stringliteral">&quot;DBA with INSDC accession $acc already registered&quot;</span>;
<a name="l00149"></a>00149         }
<a name="l00150"></a>00150         $self-&gt;{dbas_by_acc}{$acc} = $dba;
<a name="l00151"></a>00151     }
<a name="l00152"></a>00152     <span class="keywordflow">if</span> ( defined $vgc ) {
<a name="l00153"></a>00153         <span class="keywordflow">if</span> ( defined $self-&gt;{dbas_by_vgc}{$vgc} ) {
<a name="l00154"></a>00154             croak <span class="stringliteral">&quot;DBA with GC accession $vgc already registered&quot;</span>;
<a name="l00155"></a>00155         }
<a name="l00156"></a>00156         $self-&gt;{dbas_by_vgc}{$vgc} = $dba;
<a name="l00157"></a>00157         $vgc =~ s/\.[0-9]+$<span class="comment">//x;</span>
<a name="l00158"></a>00158         <span class="keywordflow">if</span> ( defined $self-&gt;{dbas_by_gc}{$vgc} ) {
<a name="l00159"></a>00159             croak <span class="stringliteral">&quot;DBA with GC set chain $vgc already registered&quot;</span>;
<a name="l00160"></a>00160         }
<a name="l00161"></a>00161         $self-&gt;{dbas_by_gc}{$vgc} = $dba;
<a name="l00162"></a>00162     }
<a name="l00163"></a>00163     push @{ $self-&gt;{dbas} }, $dba;
<a name="l00164"></a>00164     <span class="keywordflow">return</span>;
<a name="l00165"></a>00165 } ## end sub _hash_dba_from_values
<a name="l00166"></a>00166 
<a name="l00167"></a>00167 sub _invert_dba_hash {
<a name="l00168"></a>00168     my ($hash) = @_;
<a name="l00169"></a>00169     my $inv_hash;
<a name="l00170"></a>00170     <span class="keywordflow">while</span> ( my ( $item, $dbas ) = each( %{$hash} ) ) {
<a name="l00171"></a>00171         <span class="keywordflow">if</span> ( ref $dbas ne <span class="stringliteral">&#39;ARRAY&#39;</span> ) {
<a name="l00172"></a>00172             $dbas = [$dbas];
<a name="l00173"></a>00173         }
<a name="l00174"></a>00174         <span class="keywordflow">for</span> my $dba ( @{$dbas} ) {
<a name="l00175"></a>00175             my $dba_loc = _dba_to_locator($dba);
<a name="l00176"></a>00176             push @{ $inv_hash-&gt;{$dba_loc} }, $item;
<a name="l00177"></a>00177         }
<a name="l00178"></a>00178     }
<a name="l00179"></a>00179     <span class="keywordflow">return</span> $inv_hash;
<a name="l00180"></a>00180 }
<a name="l00181"></a>00181 
<a name="l00182"></a>00182 sub _registry_to_hash {
<a name="l00183"></a>00183     my ($self) = @_;
<a name="l00184"></a>00184 <span class="preprocessor">    # hash dbcs and dbas by locators</span>
<a name="l00185"></a>00185 <span class="preprocessor"></span>    my $dbc_hash;
<a name="l00186"></a>00186     my $dba_hash;
<a name="l00187"></a>00187     <span class="keywordflow">for</span> my $dba ( @{ $self-&gt;{dbas} } ) {
<a name="l00188"></a>00188         my $dbc_loc = _dbc_to_locator( $dba-&gt;dbc() );
<a name="l00189"></a>00189         $dbc_hash-&gt;{$dbc_loc} = $dba-&gt;dbc();
<a name="l00190"></a>00190         push @{ $dba_hash-&gt;{$dbc_loc} }, $dba;
<a name="l00191"></a>00191     }
<a name="l00192"></a>00192 <span class="preprocessor">    # create auxillary hashes</span>
<a name="l00193"></a>00193 <span class="preprocessor"></span>    my $acc_hash   = _invert_dba_hash( $self-&gt;{dbas_by_vacc} );
<a name="l00194"></a>00194     my $name_hash  = _invert_dba_hash( $self-&gt;{dbas_by_name} );
<a name="l00195"></a>00195     my $taxid_hash = _invert_dba_hash( $self-&gt;{dbas_by_taxid} );
<a name="l00196"></a>00196     my $gc_hash    = _invert_dba_hash( $self-&gt;{dbas_by_vgc} );
<a name="l00197"></a>00197 <span class="preprocessor">    # create array of hashes</span>
<a name="l00198"></a>00198 <span class="preprocessor"></span>    my $out_arr;
<a name="l00199"></a>00199     <span class="keywordflow">while</span> ( my ( $dbc_loc, $dbconn ) = each( %{$dbc_hash} ) ) {
<a name="l00200"></a>00200         my $dbc_h = { driver   =&gt; $dbconn-&gt;driver(),
<a name="l00201"></a>00201                       host     =&gt; $dbconn-&gt;host(),
<a name="l00202"></a>00202                       port     =&gt; $dbconn-&gt;port(),
<a name="l00203"></a>00203                       username =&gt; $dbconn-&gt;username(),
<a name="l00204"></a>00204                       password =&gt; $dbconn-&gt;password(),
<a name="l00205"></a>00205                       dbname   =&gt; $dbconn-&gt;dbname(),
<a name="l00206"></a>00206                       species  =&gt; [] };
<a name="l00207"></a>00207         <span class="keywordflow">for</span> my $dba ( @{ $dba_hash-&gt;{$dbc_loc} } ) {
<a name="l00208"></a>00208             my $dba_loc = _dba_to_locator($dba);
<a name="l00209"></a>00209             my $gc = $gc_hash-&gt;{$dba_loc};
<a name="l00210"></a>00210             <span class="keywordflow">if</span>(ref $gc eq <span class="stringliteral">&#39;ARRAY&#39;</span>) {
<a name="l00211"></a>00211                 $gc = $gc-&gt;[0];
<a name="l00212"></a>00212             }
<a name="l00213"></a>00213             push @{ $dbc_h-&gt;{species} }, {
<a name="l00214"></a>00214                 species_id =&gt; $dba-&gt;species_id(),
<a name="l00215"></a>00215                 species    =&gt; $dba-&gt;species(),
<a name="l00216"></a>00216                 taxids     =&gt; $taxid_hash-&gt;{$dba_loc},
<a name="l00217"></a>00217                 aliases    =&gt; $name_hash-&gt;{$dba_loc},
<a name="l00218"></a>00218                 accessions =&gt; $acc_hash-&gt;{$dba_loc},
<a name="l00219"></a>00219                 gc         =&gt; $gc};
<a name="l00220"></a>00220         }
<a name="l00221"></a>00221         push @{$out_arr}, $dbc_h;
<a name="l00222"></a>00222     }
<a name="l00223"></a>00223     <span class="keywordflow">return</span> $out_arr;
<a name="l00224"></a>00224 } ## end sub _registry_to_hash
<a name="l00225"></a>00225 
<a name="l00226"></a>00226 sub _registry_to_json {
<a name="l00227"></a>00227     my ($self) = @_;
<a name="l00228"></a>00228     my $out_arr = $self-&gt;_registry_to_hash();
<a name="l00229"></a>00229     my $json;
<a name="l00230"></a>00230     <span class="keywordflow">if</span> ($out_arr) {
<a name="l00231"></a>00231         $json = encode_json($out_arr);
<a name="l00232"></a>00232     } <span class="keywordflow">else</span> {
<a name="l00233"></a>00233         carp(<span class="stringliteral">&#39;No DBAs found&#39;</span>);
<a name="l00234"></a>00234     }
<a name="l00235"></a>00235     <span class="keywordflow">return</span> $json;
<a name="l00236"></a>00236 }
<a name="l00237"></a>00237 
<a name="l00238"></a>00238 sub write_registry_to_file {
<a name="l00239"></a>00239     my ( $self, $file ) = @_;
<a name="l00240"></a>00240     my $json = $self-&gt;_registry_to_json();
<a name="l00241"></a>00241     <span class="keywordflow">if</span> ($json) {
<a name="l00242"></a>00242         open( my $fh, <span class="charliteral">&#39;&gt;&#39;</span>, $file )
<a name="l00243"></a>00243           || croak <span class="stringliteral">&quot;Cannot open &#39;${file}&#39; for writing: $!&quot;</span>;
<a name="l00244"></a>00244         print $fh $json;
<a name="l00245"></a>00245         close($fh);
<a name="l00246"></a>00246     }
<a name="l00247"></a>00247     <span class="keywordflow">return</span>;
<a name="l00248"></a>00248 }
<a name="l00249"></a>00249 
<a name="l00250"></a>00250 sub clear_cache {
<a name="l00251"></a>00251     my ($self) = @_;
<a name="l00252"></a>00252     <span class="keywordflow">if</span> ( defined $self-&gt;cache_file() &amp;&amp; -e $self-&gt;cache_file() ) {
<a name="l00253"></a>00253         unlink $self-&gt;cache_file();
<a name="l00254"></a>00254     }
<a name="l00255"></a>00255     <span class="keywordflow">return</span>;
<a name="l00256"></a>00256 }
<a name="l00257"></a>00257 
<a name="l00258"></a>00258 sub _load_registry_from_json {
<a name="l00259"></a>00259     my ( $self, $json ) = @_;
<a name="l00260"></a>00260     my $reg_arr = decode_json($json);
<a name="l00261"></a>00261     <span class="keywordflow">for</span> my $dbc_h ( @{$reg_arr} ) {
<a name="l00262"></a>00262         my $dbc =
<a name="l00263"></a>00263           Bio::EnsEMBL::DBSQL::DBConnection-&gt;new(
<a name="l00264"></a>00264                                                 -driver   =&gt; $dbc_h-&gt;{driver},
<a name="l00265"></a>00265                                                 -host     =&gt; $dbc_h-&gt;{host},
<a name="l00266"></a>00266                                                 -port     =&gt; $dbc_h-&gt;{port},
<a name="l00267"></a>00267                                                 -user     =&gt; $dbc_h-&gt;{username},
<a name="l00268"></a>00268                                                 -password =&gt; $dbc_h-&gt;{password},
<a name="l00269"></a>00269                                                 -dbname   =&gt; $dbc_h-&gt;{dbname} );
<a name="l00270"></a>00270         <span class="keywordflow">for</span> my $species ( @{ $dbc_h-&gt;{species} } ) {
<a name="l00271"></a>00271             my $dba =
<a name="l00272"></a>00272               Bio::EnsEMBL::DBSQL::DBAdaptor-&gt;new(
<a name="l00273"></a>00273                                           -DBCONN     =&gt; $dbc,
<a name="l00274"></a>00274                                           -species    =&gt; $species-&gt;{species},
<a name="l00275"></a>00275                                           -species_id =&gt; $species-&gt;{species_id},
<a name="l00276"></a>00276                                           -multispecies_db =&gt; 1,
<a name="l00277"></a>00277                                           -group           =&gt; <span class="stringliteral">&#39;core&#39;</span>, );
<a name="l00278"></a>00278 <span class="preprocessor">            # no need to register - found automagically</span>
<a name="l00279"></a>00279 <span class="preprocessor"></span>            my $gc = $species-&gt;{gc};
<a name="l00280"></a>00280             <span class="keywordflow">if</span> ( ref $gc eq <span class="stringliteral">&#39;ARRAY&#39;</span> ) { $gc = $gc-&gt;[0]; }
<a name="l00281"></a>00281             ;
<a name="l00282"></a>00282             $self-&gt;_hash_dba_from_values( $dba, $species-&gt;{taxids},
<a name="l00283"></a>00283                                     $species-&gt;{aliases}, $species-&gt;{accessions},
<a name="l00284"></a>00284                                     $species-&gt;{gc} );
<a name="l00285"></a>00285         }
<a name="l00286"></a>00286     } ## end <span class="keywordflow">for</span> my $dbc_h ( @{$reg_arr...})
<a name="l00287"></a>00287     $self-&gt;_intern_db_connections();
<a name="l00288"></a>00288     <span class="keywordflow">return</span>;
<a name="l00289"></a>00289 } ## end sub _load_registry_from_json
<a name="l00290"></a>00290 
<a name="l00291"></a>00291 sub load_registry_from_file {
<a name="l00292"></a>00292     my ( $self, $file ) = @_;
<a name="l00293"></a>00293     open( my $fh, <span class="charliteral">&#39;&lt;&#39;</span>, $file ) or croak &quot;Cannot open file $file: $!&quot;;
<a name="l00294"></a>00294     my $data = do { local $/ = undef; &lt;$fh&gt; };
<a name="l00295"></a>00295     close($fh);
<a name="l00296"></a>00296     <span class="keywordflow">if</span> ( defined $data &amp;&amp; $data ne <span class="stringliteral">&#39;&#39;</span> ) {
<a name="l00297"></a>00297         $self-&gt;_load_registry_from_json($data);
<a name="l00298"></a>00298     }
<a name="l00299"></a>00299     <span class="keywordflow">return</span>;
<a name="l00300"></a>00300 }
<a name="l00301"></a>00301 
<a name="l00302"></a>00302 sub load_registry_from_url {
<a name="l00303"></a>00303     my ( $self, $url ) = @_;
<a name="l00304"></a>00304     my $json = <span class="keyword">get</span>($url);
<a name="l00305"></a>00305     croak <span class="stringliteral">&quot;Could not retrieve JSON from $url: $@&quot;</span> unless defined $json;
<a name="l00306"></a>00306     $self-&gt;_load_registry_from_json($json);
<a name="l00307"></a>00307     <span class="keywordflow">return</span>;
<a name="l00308"></a>00308 }
<a name="l00309"></a>00309 
<a name="l00310"></a>00310 sub dba_to_args {
<a name="l00311"></a>00311     my ( $self, $dba ) = @_;
<a name="l00312"></a>00312     my $dbc = $dba-&gt;dbc();
<a name="l00313"></a>00313     my $args = [ -species         =&gt; $dba-&gt;species(),
<a name="l00314"></a>00314                  -multispecies_db =&gt; $dba-&gt;is_multispecies(),
<a name="l00315"></a>00315                  -species_id      =&gt; $dba-&gt;species_id(),
<a name="l00316"></a>00316                  -group           =&gt; $dba-&gt;group(),
<a name="l00317"></a>00317                  -host            =&gt; $dbc-&gt;host(),
<a name="l00318"></a>00318                  -port            =&gt; $dbc-&gt;port(),
<a name="l00319"></a>00319                  -user            =&gt; $dbc-&gt;username(),
<a name="l00320"></a>00320                  -password        =&gt; $dbc-&gt;password(),
<a name="l00321"></a>00321                  -driver          =&gt; $dbc-&gt;driver(),
<a name="l00322"></a>00322                  -dbname          =&gt; $dbc-&gt;dbname() ];
<a name="l00323"></a>00323     <span class="keywordflow">return</span> $args;
<a name="l00324"></a>00324 }
<a name="l00325"></a>00325 
<a name="l00326"></a>00326 sub _dba_to_locator {
<a name="l00327"></a>00327     my ($dba) = @_;
<a name="l00328"></a>00328     confess(<span class="stringliteral">&quot;Argument must be Bio::EnsEMBL::DBSQL::DBAdaptor not &quot;</span> . ref($dba) )
<a name="l00329"></a>00329       unless ref($dba) eq &quot;Bio::EnsEMBL::DBSQL::DBAdaptor&quot;;
<a name="l00330"></a>00330     my $locator =
<a name="l00331"></a>00331       join( q{!-!}, $dba-&gt;species_id(), _dbc_to_locator( $dba-&gt;dbc() ) );
<a name="l00332"></a>00332     <span class="keywordflow">return</span> $locator;
<a name="l00333"></a>00333 }
<a name="l00334"></a>00334 
<a name="l00335"></a>00335 sub _dbc_to_locator {
<a name="l00336"></a>00336     my ($dbc) = @_;
<a name="l00337"></a>00337     my $locator = join( q{!-!},
<a name="l00338"></a>00338                         $dbc-&gt;host(), $dbc-&gt;dbname(), $dbc-&gt;driver(),
<a name="l00339"></a>00339                         $dbc-&gt;port(), $dbc-&gt;username() );
<a name="l00340"></a>00340     <span class="keywordflow">return</span> $locator;
<a name="l00341"></a>00341 }
<a name="l00342"></a>00342 
<a name="l00343"></a>00343 sub _intern_db_connections {
<a name="l00344"></a>00344     my ($self) = @_;
<a name="l00345"></a>00345     my $adaptors = $self-&gt;registry()-&gt;get_all_DBAdaptors();
<a name="l00346"></a>00346     my %dbc_intern;
<a name="l00347"></a>00347     <span class="keywordflow">foreach</span> my $dba ( @{$adaptors} ) {
<a name="l00348"></a>00348         my $dbc     = $dba-&gt;dbc();
<a name="l00349"></a>00349         my $locator = _dbc_to_locator($dbc);
<a name="l00350"></a>00350         <span class="keywordflow">if</span> ( exists $dbc_intern{$locator} ) {
<a name="l00351"></a>00351             $dba-&gt;dbc( $dbc_intern{$locator} );
<a name="l00352"></a>00352         } <span class="keywordflow">else</span> {
<a name="l00353"></a>00353             $dbc_intern{$locator} = $dba-&gt;dbc();
<a name="l00354"></a>00354         }
<a name="l00355"></a>00355     }
<a name="l00356"></a>00356     <span class="keywordflow">return</span>;
<a name="l00357"></a>00357 }
<a name="l00358"></a>00358 
<a name="l00359"></a>00359 sub get_all_DBConnections {
<a name="l00360"></a>00360     my ($self) = @_;
<a name="l00361"></a>00361     my $adaptors = $self-&gt;registry()-&gt;get_all_DBAdaptors();
<a name="l00362"></a>00362     my %dbcs;
<a name="l00363"></a>00363     <span class="keywordflow">foreach</span> my $dba ( @{$adaptors} ) {
<a name="l00364"></a>00364         my $dbc     = $dba-&gt;dbc();
<a name="l00365"></a>00365         $dbcs{_dbc_to_locator($dbc)} = $dbc;
<a name="l00366"></a>00366     }
<a name="l00367"></a>00367     <span class="keywordflow">return</span> [values %dbcs];
<a name="l00368"></a>00368 }
<a name="l00369"></a>00369 
<a name="l00370"></a>00370 sub get_all_DBAdaptors {
<a name="l00371"></a>00371     my ($self) = @_;
<a name="l00372"></a>00372     <span class="keywordflow">return</span> $self-&gt;{dbas};
<a name="l00373"></a>00373 }
<a name="l00374"></a>00374 
<a name="l00375"></a>00375 sub get_by_taxon_id {
<a name="l00376"></a>00376     my ( $self, $id ) = @_;
<a name="l00377"></a>00377     <span class="keywordflow">return</span> $self-&gt;{dbas_by_taxid}{$id}||[];
<a name="l00378"></a>00378 }
<a name="l00379"></a>00379 
<a name="l00380"></a>00380 sub get_by_name_exact {
<a name="l00381"></a>00381     my ( $self, $name ) = @_;
<a name="l00382"></a>00382     <span class="keywordflow">return</span> $self-&gt;{dbas_by_name}{$name};
<a name="l00383"></a>00383 }
<a name="l00384"></a>00384 
<a name="l00385"></a>00385 sub get_by_accession {
<a name="l00386"></a>00386     my ( $self, $acc ) = @_;
<a name="l00387"></a>00387     my $dba = $self-&gt;{dbas_by_vacc}{$acc};
<a name="l00388"></a>00388     <span class="keywordflow">if</span> ( !defined $dba ) {
<a name="l00389"></a>00389         $acc =~ s/\.[0-9]+$<span class="comment">//x;</span>
<a name="l00390"></a>00390         $dba = $self-&gt;{dbas_by_acc}{$acc};
<a name="l00391"></a>00391     }
<a name="l00392"></a>00392     <span class="keywordflow">return</span> $dba;
<a name="l00393"></a>00393 }
<a name="l00394"></a>00394 
<a name="l00395"></a>00395 sub get_by_assembly_accession {
<a name="l00396"></a>00396     my ( $self, $acc ) = @_;
<a name="l00397"></a>00397     my $dba = $self-&gt;{dbas_by_vgc}{$acc};
<a name="l00398"></a>00398     <span class="keywordflow">if</span> ( !defined $dba ) {
<a name="l00399"></a>00399         $acc =~ s/\.[0-9]+$<span class="comment">//x;</span>
<a name="l00400"></a>00400         $dba = $self-&gt;{dbas_by_gc}{$acc};
<a name="l00401"></a>00401     }
<a name="l00402"></a>00402     <span class="keywordflow">return</span> $dba;
<a name="l00403"></a>00403 }
<a name="l00404"></a>00404 
<a name="l00405"></a>00405 sub get_by_name_pattern {
<a name="l00406"></a>00406     my ( $self, $pattern ) = @_;
<a name="l00407"></a>00407     my %dbas;
<a name="l00408"></a>00408     <span class="keywordflow">for</span> my $name ( keys %{ $self-&gt;{dbas_by_name} } ) {
<a name="l00409"></a>00409         <span class="keywordflow">if</span> ( $name =~ m/$pattern/x ) {
<a name="l00410"></a>00410             <span class="keywordflow">for</span> my $dba ( @{ $self-&gt;{dbas_by_name}{$name} } ) {
<a name="l00411"></a>00411                 my $key = _dba_to_locator($dba);
<a name="l00412"></a>00412                 $dbas{$key} = $dba;
<a name="l00413"></a>00413             }
<a name="l00414"></a>00414         }
<a name="l00415"></a>00415     }
<a name="l00416"></a>00416     <span class="keywordflow">return</span> [ values(%dbas) ];
<a name="l00417"></a>00417 }
<a name="l00418"></a>00418 
<a name="l00419"></a>00419 sub register_ena_dbs {
<a name="l00420"></a>00420     my ( $class, $host, $port, $user, $pass, $regexp ) = @_;
<a name="l00421"></a>00421     <span class="keywordflow">if</span> ( !$regexp ) {
<a name="l00422"></a>00422         $regexp =
<a name="l00423"></a>00423           <span class="stringliteral">&#39;ena_[0-9]+_collection_core_[0-9]+_&#39;</span> . software_version() . <span class="stringliteral">&#39;_[0-9]+&#39;</span>;
<a name="l00424"></a>00424     }
<a name="l00425"></a>00425     my $str = <span class="stringliteral">&quot;DBI:mysql:host=$host;port=$port&quot;</span>;
<a name="l00426"></a>00426     my $dbh = DBI-&gt;connect( $str, $user, $pass );
<a name="l00427"></a>00427 
<a name="l00428"></a>00428     my @dbnames =
<a name="l00429"></a>00429       grep { m/$regexp/xi }
<a name="l00430"></a>00430       map  { $_-&gt;[0] } @{ $dbh-&gt;selectall_arrayref(<span class="stringliteral">&#39;SHOW DATABASES&#39;</span>) };
<a name="l00431"></a>00431 
<a name="l00432"></a>00432     <span class="keywordflow">for</span> my $db (@dbnames) {
<a name="l00433"></a>00433         _register_multispecies_core( $host, $port, $user, $pass, $db );
<a name="l00434"></a>00434     }
<a name="l00435"></a>00435     <span class="keywordflow">return</span>;
<a name="l00436"></a>00436 }
<a name="l00437"></a>00437 
<a name="l00438"></a>00438 sub _register_multispecies_core {
<a name="l00439"></a>00439     <span class="keywordflow">return</span> _register_multispecies_x(
<a name="l00440"></a>00440         <span class="stringliteral">&#39;Bio::EnsEMBL::DBSQL::DBAdaptor&#39;</span>,
<a name="l00441"></a>00441         sub {
<a name="l00442"></a>00442             my %args = @_;
<a name="l00443"></a>00443             <span class="keywordflow">return</span> Bio::EnsEMBL::DBSQL::DBAdaptor-&gt;new(%args);
<a name="l00444"></a>00444         },
<a name="l00445"></a>00445         @_ );
<a name="l00446"></a>00446 }
<a name="l00447"></a>00447 
<a name="l00448"></a>00448 sub _register_multispecies_x {
<a name="l00449"></a>00449 
<a name="l00450"></a>00450     my ( $adaptor, $closure, $host, $port, $user, $pass, $instance, $species,
<a name="l00451"></a>00451          $alias_arr_ref, $disconnect_when_idle )
<a name="l00452"></a>00452       = @_;
<a name="l00453"></a>00453     _runtime_include($adaptor);
<a name="l00454"></a>00454     _check_name($instance);
<a name="l00455"></a>00455     my %args =
<a name="l00456"></a>00456       ( _login_hash( $host, $port, $user, $pass ), <span class="stringliteral">&#39;-DBNAME&#39;</span>, $instance );
<a name="l00457"></a>00457     my @species_array = @{ _query_multispecies_db( $species, %args ) };
<a name="l00458"></a>00458     my @dbas;
<a name="l00459"></a>00459 
<a name="l00460"></a>00460     <span class="keywordflow">foreach</span> my $species_args (@species_array) {
<a name="l00461"></a>00461         my %new_args = ( %args, %{$species_args}, <span class="stringliteral">&#39;-MULTISPECIES_DB&#39;</span> =&gt; 1 );
<a name="l00462"></a>00462         $new_args{-DISCONNECT_WHEN_INACTIVE} = 1 <span class="keywordflow">if</span> $disconnect_when_idle;
<a name="l00463"></a>00463         my $dba = $closure-&gt;(%new_args);
<a name="l00464"></a>00464         <span class="keywordflow">if</span> ( defined $species &amp;&amp; defined $alias_arr_ref ) {
<a name="l00465"></a>00465             _add_aliases( $species, $alias_arr_ref );
<a name="l00466"></a>00466         }
<a name="l00467"></a>00467         push( @dbas, $dba );
<a name="l00468"></a>00468     }
<a name="l00469"></a>00469     <span class="keywordflow">return</span> \@dbas;
<a name="l00470"></a>00470 }
<a name="l00471"></a>00471 
<a name="l00472"></a>00472 sub _query_multispecies_db {
<a name="l00473"></a>00473     my ( $species, %db_args ) = @_;
<a name="l00474"></a>00474     my @species_array;
<a name="l00475"></a>00475     eval {
<a name="l00476"></a>00476         my $dbh = Bio::EnsEMBL::DBSQL::DBConnection-&gt;new(%db_args);
<a name="l00477"></a>00477         my $sql;
<a name="l00478"></a>00478         my @params = (<span class="stringliteral">&#39;species.db_name&#39;</span>);
<a name="l00479"></a>00479 
<a name="l00480"></a>00480         <span class="keywordflow">if</span> ( defined $species ) {
<a name="l00481"></a>00481             $sql =
<a name="l00482"></a>00482 <span class="stringliteral">&#39;select meta_value, species_id from meta where meta_key =? and meta_value =?&#39;</span>;
<a name="l00483"></a>00483             push( @params, $species );
<a name="l00484"></a>00484         } <span class="keywordflow">else</span> {
<a name="l00485"></a>00485             $sql =
<a name="l00486"></a>00486 <span class="stringliteral">&#39;select distinct meta_value, species_id from meta where meta_key =? and species_id is not null&#39;</span>;
<a name="l00487"></a>00487         }
<a name="l00488"></a>00488         my $sth = $dbh-&gt;prepare($sql);
<a name="l00489"></a>00489         $sth-&gt;execute(@params);
<a name="l00490"></a>00490         <span class="keywordflow">while</span> ( my $row = $sth-&gt;fetchrow_arrayref() ) {
<a name="l00491"></a>00491             my ( $species_name, $species_id ) = @{$row};
<a name="l00492"></a>00492             push( @species_array, {
<a name="l00493"></a>00493                      <span class="stringliteral">&#39;-SPECIES&#39;</span>    =&gt; $species_name,
<a name="l00494"></a>00494                      <span class="stringliteral">&#39;-SPECIES_ID&#39;</span> =&gt; $species_id } );
<a name="l00495"></a>00495         }
<a name="l00496"></a>00496         $sth-&gt;finish();
<a name="l00497"></a>00497     };
<a name="l00498"></a>00498     <span class="keywordflow">throw</span>(<span class="stringliteral">&quot;Error detected: $@&quot;</span>) <span class="keywordflow">if</span> $@;
<a name="l00499"></a>00499     <span class="keywordflow">return</span> \@species_array;
<a name="l00500"></a>00500 } ## end sub _query_multispecies_db
<a name="l00501"></a>00501 
<a name="l00502"></a>00502 sub _add_aliases {
<a name="l00503"></a>00503     my ( $species_name, $alias_ref ) = @_;
<a name="l00504"></a>00504     <span class="keywordflow">if</span> ( defined $alias_ref ) {
<a name="l00505"></a>00505         <span class="keywordflow">if</span> ( ref($alias_ref) ne <span class="stringliteral">&#39;ARRAY&#39;</span> ) {
<a name="l00506"></a>00506             $alias_ref = [$alias_ref];
<a name="l00507"></a>00507         }
<a name="l00508"></a>00508 
<a name="l00509"></a>00509         <span class="keywordflow">if</span> ( scalar( @{$alias_ref} ) &gt; 0 ) {
<a name="l00510"></a>00510             Bio::EnsEMBL::Utils::ConfigRegistry-&gt;add_alias(
<a name="l00511"></a>00511                                                       -species =&gt; $species_name,
<a name="l00512"></a>00512                                                       -alias   =&gt; $alias_ref );
<a name="l00513"></a>00513         }
<a name="l00514"></a>00514     }
<a name="l00515"></a>00515     <span class="keywordflow">return</span>;
<a name="l00516"></a>00516 }
<a name="l00517"></a>00517 
<a name="l00518"></a>00518 sub _runtime_include {
<a name="l00519"></a>00519     my ($full_module) = @_;
<a name="l00520"></a>00520 <span class="preprocessor">    ## no critic (ProhibitStringyEval)</span>
<a name="l00521"></a>00521 <span class="preprocessor"></span>    eval <span class="stringliteral">&quot;use ${full_module}&quot;</span>
<a name="l00522"></a>00522       || <span class="keywordflow">throw</span>(<span class="stringliteral">&quot;Cannot import module &#39;${full_module}&#39;: $@&quot;</span>)
<a name="l00523"></a>00523       <span class="keywordflow">if</span> $@;
<a name="l00524"></a>00524 <span class="preprocessor">    ## use critic</span>
<a name="l00525"></a>00525 <span class="preprocessor"></span>    <span class="keywordflow">return</span>;
<a name="l00526"></a>00526 }
<a name="l00527"></a>00527 <span class="preprocessor">#Throws a wobbly if the db name is longer than 64 characters</span>
<a name="l00528"></a>00528 <span class="preprocessor"></span>sub _check_name {
<a name="l00529"></a>00529     my ($name) = @_;
<a name="l00530"></a>00530     my $length = length($name);
<a name="l00531"></a>00531     <span class="keywordflow">throw</span>(
<a name="l00532"></a>00532 <span class="stringliteral">&quot;Invalid length for database name used. Max is 64. The name ${name}- was ${length}&quot;</span>
<a name="l00533"></a>00533     ) <span class="keywordflow">if</span> $length &gt; 64;
<a name="l00534"></a>00534     <span class="keywordflow">return</span>;
<a name="l00535"></a>00535 }
<a name="l00536"></a>00536 
<a name="l00537"></a>00537 sub _login_hash {
<a name="l00538"></a>00538     my ( $host, $port, $user, $pass ) = @_;
<a name="l00539"></a>00539     my %details = ( -HOST =&gt; $host, -PORT =&gt; $port, -USER =&gt; $user );
<a name="l00540"></a>00540     $details{-PASS} = $pass <span class="keywordflow">if</span> defined $pass;
<a name="l00541"></a>00541     <span class="keywordflow">return</span> %details;
<a name="l00542"></a>00542 }
<a name="l00543"></a>00543 
<a name="l00544"></a>00544 1;
<a name="l00545"></a>00545 
<a name="l00546"></a>00546 __END__
<a name="l00547"></a>00547 
<a name="l00548"></a>00548 =pod
<a name="l00549"></a>00549 
<a name="l00550"></a>00550 =head1 NAME
<a name="l00551"></a>00551 
<a name="l00552"></a>00552 Bio::EnsEMBL::ENA::ENARegistryHelper
<a name="l00553"></a>00553 
<a name="l00554"></a>00554 =head1 SYNOPSIS
<a name="l00555"></a>00555 
<a name="l00556"></a>00556 <span class="preprocessor"># creation from a database server</span>
<a name="l00557"></a>00557 <span class="preprocessor"></span>Bio::EnsEMBL::ENA::ENARegistryHelper-&gt;register_ena_dbs( $conf-&gt;{host},
<a name="l00558"></a>00558        $conf-&gt;{port}, $conf-&gt;{user}, $conf-&gt;{pass}, $conf-&gt;{db});
<a name="l00559"></a>00559 my $helper = Bio::EnsEMBL::ENA::ENARegistryHelper-&gt;new();
<a name="l00560"></a>00560 my $dbas = $helper-&gt;registry()-&gt;get_all_DBAdaptors();
<a name="l00561"></a>00561 $dbas = $helper-&gt;get_by_taxon_id(388919);
<a name="l00562"></a>00562 $dbas = $helper-&gt;get_by_name_pattern(<span class="stringliteral">&quot;Escherichia.*&quot;</span>);
<a name="l00563"></a>00563 
<a name="l00564"></a>00564 <span class="preprocessor"># creation from a URL</span>
<a name="l00565"></a>00565 <span class="preprocessor"></span>my $helper = Bio::EnsEMBL::ENA::ENARegistryHelper-&gt;new(-URL=&gt;<span class="stringliteral">&#39;http://www.ebi.ac.uk/ena/ena_genomes_registry.json&#39;</span>);
<a name="l00566"></a>00566 
<a name="l00567"></a>00567 =head1 DESCRIPTION
<a name="l00568"></a>00568 
<a name="l00569"></a>00569 This module is a helper that provides additional methods to aid navigating a registry of &gt;2000 species across &gt;10 databases. 
<a name="l00570"></a>00570 It does not replace the Registry but provides some additional methods <span class="keywordflow">for</span> finding species e.g. by searching <span class="keywordflow">for</span> species that 
<a name="l00571"></a>00571 have an alias that match a regular expression, or species which are derived from a specific ENA/INSDC accession, or species
<a name="l00572"></a>00572 that belong to a particular part of the bacterial taxonomy. 
<a name="l00573"></a>00573 
<a name="l00574"></a>00574 There are a number of ways of creating a helper. The simplest and fastest is to supply a URL <span class="keywordflow">for</span> a JSON resource that contains all the details
<a name="l00575"></a>00575 needed to connect to the <span class="keyword">public</span> ENA databases e.g.
<a name="l00576"></a>00576 
<a name="l00577"></a>00577     my $helper = Bio::EnsEMBL::ENA::ENARegistryHelper-&gt;new(-URL=&gt;<span class="stringliteral">&#39;http://www.ebi.ac.uk/ena/ena_genomes_registry.json&#39;</span>);
<a name="l00578"></a>00578 
<a name="l00579"></a>00579 Alternatively, a local file containing the required JSON can be specified instead:
<a name="l00580"></a>00580 
<a name="l00581"></a>00581     my $helper = Bio::EnsEMBL::ENA::ENARegistryHelper-&gt;new(-FILE=&gt;<span class="stringliteral">&quot;/path/to/reg.json&quot;</span>);
<a name="l00582"></a>00582 
<a name="l00583"></a>00583 Finally, a Registry already loaded with the ENA core databases can be supplied:
<a name="l00584"></a>00584 
<a name="l00585"></a>00585     my $helper = Bio::EnsEMBL::ENA::ENARegistryHelper-&gt;new(-REGISTRY=&gt;<span class="stringliteral">&#39;Bio::EnsEMBL::Regsitry&#39;</span>);
<a name="l00586"></a>00586 
<a name="l00587"></a>00587 If the standard Registry is used, the argument can be omitted completely:
<a name="l00588"></a>00588 
<a name="l00589"></a>00589     my $helper = Bio::EnsEMBL::ENA::ENARegistryHelper-&gt;new(-REGISTRY=&gt;<span class="stringliteral">&#39;Bio::EnsEMBL::Regsitry&#39;</span>);
<a name="l00590"></a>00590 
<a name="l00591"></a>00591 To populate the registry with just the ENA databases <span class="keywordflow">for</span> the current software release on a specified server, the following method can be used:
<a name="l00592"></a>00592 
<a name="l00593"></a>00593     Bio::EnsEMBL::ENA::ENARegistryHelper-&gt;register_ena_dbs( $host,
<a name="l00594"></a>00594        $port, $user, $pass);
<a name="l00595"></a>00595 
<a name="l00596"></a>00596 Once a helper has been created, there are various methods to retreive DBAdaptors <span class="keywordflow">for</span> species of interest:
<a name="l00597"></a>00597 
<a name="l00598"></a>00598 1. To find species by name - all DBAdaptors <span class="keywordflow">for</span> species with a name or alias matching the supplied <span class="keywordtype">string</span>:
<a name="l00599"></a>00599 
<a name="l00600"></a>00600     $dbas = $helper-&gt;get_by_name_exact(<span class="stringliteral">&#39;Escherichia coli str. K-12 substr. MG1655&#39;</span>);
<a name="l00601"></a>00601 
<a name="l00602"></a>00602 2. To find species by name pattern - all DBAdaptors <span class="keywordflow">for</span> species with a name or alias matching the supplied regexp:
<a name="l00603"></a>00603 
<a name="l00604"></a>00604     $dbas = $helper-&gt;get_by_name_exact(<span class="stringliteral">&#39;Escherichia coli .*);</span>
<a name="l00605"></a>00605 <span class="stringliteral"></span>
<a name="l00606"></a>00606 <span class="stringliteral">3. To find species with the supplied taxonomy ID:</span>
<a name="l00607"></a>00607 <span class="stringliteral"></span>
<a name="l00608"></a>00608 <span class="stringliteral">    $dbas = $helper-&gt;get_by_taxon_id(388919);</span>
<a name="l00609"></a>00609 <span class="stringliteral">    </span>
<a name="l00610"></a>00610 <span class="stringliteral">4. In coordination with a taxonomy node adaptor to get DBAs for all descendants of a node:</span>
<a name="l00611"></a>00611 <span class="stringliteral"></span>
<a name="l00612"></a>00612 <span class="stringliteral">    my $node = $node_adaptor-&gt;fetch_by_taxon_id(511145);</span>
<a name="l00613"></a>00613 <span class="stringliteral">    for my $child (@{$node_adaptor-&gt;fetch_descendants($node)}) {</span>
<a name="l00614"></a>00614 <span class="stringliteral">        my $dbas = $helper-&gt;get_by_taxon_id($node-&gt;taxon_id())</span>
<a name="l00615"></a>00615 <span class="stringliteral">        if(defined $dbas) {</span>
<a name="l00616"></a>00616 <span class="stringliteral">            for my $dba (@{$dbas}) {</span>
<a name="l00617"></a>00617 <span class="stringliteral">                # do something with the $dba</span>
<a name="l00618"></a>00618 <span class="stringliteral">            }</span>
<a name="l00619"></a>00619 <span class="stringliteral">        }</span>
<a name="l00620"></a>00620 <span class="stringliteral">    }</span>
<a name="l00621"></a>00621 <span class="stringliteral"></span>
<a name="l00622"></a>00622 <span class="stringliteral">The retrieved DBAdaptors can then be used as normal e.g.</span>
<a name="l00623"></a>00623 <span class="stringliteral"></span>
<a name="l00624"></a>00624 <span class="stringliteral">    for my $gene (@{$dba-&gt;get_GeneAdaptor()-&gt;fetch_all_by_biotype(&#39;</span>protein_coding<span class="stringliteral">&#39;)}) {</span>
<a name="l00625"></a>00625 <span class="stringliteral">        print $gene-&gt;external_name.&quot;\n&quot;;</span>
<a name="l00626"></a>00626 <span class="stringliteral">    }</span>
<a name="l00627"></a>00627 <span class="stringliteral"></span>
<a name="l00628"></a>00628 <span class="stringliteral">Once retrieved, the arguments needed for constructing a DBAdaptor directly can be dumped for later use e.g.</span>
<a name="l00629"></a>00629 <span class="stringliteral"></span>
<a name="l00630"></a>00630 <span class="stringliteral">    my $args = $helper-&gt;dba_to_args($dba);</span>
<a name="l00631"></a>00631 <span class="stringliteral">    ... store and retrieve $args for use in another script ... </span>
<a name="l00632"></a>00632 <span class="stringliteral">    my $resurrected_dba = Bio::EnsEMBL::DBSQL::DBAdaptor-&gt;new(@$args);</span>
<a name="l00633"></a>00633 <span class="stringliteral"></span>
<a name="l00634"></a>00634 <span class="stringliteral">=head1 ATTRIBUTES</span>
<a name="l00635"></a>00635 <span class="stringliteral"></span>
<a name="l00636"></a>00636 <span class="stringliteral">=head2 registry</span>
<a name="l00637"></a>00637 <span class="stringliteral"></span>
<a name="l00638"></a>00638 <span class="stringliteral">  Arg [1]     : Registry module to use (default is Bio::EnsEMBL::Registry)</span>
<a name="l00639"></a>00639 <span class="stringliteral">  Description : Sets and retrieves the Registry </span>
<a name="l00640"></a>00640 <span class="stringliteral">  Returntype  : Registry if set; otherwise undef</span>
<a name="l00641"></a>00641 <span class="stringliteral">  Exceptions  : if an attempt is made to set the value more than once</span>
<a name="l00642"></a>00642 <span class="stringliteral">  Status      : Stable</span>
<a name="l00643"></a>00643 <span class="stringliteral"></span>
<a name="l00644"></a>00644 <span class="stringliteral">=head2 cache_file</span>
<a name="l00645"></a>00645 <span class="stringliteral">  Arg [1]     : File to use for local caching</span>
<a name="l00646"></a>00646 <span class="stringliteral">  Description : Sets and retrieves the local cache file </span>
<a name="l00647"></a>00647 <span class="stringliteral">  Returntype  : File name if set; otherwise undef</span>
<a name="l00648"></a>00648 <span class="stringliteral">  Status      : Stable</span>
<a name="l00649"></a>00649 <span class="stringliteral"></span>
<a name="l00650"></a>00650 <span class="stringliteral">=head1 SUBROUTINES/METHODS</span>
<a name="l00651"></a>00651 <span class="stringliteral"></span>
<a name="l00652"></a>00652 <span class="stringliteral">=head2 new</span>
<a name="l00653"></a>00653 <span class="stringliteral"></span>
<a name="l00654"></a>00654 <span class="stringliteral">  Description       : Creates a new instance of this object. </span>
<a name="l00655"></a>00655 <span class="stringliteral">  Arg [-REGISTRY]   : (optional) Registry module to use (default is Bio::EnsEMBL::Registry)</span>
<a name="l00656"></a>00656 <span class="stringliteral">  Arg [-NO_CACHE]   : (optional) int 1</span>
<a name="l00657"></a>00657 <span class="stringliteral">               This option will turn off the use of a local cache file for storing species details</span>
<a name="l00658"></a>00658 <span class="stringliteral">  Arg [-CACHE_FILE] : (optional) String (default .ena_registry_cache.json)</span>
<a name="l00659"></a>00659 <span class="stringliteral">                This option allows use of a user specified local cache file</span>
<a name="l00660"></a>00660 <span class="stringliteral">  Arg [-URL]        : (optional) </span>
<a name="l00661"></a>00661 <span class="stringliteral">              This option allows the use of a remote resource supplying species details in JSON format. This is not used if a local cache exists and -NO_CACHE is not set</span>
<a name="l00662"></a>00662 <span class="stringliteral">  Arg [-FILE]       : (optional) </span>
<a name="l00663"></a>00663 <span class="stringliteral">              This option allows the use of a file supplying species details in JSON format. This is not used if a local cache exists and -NO_CACHE is not set</span>
<a name="l00664"></a>00664 <span class="stringliteral">  Returntype        : Instance of helper</span>
<a name="l00665"></a>00665 <span class="stringliteral">  Status            : Stable</span>
<a name="l00666"></a>00666 <span class="stringliteral"></span>
<a name="l00667"></a>00667 <span class="stringliteral">  Example           : </span>
<a name="l00668"></a>00668 <span class="stringliteral">  my $helper = Bio::EnsEMBL::ENA::ENARegistryHelper-&gt;new(</span>
<a name="l00669"></a>00669 <span class="stringliteral">                                        -REGISTRY =&gt; $reg);</span>
<a name="l00670"></a>00670 <span class="stringliteral"></span>
<a name="l00671"></a>00671 <span class="stringliteral">=head2 get_all_DBAdaptors</span>
<a name="l00672"></a>00672 <span class="stringliteral">    Description : Return all database adaptors that have been retrieved from registry</span>
<a name="l00673"></a>00673 <span class="stringliteral">    Argument    : None</span>
<a name="l00674"></a>00674 <span class="stringliteral">    Exceptions  : None</span>
<a name="l00675"></a>00675 <span class="stringliteral">    Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor</span>
<a name="l00676"></a>00676 <span class="stringliteral"></span>
<a name="l00677"></a>00677 <span class="stringliteral">=head2 get_all_DBConnections</span>
<a name="l00678"></a>00678 <span class="stringliteral">    Description : Return all database connections used by the DBAs retrieved from the registry</span>
<a name="l00679"></a>00679 <span class="stringliteral">    Argument    : None</span>
<a name="l00680"></a>00680 <span class="stringliteral">    Exceptions  : None</span>
<a name="l00681"></a>00681 <span class="stringliteral">    Return type : Arrayref of Bio::EnsEMBL::DBSQL::DBConnection</span>
<a name="l00682"></a>00682 <span class="stringliteral"></span>
<a name="l00683"></a>00683 <span class="stringliteral">=head2 get_by_accession</span>
<a name="l00684"></a>00684 <span class="stringliteral">    Description : Returns the database adaptor that contains a seq_region with the supplied INSDC accession</span>
<a name="l00685"></a>00685 <span class="stringliteral">    Argument    : Int</span>
<a name="l00686"></a>00686 <span class="stringliteral">    Exceptions  : None</span>
<a name="l00687"></a>00687 <span class="stringliteral">    Return type : Bio::EnsEMBL::DBSQL::DatabaseAdaptor</span>
<a name="l00688"></a>00688 <span class="stringliteral">    </span>
<a name="l00689"></a>00689 <span class="stringliteral">=head2 get_by_assembly_accession</span>
<a name="l00690"></a>00690 <span class="stringliteral">    Description : Returns the database adaptor that contains the assembly with the supplied INSDC assembly accession</span>
<a name="l00691"></a>00691 <span class="stringliteral">    Argument    : Int</span>
<a name="l00692"></a>00692 <span class="stringliteral">    Exceptions  : None</span>
<a name="l00693"></a>00693 <span class="stringliteral">    Return type : Bio::EnsEMBL::DBSQL::DatabaseAdaptor</span>
<a name="l00694"></a>00694 <span class="stringliteral"></span>
<a name="l00695"></a>00695 <span class="stringliteral">=head2 get_by_taxon_id</span>
<a name="l00696"></a>00696 <span class="stringliteral">    Description : Returns all database adaptors that have the supplied taxonomy ID</span>
<a name="l00697"></a>00697 <span class="stringliteral">    Argument    : Int</span>
<a name="l00698"></a>00698 <span class="stringliteral">    Exceptions  : None</span>
<a name="l00699"></a>00699 <span class="stringliteral">    Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor</span>
<a name="l00700"></a>00700 <span class="stringliteral"></span>
<a name="l00701"></a>00701 <span class="stringliteral">=head2 get_by_name_exact</span>
<a name="l00702"></a>00702 <span class="stringliteral">    Description : Return all database adaptors that have the supplied string as an alias/name</span>
<a name="l00703"></a>00703 <span class="stringliteral">    Argument    : String</span>
<a name="l00704"></a>00704 <span class="stringliteral">    Exceptions  : None</span>
<a name="l00705"></a>00705 <span class="stringliteral">    Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor</span>
<a name="l00706"></a>00706 <span class="stringliteral"></span>
<a name="l00707"></a>00707 <span class="stringliteral">=head2 get_by_name_pattern</span>
<a name="l00708"></a>00708 <span class="stringliteral">    Description : Return all database adaptors that have an alias/name that match the supplied regexp</span>
<a name="l00709"></a>00709 <span class="stringliteral">    Argument    : String</span>
<a name="l00710"></a>00710 <span class="stringliteral">    Exceptions  : None</span>
<a name="l00711"></a>00711 <span class="stringliteral">    Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor</span>
<a name="l00712"></a>00712 <span class="stringliteral"></span>
<a name="l00713"></a>00713 <span class="stringliteral">=head2 register_ena_dbs</span>
<a name="l00714"></a>00714 <span class="stringliteral">    Description : Helper method to load the registry with all ENA databases on the supplied server</span>
<a name="l00715"></a>00715 <span class="stringliteral">    Argument    : Host</span>
<a name="l00716"></a>00716 <span class="stringliteral">    Argument    : Port</span>
<a name="l00717"></a>00717 <span class="stringliteral">    Argument    : User</span>
<a name="l00718"></a>00718 <span class="stringliteral">    Argument    : Password</span>
<a name="l00719"></a>00719 <span class="stringliteral">    Argument    : (optional) String with ENA database regexp (default is ena_[0-9]+_collection_core_.*)</span>
<a name="l00720"></a>00720 <span class="stringliteral">    Exceptions  : None</span>
<a name="l00721"></a>00721 <span class="stringliteral">    Return type : None</span>
<a name="l00722"></a>00722 <span class="stringliteral"></span>
<a name="l00723"></a>00723 <span class="stringliteral">=head2  update_from_registry</span>
<a name="l00724"></a>00724 <span class="stringliteral">    Description : Update internal hashes of database adaptors by name/taxids from the registry. Invoke when registry has been updated independently.</span>
<a name="l00725"></a>00725 <span class="stringliteral">    Exceptions  : None</span>
<a name="l00726"></a>00726 <span class="stringliteral"></span>
<a name="l00727"></a>00727 <span class="stringliteral">=head2 register_dba</span>
<a name="l00728"></a>00728 <span class="stringliteral">    Description : Add a single DBAdaptor to the registry and the internal hashes of details</span>
<a name="l00729"></a>00729 <span class="stringliteral">    Argument    : Bio::EnsEMBL::DBAdaptor</span>
<a name="l00730"></a>00730 <span class="stringliteral">    Return      : None</span>
<a name="l00731"></a>00731 <span class="stringliteral"></span>
<a name="l00732"></a>00732 <span class="stringliteral">=head2 write_registry_to_file</span>
<a name="l00733"></a>00733 <span class="stringliteral">    Description : Write the contents of the registry and species lists to a JSON file</span>
<a name="l00734"></a>00734 <span class="stringliteral">    Argument    : File name</span>
<a name="l00735"></a>00735 <span class="stringliteral">    Return      : None</span>
<a name="l00736"></a>00736 <span class="stringliteral"></span>
<a name="l00737"></a>00737 <span class="stringliteral">=head2 dba_to_args</span>
<a name="l00738"></a>00738 <span class="stringliteral">    Description : Dump the arguments needed for contructing a DBA</span>
<a name="l00739"></a>00739 <span class="stringliteral">    Argument    : Bio::EnsEMBL::DBAdaptor</span>
<a name="l00740"></a>00740 <span class="stringliteral">    Return      : Arrayref of args</span>
<a name="l00741"></a>00741 <span class="stringliteral"></span>
<a name="l00742"></a>00742 <span class="stringliteral">=head1 INTERNAL METHODS</span>
<a name="l00743"></a>00743 <span class="stringliteral"></span>
<a name="l00744"></a>00744 <span class="stringliteral">=head2 _hash_dba</span>
<a name="l00745"></a>00745 <span class="stringliteral">    Description : Add details from a DBAdaptor to the internal hashes of details</span>
<a name="l00746"></a>00746 <span class="stringliteral">    Argument    : Bio::EnsEMBL::DBAdaptor</span>
<a name="l00747"></a>00747 <span class="stringliteral"></span>
<a name="l00748"></a>00748 <span class="stringliteral">=head2 _hash_dba_from_values</span>
<a name="l00749"></a>00749 <span class="stringliteral">    Description : Add supplied details to the internal hashes of details</span>
<a name="l00750"></a>00750 <span class="stringliteral">    Argument    : Bio::EnsEMBL::DBAdaptor to hash</span>
<a name="l00751"></a>00751 <span class="stringliteral">    Argument    : Arrayref of taxonomy IDs to use as keys</span>
<a name="l00752"></a>00752 <span class="stringliteral">    Argument    : Arrayref of aliases to use as keys</span>
<a name="l00753"></a>00753 <span class="stringliteral">    Argument    : Arrayref of ENA accessions to use as keys</span>
<a name="l00754"></a>00754 <span class="stringliteral"></span>
<a name="l00755"></a>00755 <span class="stringliteral">=head2 _registry_to_hash</span>
<a name="l00756"></a>00756 <span class="stringliteral">    Description : Generate a hash array structure for the current registry and species details for turning to JSON</span>
<a name="l00757"></a>00757 <span class="stringliteral">    Returns     : Arrayref of hashes</span>
<a name="l00758"></a>00758 <span class="stringliteral"></span>
<a name="l00759"></a>00759 <span class="stringliteral">=head2 _load_registry_from_json</span>
<a name="l00760"></a>00760 <span class="stringliteral">    Description : load the registry from the supplied JSON string</span>
<a name="l00761"></a>00761 <span class="stringliteral">    Argument    : JSON string</span>
<a name="l00762"></a>00762 <span class="stringliteral"></span>
<a name="l00763"></a>00763 <span class="stringliteral">=head2 _dba_to_locator</span>
<a name="l00764"></a>00764 <span class="stringliteral">    Description : return a hash key for a DBAdaptor</span>
<a name="l00765"></a>00765 <span class="stringliteral">    Argument    : Bio::EnsEMBL::DBAdaptor</span>
<a name="l00766"></a>00766 <span class="stringliteral"></span>
<a name="l00767"></a>00767 <span class="stringliteral">=head2 _dbc_to_locator</span>
<a name="l00768"></a>00768 <span class="stringliteral">    Description : return a hash key for a DBConnection</span>
<a name="l00769"></a>00769 <span class="stringliteral">    Argument    : Bio::EnsEMBL::DBConnection</span>
<a name="l00770"></a>00770 <span class="stringliteral"></span>
<a name="l00771"></a>00771 <span class="stringliteral">=head2 _register_multispecies_core</span>
<a name="l00772"></a>00772 <span class="stringliteral">    Description : Register core dbas for all species in the supplied database</span>
<a name="l00773"></a>00773 <span class="stringliteral"></span>
<a name="l00774"></a>00774 <span class="stringliteral">=head2 _register_multispecies_x</span>
<a name="l00775"></a>00775 <span class="stringliteral">    Description : Register specified dba type for all species in the supplied database</span>
<a name="l00776"></a>00776 <span class="stringliteral"></span>
<a name="l00777"></a>00777 <span class="stringliteral">=head2 _query_multispecies_db</span>
<a name="l00778"></a>00778 <span class="stringliteral">    Description : Find all species in the multispecies database</span>
<a name="l00779"></a>00779 <span class="stringliteral"></span>
<a name="l00780"></a>00780 <span class="stringliteral">=head2 _add_aliases</span>
<a name="l00781"></a>00781 <span class="stringliteral">    Description : Registry all aliases</span>
<a name="l00782"></a>00782 <span class="stringliteral">    Argument    : Species name</span>
<a name="l00783"></a>00783 <span class="stringliteral">    Argument    : Arrayref of alias strings</span>
<a name="l00784"></a>00784 <span class="stringliteral"></span>
<a name="l00785"></a>00785 <span class="stringliteral">=head2 _runtime_include</span>
<a name="l00786"></a>00786 <span class="stringliteral">    Description : Load the specified module (usually a DatabaseAdaptor)</span>
<a name="l00787"></a>00787 <span class="stringliteral">    Argument    : Module name</span>
<a name="l00788"></a>00788 <span class="stringliteral"></span>
<a name="l00789"></a>00789 <span class="stringliteral">=head2 _check_name</span>
<a name="l00790"></a>00790 <span class="stringliteral">    Description : Check that the name is not longer than 64 characters</span>
<a name="l00791"></a>00791 <span class="stringliteral">    Argument    : Name to check</span>
<a name="l00792"></a>00792 <span class="stringliteral"></span>
<a name="l00793"></a>00793 <span class="stringliteral">=head2 _login_hash</span>
<a name="l00794"></a>00794 <span class="stringliteral">    Description : Generate dbadaptor login hash from supplied arguments</span>
<a name="l00795"></a>00795 <span class="stringliteral">    Argument    : Host</span>
<a name="l00796"></a>00796 <span class="stringliteral">    Argument    : Port</span>
<a name="l00797"></a>00797 <span class="stringliteral">    Argument    : User</span>
<a name="l00798"></a>00798 <span class="stringliteral">    Argument    : Password</span>
<a name="l00799"></a>00799 <span class="stringliteral"></span>
<a name="l00800"></a>00800 <span class="stringliteral">=head2 _intern_db_connections()</span>
<a name="l00801"></a>00801 <span class="stringliteral">  Description : Go through all available DBAdaptors of registry and ensure they use the same</span>
<a name="l00802"></a>00802 <span class="stringliteral">                DBConnection instance.</span>
<a name="l00803"></a>00803 <span class="stringliteral">  Exceptions  : None</span>
<a name="l00804"></a>00804 <span class="stringliteral">  Status      : At Risk</span>
<a name="l00805"></a>00805 <span class="stringliteral">  </span>
<a name="l00806"></a>00806 <span class="stringliteral">=head1 AUTHOR</span>
<a name="l00807"></a>00807 <span class="stringliteral"></span>
<a name="l00808"></a>00808 <span class="stringliteral">dstaines</span>
<a name="l00809"></a>00809 <span class="stringliteral"></span>
<a name="l00810"></a>00810 <span class="stringliteral">=head1 MAINTANER</span>
<a name="l00811"></a>00811 <span class="stringliteral"></span>
<a name="l00812"></a>00812 <span class="stringliteral">$Author$</span>
<a name="l00813"></a>00813 <span class="stringliteral"></span>
<a name="l00814"></a>00814 <span class="stringliteral">=head1 VERSION</span>
<a name="l00815"></a>00815 <span class="stringliteral"></span>
<a name="l00816"></a>00816 <span class="stringliteral">$Revision$</span>
<a name="l00817"></a>00817 <span class="stringliteral"></span>
<a name="l00818"></a>00818 <span class="stringliteral">=cut</span>
</pre></div></div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="ENARegistryHelper_8pm.html">ENARegistryHelper.pm</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Files</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
