<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>ensembl-ena: Bio/EnsEMBL/ENA/DBSQL/TaxonomyNodeAdaptor.pm Source File</title>
<link href="/info/docs/Doxygen/doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('TaxonomyNodeAdaptor_8pm.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">TaxonomyNodeAdaptor.pm</div>  </div>
</div>
<div class="contents">
<a href="TaxonomyNodeAdaptor_8pm.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 
<a name="l00002"></a>00002 =pod
<a name="l00003"></a>00003 =head1 LICENSE
<a name="l00004"></a>00004 
<a name="l00005"></a>00005   Copyright (c) 1999-2011 The European Bioinformatics Institute and
<a name="l00006"></a>00006   Genome Research Limited.  All rights reserved.
<a name="l00007"></a>00007 
<a name="l00008"></a>00008   This software is distributed under a modified Apache license.
<a name="l00009"></a>00009   For license details, please see
<a name="l00010"></a>00010 
<a name="l00011"></a>00011     http:<span class="comment">//www.ensembl.org/info/about/code_licence.html</span>
<a name="l00012"></a>00012 
<a name="l00013"></a>00013 =head1 CONTACT
<a name="l00014"></a>00014 
<a name="l00015"></a>00015   Please email comments or questions to the <span class="keyword">public</span> Ensembl
<a name="l00016"></a>00016   developers list at &lt;dev@ensembl.org&gt;.
<a name="l00017"></a>00017 
<a name="l00018"></a>00018   Questions may also be sent to the Ensembl help desk at
<a name="l00019"></a>00019   &lt;helpdesk@ensembl.org&gt;.
<a name="l00020"></a>00020  
<a name="l00021"></a>00021 =cut
<a name="l00022"></a>00022 
<a name="l00023"></a>00023 <span class="keyword">package </span>Bio::EnsEMBL::ENA::DBSQL::TaxonomyNodeAdaptor;
<a name="l00024"></a>00024 
<a name="l00025"></a>00025 use strict;
<a name="l00026"></a>00026 use warnings;
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 use Bio::EnsEMBL::Utils::Exception qw( <span class="keywordflow">throw</span> );
<a name="l00029"></a>00029 use Bio::EnsEMBL::Utils::Scalar qw( assert_ref );
<a name="l00030"></a>00030 use Bio::EnsEMBL::Utils::SqlHelper;
<a name="l00031"></a>00031 use Bio::EnsEMBL::ENA::TaxonomyNode;
<a name="l00032"></a>00032 use Data::Dumper;
<a name="l00033"></a>00033 
<a name="l00034"></a>00034 use base qw( Bio::EnsEMBL::DBSQL::BaseAdaptor );
<a name="l00035"></a>00035 
<a name="l00036"></a>00036 sub helper {
<a name="l00037"></a>00037     my ($self) = @_;
<a name="l00038"></a>00038     <span class="keywordflow">if</span> ( !defined $self-&gt;{helper} ) {
<a name="l00039"></a>00039         $self-&gt;{helper} =
<a name="l00040"></a>00040           Bio::EnsEMBL::Utils::SqlHelper-&gt;new( -DB_CONNECTION =&gt; $self-&gt;dbc() );
<a name="l00041"></a>00041     }
<a name="l00042"></a>00042     <span class="keywordflow">return</span> $self-&gt;{helper};
<a name="l00043"></a>00043 }
<a name="l00044"></a>00044 
<a name="l00045"></a>00045 sub mapper {
<a name="l00046"></a>00046     my ($self) = @_;
<a name="l00047"></a>00047     <span class="keywordflow">if</span> ( !defined $self-&gt;{mapper} ) {
<a name="l00048"></a>00048         $self-&gt;{mapper} = sub {
<a name="l00049"></a>00049             my @row   = @{ shift @_ };
<a name="l00050"></a>00050             my $value = shift;
<a name="l00051"></a>00051             <span class="keywordflow">if</span> ( !$value ) {
<a name="l00052"></a>00052                 $value =
<a name="l00053"></a>00053                   Bio::EnsEMBL::ENA::TaxonomyNode-&gt;new( -TAXON_ID    =&gt; $row[0],
<a name="l00054"></a>00054                                                         -PARENT_ID   =&gt; $row[1],
<a name="l00055"></a>00055                                                         -RANK        =&gt; $row[2],
<a name="l00056"></a>00056                                                         -ROOT_ID     =&gt; $row[3],
<a name="l00057"></a>00057                                                         -LEFT_INDEX  =&gt; $row[4],
<a name="l00058"></a>00058                                                         -RIGHT_INDEX =&gt; $row[5],
<a name="l00059"></a>00059                                                         -ADAPTOR     =&gt; $self );
<a name="l00060"></a>00060             }
<a name="l00061"></a>00061             push @{ $value-&gt;names()-&gt;{ $row[6] } }, $row[7];
<a name="l00062"></a>00062             <span class="keywordflow">return</span> $value;
<a name="l00063"></a>00063         };
<a name="l00064"></a>00064     }
<a name="l00065"></a>00065     <span class="keywordflow">return</span> $self-&gt;{mapper};
<a name="l00066"></a>00066 }
<a name="l00067"></a>00067 
<a name="l00068"></a>00068 my $base_sql =
<a name="l00069"></a>00069 q{select n.taxon_id, n.parent_id, n.rank, n.root_id, n.left_index, n.right_index,
<a name="l00070"></a>00070    na.name_class, na.name
<a name="l00071"></a>00071    from ncbi_taxa_node n 
<a name="l00072"></a>00072    join ncbi_taxa_name na <span class="keyword">using</span> (taxon_id)};
<a name="l00073"></a>00073 
<a name="l00074"></a>00074 sub fetch_by_coredbadaptor {
<a name="l00075"></a>00075     my ( $self, $core_dba ) = @_;
<a name="l00076"></a>00076 <span class="preprocessor">    # get taxid from core</span>
<a name="l00077"></a>00077 <span class="preprocessor"></span>    my $taxid = $core_dba-&gt;get_MetaContainer()-&gt;get_taxonomy_id();
<a name="l00078"></a>00078     <span class="keywordflow">if</span> ( !defined $taxid ) {
<a name="l00079"></a>00079         <span class="keywordflow">throw</span>(
<a name="l00080"></a>00080             <span class="stringliteral">&quot;Could not find taxonomy ID for database &quot;</span> . $core_dba-&gt;species() );
<a name="l00081"></a>00081     }
<a name="l00082"></a>00082     my $node = $self-&gt;fetch_by_taxon_id($taxid);
<a name="l00083"></a>00083     <span class="keywordflow">if</span> ($node) {
<a name="l00084"></a>00084         push @{$node-&gt;dba()},$core_dba;
<a name="l00085"></a>00085     }
<a name="l00086"></a>00086     <span class="keywordflow">return</span> $node;
<a name="l00087"></a>00087 }
<a name="l00088"></a>00088 
<a name="l00089"></a>00089 sub fetch_by_taxon_id {
<a name="l00090"></a>00090     my ( $self, $taxon_id ) = @_;
<a name="l00091"></a>00091     my $node =
<a name="l00092"></a>00092       $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00093"></a>00093                                      -SQL =&gt; $base_sql . q{ where taxon_id = ?},
<a name="l00094"></a>00094                                      -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00095"></a>00095                                      -PARAMS   =&gt; [$taxon_id] );
<a name="l00096"></a>00096     <span class="keywordflow">return</span> $node-&gt;{$taxon_id};
<a name="l00097"></a>00097 }
<a name="l00098"></a>00098 
<a name="l00099"></a>00099 sub fetch_by_taxon_ids {
<a name="l00100"></a>00100     my ( $self, $taxon_ids ) = @_;
<a name="l00101"></a>00101     my $list = join <span class="charliteral">&#39;,&#39;</span>, @$taxon_ids;
<a name="l00102"></a>00102     my $node =
<a name="l00103"></a>00103       $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00104"></a>00104                              -SQL =&gt; $base_sql . qq{ where taxon_id in ($list)},
<a name="l00105"></a>00105                              -CALLBACK =&gt; $self-&gt;mapper() );
<a name="l00106"></a>00106     <span class="keywordflow">return</span> [ values %{$node} ];
<a name="l00107"></a>00107 }
<a name="l00108"></a>00108 
<a name="l00109"></a>00109 sub fetch_by_rank {
<a name="l00110"></a>00110     my ( $self, $rank ) = @_;
<a name="l00111"></a>00111     my $node =
<a name="l00112"></a>00112       $self-&gt;helper()-&gt;execute_into_hash(-SQL =&gt; $base_sql . q{ where rank = ?},
<a name="l00113"></a>00113                                          -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00114"></a>00114                                          -PARAMS   =&gt; [$rank] );
<a name="l00115"></a>00115     <span class="keywordflow">return</span> [ values %{$node} ];
<a name="l00116"></a>00116 }
<a name="l00117"></a>00117 
<a name="l00118"></a>00118 sub fetch_by_name {
<a name="l00119"></a>00119     my ( $self, $name ) = @_;
<a name="l00120"></a>00120     my $node =
<a name="l00121"></a>00121       $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00122"></a>00122          -SQL =&gt; $base_sql
<a name="l00123"></a>00123            . q{ join ncbi_taxa_name nan <span class="keyword">using</span> (taxon_id) where nan.name like ?},
<a name="l00124"></a>00124          -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00125"></a>00125          -PARAMS   =&gt; [$name] );
<a name="l00126"></a>00126     <span class="keywordflow">return</span> [ values %{$node} ];
<a name="l00127"></a>00127 }
<a name="l00128"></a>00128 
<a name="l00129"></a>00129 sub fetch_by_name_and_class {
<a name="l00130"></a>00130     my ( $self, $name, $class ) = @_;
<a name="l00131"></a>00131     my $node =
<a name="l00132"></a>00132       $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00133"></a>00133         -SQL =&gt; $base_sql
<a name="l00134"></a>00134           . q{ join ncbi_taxa_name nan <span class="keyword">using</span> (taxon_id) where nan.name like ? and nan.name_class like ?},
<a name="l00135"></a>00135         -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00136"></a>00136         -PARAMS   =&gt; [ $name, $class ] );
<a name="l00137"></a>00137     <span class="keywordflow">return</span> [ values %{$node} ];
<a name="l00138"></a>00138 }
<a name="l00139"></a>00139 
<a name="l00140"></a>00140 sub fetch_by_name_and_rank {
<a name="l00141"></a>00141     my ( $self, $name, $rank ) = @_;
<a name="l00142"></a>00142     my $node =
<a name="l00143"></a>00143       $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00144"></a>00144         -SQL =&gt; $base_sql
<a name="l00145"></a>00145           . q{ join ncbi_taxa_name nan <span class="keyword">using</span> (taxon_id) where nan.name like ? and n.rank = ?},
<a name="l00146"></a>00146         -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00147"></a>00147         -PARAMS   =&gt; [ $name, $rank ] );
<a name="l00148"></a>00148     <span class="keywordflow">return</span> [ values %{$node} ];
<a name="l00149"></a>00149 }
<a name="l00150"></a>00150 
<a name="l00151"></a>00151 sub fetch_parent {
<a name="l00152"></a>00152     my ( $self, $child ) = @_;
<a name="l00153"></a>00153     my $node =
<a name="l00154"></a>00154       $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00155"></a>00155                                      -SQL =&gt; $base_sql . q{ where taxon_id = ?},
<a name="l00156"></a>00156                                      -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00157"></a>00157                                      -PARAMS   =&gt; [ $child-&gt;parent_id() ] );
<a name="l00158"></a>00158     <span class="keywordflow">return</span> $node-&gt;{ $child-&gt;parent_id() };
<a name="l00159"></a>00159 }
<a name="l00160"></a>00160 
<a name="l00161"></a>00161 sub fetch_root {
<a name="l00162"></a>00162     my ( $self, $child ) = @_;
<a name="l00163"></a>00163     my $node =
<a name="l00164"></a>00164       $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00165"></a>00165                                      -SQL =&gt; $base_sql . q{ where taxon_id = ?},
<a name="l00166"></a>00166                                      -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00167"></a>00167                                      -PARAMS   =&gt; [ $child-&gt;root_id() ] );
<a name="l00168"></a>00168     <span class="keywordflow">return</span> $node-&gt;{ $child-&gt;root_id() };
<a name="l00169"></a>00169 }
<a name="l00170"></a>00170 
<a name="l00171"></a>00171 sub fetch_children {
<a name="l00172"></a>00172     my ( $self, $parent ) = @_;
<a name="l00173"></a>00173     my $node =
<a name="l00174"></a>00174       $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00175"></a>00175                                     -SQL =&gt; $base_sql . q{ where parent_id = ?},
<a name="l00176"></a>00176                                     -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00177"></a>00177                                     -PARAMS   =&gt; [ $parent-&gt;taxon_id() ] );
<a name="l00178"></a>00178     <span class="keywordflow">return</span> [ sort { $a-&gt;num_descendants() &lt;=&gt; $b-&gt;num_descendants() }
<a name="l00179"></a>00179              values %{$node} ];
<a name="l00180"></a>00180 }
<a name="l00181"></a>00181 
<a name="l00182"></a>00182 sub fetch_descendants {
<a name="l00183"></a>00183     my ( $self, $node ) = @_;
<a name="l00184"></a>00184     my $nodes = $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00185"></a>00185         -SQL =&gt; $base_sql . q{ join ncbi_taxa_node parent 
<a name="l00186"></a>00186             on (n.left_index between parent.left_index and parent.right_index and n.taxon_id&lt;&gt;parent.taxon_id)
<a name="l00187"></a>00187             where parent.taxon_id=?},
<a name="l00188"></a>00188         -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00189"></a>00189         -PARAMS   =&gt; [ $node-&gt;taxon_id() ] );
<a name="l00190"></a>00190     return [ sort { $a-&gt;num_descendants() &lt;=&gt; $b-&gt;num_descendants() }
<a name="l00191"></a>00191              values %{$nodes} ];
<a name="l00192"></a>00192 }
<a name="l00193"></a>00193 
<a name="l00194"></a>00194 sub fetch_leaf_nodes {
<a name="l00195"></a>00195     my ( $self, $node ) = @_;
<a name="l00196"></a>00196     my $nodes = $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00197"></a>00197         -SQL =&gt; $base_sql . q{ join ncbi_taxa_node parent 
<a name="l00198"></a>00198             on (n.left_index between parent.left_index and parent.right_index and n.taxon_id&lt;&gt;parent.taxon_id)
<a name="l00199"></a>00199             where parent.taxon_id=? and n.right_index=n.left_index+1},
<a name="l00200"></a>00200         -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00201"></a>00201         -PARAMS   =&gt; [ $node-&gt;taxon_id() ] );
<a name="l00202"></a>00202     return [ values %{$nodes} ];
<a name="l00203"></a>00203 }
<a name="l00204"></a>00204 
<a name="l00205"></a>00205 sub fetch_ancestors {
<a name="l00206"></a>00206     my ( $self, $node ) = @_;
<a name="l00207"></a>00207     my $nodes = $self-&gt;helper()-&gt;execute_into_hash(
<a name="l00208"></a>00208         -SQL =&gt; $base_sql . q{ join ncbi_taxa_node child
<a name="l00209"></a>00209                                             on (child.left_index between n.left_index and n.right_index and n.taxon_id&lt;&gt;child.taxon_id)
<a name="l00210"></a>00210                                             where child.taxon_id=?},
<a name="l00211"></a>00211         -CALLBACK =&gt; $self-&gt;mapper(),
<a name="l00212"></a>00212         -PARAMS   =&gt; [ $node-&gt;taxon_id() ] );
<a name="l00213"></a>00213     return [ sort { $a-&gt;num_descendants() &lt;=&gt; $b-&gt;num_descendants() }
<a name="l00214"></a>00214              values %{$nodes} ];
<a name="l00215"></a>00215 }
<a name="l00216"></a>00216 
<a name="l00217"></a>00217 sub add_descendants {
<a name="l00218"></a>00218     my ( $self, $node ) = @_;
<a name="l00219"></a>00219 <span class="preprocessor">    # hash nodes</span>
<a name="l00220"></a>00220 <span class="preprocessor"></span>    $self-&gt;associate_nodes( [ $node, @{ $self-&gt;fetch_descendants($node) } ] );
<a name="l00221"></a>00221     <span class="keywordflow">return</span>;
<a name="l00222"></a>00222 }
<a name="l00223"></a>00223 
<a name="l00224"></a>00224 sub associate_nodes {
<a name="l00225"></a>00225     my ( $self, $nodes ) = @_;
<a name="l00226"></a>00226 <span class="preprocessor">    # sort by left index</span>
<a name="l00227"></a>00227 <span class="preprocessor"></span>    my @nodes     = sort { $a-&gt;{left_index} &lt;=&gt; $b-&gt;{left_index} } @$nodes;
<a name="l00228"></a>00228     my $root      = $nodes[0];
<a name="l00229"></a>00229     my $last_node = undef;
<a name="l00230"></a>00230 <span class="preprocessor">    # walk through them finding parents which enclose them</span>
<a name="l00231"></a>00231 <span class="preprocessor"></span><span class="preprocessor">    # reset parent/child relationships</span>
<a name="l00232"></a>00232 <span class="preprocessor"></span>    <span class="keywordflow">for</span> my $node (@nodes) {
<a name="l00233"></a>00233         $node-&gt;{parent_id} = 0;
<a name="l00234"></a>00234         $node-&gt;{parent}    = undef;
<a name="l00235"></a>00235         $node-&gt;{children}  = [];
<a name="l00236"></a>00236         <span class="keywordflow">if</span> ( defined $last_node ) {
<a name="l00237"></a>00237             my $parent = $self-&gt;_get_parent_by_index( $node, $last_node );
<a name="l00238"></a>00238             <span class="keywordflow">if</span> ( defined $parent ) {
<a name="l00239"></a>00239                 $node-&gt;{parent_id} = $parent-&gt;taxon_id();
<a name="l00240"></a>00240                 $node-&gt;{parent}    = $parent;
<a name="l00241"></a>00241                 push @{ $parent-&gt;{children} }, $node;
<a name="l00242"></a>00242             }
<a name="l00243"></a>00243         }
<a name="l00244"></a>00244         $node-&gt;{root_id} = $root-&gt;taxon_id();
<a name="l00245"></a>00245         $node-&gt;{root}    = $root;
<a name="l00246"></a>00246         $last_node       = $node;
<a name="l00247"></a>00247     }
<a name="l00248"></a>00248     $self-&gt;set_num_descendants($root);
<a name="l00249"></a>00249     <span class="keywordflow">return</span> $root;
<a name="l00250"></a>00250 } ## end sub associate_nodes
<a name="l00251"></a>00251 
<a name="l00252"></a>00252 sub set_num_descendants {
<a name="l00253"></a>00253     my ( $self, $node ) = @_;
<a name="l00254"></a>00254     $node-&gt;{num_descendants} = 0;
<a name="l00255"></a>00255     <span class="keywordflow">for</span> my $child ( @{ $node-&gt;children() } ) {
<a name="l00256"></a>00256         $node-&gt;{num_descendants} += 1 + $self-&gt;set_num_descendants($child);
<a name="l00257"></a>00257     }
<a name="l00258"></a>00258     <span class="keywordflow">return</span> $node-&gt;num_descendants();
<a name="l00259"></a>00259 }
<a name="l00260"></a>00260 
<a name="l00261"></a>00261 sub _get_parent_by_index {
<a name="l00262"></a>00262     my ( $self, $leaf_node, $node ) = @_;
<a name="l00263"></a>00263     my $parent;
<a name="l00264"></a>00264     <span class="keywordflow">if</span> (    $node-&gt;{left_index} &lt; $leaf_node-&gt;{left_index}
<a name="l00265"></a>00265          &amp;&amp; $node-&gt;{right_index} &gt; $leaf_node-&gt;{right_index} )
<a name="l00266"></a>00266     {
<a name="l00267"></a>00267         $parent = $node;
<a name="l00268"></a>00268     } elsif ( defined $node-&gt;{parent} ) {
<a name="l00269"></a>00269         $parent = $self-&gt;_get_parent_by_index( $leaf_node, $node-&gt;{parent} );
<a name="l00270"></a>00270     }
<a name="l00271"></a>00271     <span class="keywordflow">return</span> $parent;
<a name="l00272"></a>00272 }
<a name="l00273"></a>00273 
<a name="l00274"></a>00274 sub build_pruned_tree {
<a name="l00275"></a>00275     my ( $self, $root, $leaf_nodes ) = @_;
<a name="l00276"></a>00276 <span class="preprocessor">    # get the test set and winnow</span>
<a name="l00277"></a>00277 <span class="preprocessor"></span>    my @nodes =
<a name="l00278"></a>00278       grep { $_-&gt;taxon_id() != $root-&gt;taxon_id() }
<a name="l00279"></a>00279       @{ $self-&gt;_restrict_set( $leaf_nodes, $self-&gt;fetch_descendants($root) ) };
<a name="l00280"></a>00280     $self-&gt;associate_nodes( [ $root, @nodes ] );
<a name="l00281"></a>00281     <span class="keywordflow">return</span>;
<a name="l00282"></a>00282 }
<a name="l00283"></a>00283 
<a name="l00284"></a>00284 sub collapse_tree {
<a name="l00285"></a>00285     my ( $self, $root, $check ) = @_;
<a name="l00286"></a>00286     my @nodes;
<a name="l00287"></a>00287     $check ||= sub {
<a name="l00288"></a>00288         my ($node) = @_;
<a name="l00289"></a>00289         <span class="keywordflow">return</span> (    $node-&gt;taxon_id() eq $root-&gt;taxon_id()
<a name="l00290"></a>00290                  || ($node-&gt;dba() &amp;&amp; scalar(@{$node-&gt;dba()}) &gt; 0)
<a name="l00291"></a>00291                  || scalar(@{$node-&gt;children()})!=1
<a name="l00292"></a>00292                  || scalar( @{ $node-&gt;parent()-&gt;children() })&gt;1 );
<a name="l00293"></a>00293     };
<a name="l00294"></a>00294     $root-&gt;traverse_tree(
<a name="l00295"></a>00295         sub {
<a name="l00296"></a>00296             my ( $node, $depth ) = @_;
<a name="l00297"></a>00297             <span class="keywordflow">if</span> ( $check-&gt;($node) ) {
<a name="l00298"></a>00298                 push @nodes, $node;
<a name="l00299"></a>00299             }
<a name="l00300"></a>00300             <span class="keywordflow">return</span>;
<a name="l00301"></a>00301         } );
<a name="l00302"></a>00302     $self-&gt;associate_nodes( [ $root, @nodes ] );
<a name="l00303"></a>00303     <span class="keywordflow">return</span>;
<a name="l00304"></a>00304 }
<a name="l00305"></a>00305 
<a name="l00306"></a>00306 sub _build_node_array {
<a name="l00307"></a>00307     my ( $self, $leaf_nodes ) = @_;
<a name="l00308"></a>00308     my $array;
<a name="l00309"></a>00309     my $lmin = 0;
<a name="l00310"></a>00310     my $rmax = 0;
<a name="l00311"></a>00311     <span class="keywordflow">for</span> my $node (@$leaf_nodes) {
<a name="l00312"></a>00312         <span class="keywordflow">if</span> ( !$node-&gt;{left_index} || !$node-&gt;{right_index} ) {
<a name="l00313"></a>00313             print Dumper($node);
<a name="l00314"></a>00314             <span class="keywordflow">throw</span> <span class="stringliteral">&quot;Indices not set for &quot;</span> . $node-&gt;to_string();
<a name="l00315"></a>00315         }
<a name="l00316"></a>00316         $array-&gt;[ $node-&gt;{left_index} ]  = 1;
<a name="l00317"></a>00317         $array-&gt;[ $node-&gt;{right_index} ] = 1;
<a name="l00318"></a>00318         $lmin =
<a name="l00319"></a>00319           ( $lmin == 0 || $node-&gt;{left_index} &lt; $lmin )
<a name="l00320"></a>00320           ? $node-&gt;{left_index}
<a name="l00321"></a>00321           : $lmin;
<a name="l00322"></a>00322         $rmax = ( $rmax &lt; $node-&gt;{right_index} ) ? $node-&gt;{right_index} : $rmax;
<a name="l00323"></a>00323     }
<a name="l00324"></a>00324     <span class="keywordflow">return</span> [ $array, $lmin, $rmax ];
<a name="l00325"></a>00325 }
<a name="l00326"></a>00326 
<a name="l00327"></a>00327 sub _restrict_set {
<a name="l00328"></a>00328     my ( $self, $leaf_nodes, $input_set ) = @_;
<a name="l00329"></a>00329 <span class="preprocessor">    # build sparse array</span>
<a name="l00330"></a>00330 <span class="preprocessor"></span>    my ( $node_array, $lmin, $rmax ) =
<a name="l00331"></a>00331       @{ $self-&gt;_build_node_array($leaf_nodes) };
<a name="l00332"></a>00332 <span class="preprocessor">    # hash leaf nodes by taxon_id to allow replacement</span>
<a name="l00333"></a>00333 <span class="preprocessor"></span>    my %leaves = map { $_-&gt;taxon_id() =&gt; $_ } @$leaf_nodes;
<a name="l00334"></a>00334     my @output_set;
<a name="l00335"></a>00335 <span class="preprocessor">    # iterate over everything in the min/max range</span>
<a name="l00336"></a>00336 <span class="preprocessor"></span>    <span class="keywordflow">for</span> my $node (
<a name="l00337"></a>00337         grep {
<a name="l00338"></a>00338                  $_-&gt;{right_index} &gt;= $lmin
<a name="l00339"></a>00339               || $_-&gt;{left_index} &lt;= $rmax
<a name="l00340"></a>00340         } @$input_set )
<a name="l00341"></a>00341     {
<a name="l00342"></a>00342         my $leaf = $leaves{ $node-&gt;taxon_id() };
<a name="l00343"></a>00343         <span class="keywordflow">if</span> ( defined $leaf ) {
<a name="l00344"></a>00344             push @output_set, $leaf;
<a name="l00345"></a>00345         } <span class="keywordflow">else</span> {
<a name="l00346"></a>00346             <span class="keywordflow">for</span> ( my $i = $node-&gt;{left_index};
<a name="l00347"></a>00347                   $i &lt;= $node-&gt;{right_index}; $i++ )
<a name="l00348"></a>00348             {
<a name="l00349"></a>00349                 <span class="keywordflow">if</span> ( $node_array-&gt;[$i] ) {
<a name="l00350"></a>00350                     push @output_set, $node;
<a name="l00351"></a>00351                     last;
<a name="l00352"></a>00352                 }
<a name="l00353"></a>00353             }
<a name="l00354"></a>00354         }
<a name="l00355"></a>00355     }
<a name="l00356"></a>00356     <span class="keywordflow">return</span> \@output_set;
<a name="l00357"></a>00357 } ## end sub _restrict_set
<a name="l00358"></a>00358 
<a name="l00359"></a>00359 1;
<a name="l00360"></a>00360 
<a name="l00361"></a>00361 __END__
<a name="l00362"></a>00362 
<a name="l00363"></a>00363 =pod
<a name="l00364"></a>00364 
<a name="l00365"></a>00365 =head1 NAME
<a name="l00366"></a>00366 
<a name="l00367"></a>00367 Bio::EnsEMBL::ENA::DBSQL::TaxonomyNodeAdaptor
<a name="l00368"></a>00368 
<a name="l00369"></a>00369 =head1 SYNOPSIS
<a name="l00370"></a>00370 
<a name="l00371"></a>00371 <span class="preprocessor">#create an adaptor (Registry cannot be used currently)</span>
<a name="l00372"></a>00372 <span class="preprocessor"></span>my $tax_dba =  Bio::EnsEMBL::ENA::DBSQL::TaxonomyDBAdaptor-&gt;new(
<a name="l00373"></a>00373                                       -user   =&gt; $tax_user,
<a name="l00374"></a>00374                                       -pass   =&gt; $tax_pass,
<a name="l00375"></a>00375                                       -dbname =&gt; $tax_db,
<a name="l00376"></a>00376                                       -host   =&gt; $tax_host,
<a name="l00377"></a>00377                                       -port   =&gt; $tax_port);                                      
<a name="l00378"></a>00378 my $node_adaptor = $tax_dba-&gt;get_TaxonomyNodeAdaptor();
<a name="l00379"></a>00379 <span class="preprocessor"># get a node for an ID</span>
<a name="l00380"></a>00380 <span class="preprocessor"></span>my $node = $node_adaptor-&gt;fetch_by_taxon_id(511145);
<a name="l00381"></a>00381 <span class="preprocessor"># print some info</span>
<a name="l00382"></a>00382 <span class="preprocessor"></span>printf <span class="stringliteral">&quot;Node %d is %s %s\n&quot;</span>,$node-&gt;taxon_id(),$node-&gt;rank(),$node-&gt;names()-&gt;{<span class="stringliteral">&#39;scientific name&#39;</span>}-&gt;[0];
<a name="l00383"></a>00383 <span class="preprocessor"># Finding ancestors</span>
<a name="l00384"></a>00384 <span class="preprocessor"></span>my @lineage = @{$node_adaptor-&gt;fetch_ancestors($node)};
<a name="l00385"></a>00385 <span class="keywordflow">for</span> my $node (@lineage) {
<a name="l00386"></a>00386     printf <span class="stringliteral">&quot;Node %d is %s %s\n&quot;</span>,$node-&gt;taxon_id(),$node-&gt;rank(),$node-&gt;names()-&gt;{<span class="stringliteral">&#39;scientific name&#39;</span>}-&gt;[0];
<a name="l00387"></a>00387 }
<a name="l00388"></a>00388 
<a name="l00389"></a>00389 =head1 DESCRIPTION
<a name="l00390"></a>00390 
<a name="l00391"></a>00391 A database adaptor allowing access to the nodes of the NCBI taxonomy. It is currently managed independently of the registry:
<a name="l00392"></a>00392 my $tax_dba =  Bio::EnsEMBL::ENA::DBSQL::TaxonomyDBAdaptor-&gt;new(
<a name="l00393"></a>00393                                       -user   =&gt; $tax_user,
<a name="l00394"></a>00394                                       -pass   =&gt; $tax_pass,
<a name="l00395"></a>00395                                       -dbname =&gt; $tax_db,
<a name="l00396"></a>00396                                       -host   =&gt; $tax_host,
<a name="l00397"></a>00397                                       -port   =&gt; $tax_port);
<a name="l00398"></a>00398                                                                           
<a name="l00399"></a>00399 my $node_adaptor = $tax_dba-&gt;get_TaxonomyNodeAdaptor();
<a name="l00400"></a>00400 <span class="preprocessor"># or alternatively</span>
<a name="l00401"></a>00401 <span class="preprocessor"></span>my $node_adaptor = Bio::EnsEMBL::ENA::DBSQL::TaxonomyNodeAdaptor-&gt;new($tax_dba);
<a name="l00402"></a>00402 
<a name="l00403"></a>00403 The TaxonomyNodeAdaptor handles TaxonomyNode objects, which represent nodes in the NCBI taxonomy
<a name="l00404"></a>00404 
<a name="l00405"></a>00405 There are two main practical uses <span class="keywordflow">for</span> the :
<a name="l00406"></a>00406 1. Accessing the taxonomy database to find species of interest and to then find those within the ENA databases e.g. to find all 
<a name="l00407"></a>00407 descendants of the node representing Escherichia coli:
<a name="l00408"></a>00408     my $node = $node_adaptor-&gt;fetch_by_taxon_id(562);
<a name="l00409"></a>00409     <span class="keywordflow">for</span> my $child (@{$node_adaptor-&gt;fetch_descendants($node)}) {
<a name="l00410"></a>00410         my $dbas = $helper-&gt;get_by_taxon_id($node-&gt;taxon_id())
<a name="l00411"></a>00411         <span class="keywordflow">if</span>(defined $dbas) {
<a name="l00412"></a>00412             <span class="keywordflow">for</span> my $dba (@{$dbas}) {
<a name="l00413"></a>00413 <span class="preprocessor">                # do something with the $dba</span>
<a name="l00414"></a>00414 <span class="preprocessor"></span>            }
<a name="l00415"></a>00415         }
<a name="l00416"></a>00416     }
<a name="l00417"></a>00417 
<a name="l00418"></a>00418 2. Placing ENA species into a taxonomic hierarchy and manipulating that hierarchy (mainly <span class="keywordflow">for</span> use in the web interface)
<a name="l00419"></a>00419     <span class="preprocessor"># iterate over all dbas from a registry or ENARegistryHelper</span>
<a name="l00420"></a>00420 <span class="preprocessor"></span>    <span class="keywordflow">for</span> my $dba (@{$helper-&gt;get_all_DBAdaptors()}) {
<a name="l00421"></a>00421         my $node = $node_adaptor-&gt;fetch_by_coredbadaptor($dba);
<a name="l00422"></a>00422 <span class="preprocessor">        # add the node to a hash set of leaf nodes if not already there</span>
<a name="l00423"></a>00423 <span class="preprocessor"></span>        <span class="keywordflow">if</span> ( $leaf_nodes-&gt;{ $node-&gt;taxon_id() } ) {
<a name="l00424"></a>00424             $node = $leaf_nodes-&gt;{ $node-&gt;taxon_id() };
<a name="l00425"></a>00425         } <span class="keywordflow">else</span> {
<a name="l00426"></a>00426             $leaf_nodes-&gt;{ $node-&gt;taxon_id() } = $node;
<a name="l00427"></a>00427         }
<a name="l00428"></a>00428 <span class="preprocessor">        # attach the DBA to the node</span>
<a name="l00429"></a>00429 <span class="preprocessor"></span>        push @{$node-&gt;dba()}, $dba;
<a name="l00430"></a>00430     }
<a name="l00431"></a>00431     my @leaf_nodes = values(%$leaf_nodes);
<a name="l00432"></a>00432 <span class="preprocessor">    # get the root of the tree</span>
<a name="l00433"></a>00433 <span class="preprocessor"></span>    my $root = $node_adaptor-&gt;fetch_by_taxon_id(1);
<a name="l00434"></a>00434 <span class="preprocessor">    # get rid of all branches of the taxonomy that don&#39;t involve a node from the dba</span>
<a name="l00435"></a>00435 <span class="preprocessor"></span>    $node_adaptor-&gt;build_pruned_tree( $root, \@leaf_nodes );
<a name="l00436"></a>00436 <span class="preprocessor">    # collapse </span>
<a name="l00437"></a>00437 <span class="preprocessor"></span>    $node_adaptor-&gt;collapse_tree($root);
<a name="l00438"></a>00438 
<a name="l00439"></a>00439 Common tasks involve
<a name="l00440"></a>00440 
<a name="l00441"></a>00441 =head1 ATTRIBUTES
<a name="l00442"></a>00442 
<a name="l00443"></a>00443 =head2 helper
<a name="l00444"></a>00444 
<a name="l00445"></a>00445 =head2 mapper
<a name="l00446"></a>00446 
<a name="l00447"></a>00447 =head1 SUBROUTINES/METHODS
<a name="l00448"></a>00448 
<a name="l00449"></a>00449 =head2 fetch_by_taxon_id
<a name="l00450"></a>00450 
<a name="l00451"></a>00451 Description : Fetch a single node corresponding to the supplied taxon ID.
<a name="l00452"></a>00452 Argument    : Int
<a name="l00453"></a>00453 Return type : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00454"></a>00454 
<a name="l00455"></a>00455 =head2 fetch_by_coredbadaptor
<a name="l00456"></a>00456 
<a name="l00457"></a>00457 Description : Fetch a single node corresponding to the taxonomy ID found in the supplied Ensembl DBAdaptor
<a name="l00458"></a>00458 Argument    : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00459"></a>00459 Return type : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00460"></a>00460 
<a name="l00461"></a>00461 =head2 fetch_by_taxon_ids
<a name="l00462"></a>00462 
<a name="l00463"></a>00463 Description : Fetch an array of nodes corresponding to the supplied taxonomy IDs
<a name="l00464"></a>00464 Argument    : Arrayref of Int
<a name="l00465"></a>00465 Return type : Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00466"></a>00466 
<a name="l00467"></a>00467 =head2 fetch_by_name
<a name="l00468"></a>00468 
<a name="l00469"></a>00469 Description : Fetch an array of nodes which have a name LIKE the supplied String
<a name="l00470"></a>00470 Argument    : Name as String (can contain SQL wildcards)
<a name="l00471"></a>00471 Return type : Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00472"></a>00472 
<a name="l00473"></a>00473 =head2 fetch_by_name_and_class
<a name="l00474"></a>00474 
<a name="l00475"></a>00475 Description : Fetch an array of nodes which have a name and name <span class="keyword">class </span>LIKE the supplied Strings
<a name="l00476"></a>00476 Argument    : Name as String (can contain SQL wildcards)
<a name="l00477"></a>00477 Argument    : Name class as String (can contain SQL wildcards)
<a name="l00478"></a>00478 Return type : Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00479"></a>00479 
<a name="l00480"></a>00480 =head2 fetch_by_rank
<a name="l00481"></a>00481 
<a name="l00482"></a>00482 Description : Fetch an array of nodes which have the supplied rank
<a name="l00483"></a>00483 Argument    : String
<a name="l00484"></a>00484 Return type : Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00485"></a>00485 
<a name="l00486"></a>00486 =head2 fetch_by_name_and_rank
<a name="l00487"></a>00487 
<a name="l00488"></a>00488 Description : Fetch an array of nodes which have the supplied rank and a name LIKE the supplied String
<a name="l00489"></a>00489 Argument    : Name as String (can contain SQL wildcards)
<a name="l00490"></a>00490 Argument    : Rank as String
<a name="l00491"></a>00491 Return type : Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00492"></a>00492 
<a name="l00493"></a>00493 =head2 fetch_parent
<a name="l00494"></a>00494  
<a name="l00495"></a>00495 Description : Fetch a single node corresponding to the parent of the node
<a name="l00496"></a>00496 Argument    : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00497"></a>00497 Return type : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00498"></a>00498 
<a name="l00499"></a>00499 =head2 fetch_root
<a name="l00500"></a>00500 
<a name="l00501"></a>00501 Description : Fetch a single node corresponding to the root of the supplied node
<a name="l00502"></a>00502 Argument    : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00503"></a>00503 Return type : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00504"></a>00504 
<a name="l00505"></a>00505 =head2 fetch_children
<a name="l00506"></a>00506 
<a name="l00507"></a>00507 Description : Fetch an array of nodes for which the parent is the supplied node
<a name="l00508"></a>00508 Argument    : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00509"></a>00509 Return type : Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00510"></a>00510 
<a name="l00511"></a>00511 =head2 fetch_descendants
<a name="l00512"></a>00512 
<a name="l00513"></a>00513 Description : Fetch an array of nodes below the supplied node on the tree
<a name="l00514"></a>00514 Argument    : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00515"></a>00515 Return type : Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00516"></a>00516 
<a name="l00517"></a>00517 =head2 fetch_leaf_nodes
<a name="l00518"></a>00518 
<a name="l00519"></a>00519 Description : Fetch an array of nodes which are the leaves below the supplied node on the tree
<a name="l00520"></a>00520 Argument    : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00521"></a>00521 Return type : Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00522"></a>00522 
<a name="l00523"></a>00523 =head2 fetch_ancestors
<a name="l00524"></a>00524 
<a name="l00525"></a>00525 Description : Fetch an array of nodes which are above the supplied node on the tree (ordered ascending up the tree)
<a name="l00526"></a>00526 Argument    : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00527"></a>00527 Return type : Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00528"></a>00528 
<a name="l00529"></a>00529 =head2 add_descendants
<a name="l00530"></a>00530 
<a name="l00531"></a>00531 Description : Fetch descendants for the supplied node and assemble into a tree
<a name="l00532"></a>00532 Argument    : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00533"></a>00533 Return type : None
<a name="l00534"></a>00534 
<a name="l00535"></a>00535 =head2 associate_nodes
<a name="l00536"></a>00536 
<a name="l00537"></a>00537 Description : Associate the supplied nodes into a tree based on left/right indices and return the root node
<a name="l00538"></a>00538 Argument    : Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00539"></a>00539 Return type : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00540"></a>00540 
<a name="l00541"></a>00541 =head2 set_num_descendants
<a name="l00542"></a>00542 
<a name="l00543"></a>00543 Description : Traverse the supplied tree setting number of descendants based on current tree structure
<a name="l00544"></a>00544 Argument    : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00545"></a>00545 Return type : None
<a name="l00546"></a>00546 
<a name="l00547"></a>00547 =head2 build_pruned_tree
<a name="l00548"></a>00548 Description : Restrict supplied tree by list of key leaf nodes
<a name="l00549"></a>00549 Argument    : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00550"></a>00550 Argument    : Array ref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00551"></a>00551 Return type : None
<a name="l00552"></a>00552 
<a name="l00553"></a>00553 =head2 collapse_tree
<a name="l00554"></a>00554 Description : Remove all nodes from tree that do not fit the required criteria. By default, these are not branch points, the root node, leaf nodes or nodes with database adaptors.
<a name="l00555"></a>00555 Argument    : Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00556"></a>00556 Argument    : Array ref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00557"></a>00557 Argument    : (Optional) Subroutine reference which is used to define whether to retain a node
<a name="l00558"></a>00558 Return type : None
<a name="l00559"></a>00559 
<a name="l00560"></a>00560 =head1 INTERNAL METHODS
<a name="l00561"></a>00561 
<a name="l00562"></a>00562 =head2 _get_parent_by_index
<a name="l00563"></a>00563 Description: Find the parent of a node, initially based on left/right index to find direct parent, and then by recursion through parents to work back to the closest common ancestor.
<a name="l00564"></a>00564 Argument: Current node
<a name="l00565"></a>00565 Argument: Prospective parent
<a name="l00566"></a>00566 Return type: Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00567"></a>00567 
<a name="l00568"></a>00568 =head2 _build_node_array
<a name="l00569"></a>00569 Description: Build a sparse array based on left/right indices of a set of nodes to allow speedy lookup
<a name="l00570"></a>00570 Argument: Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00571"></a>00571 Return type: Arrayref containing arrayref of 1/0 by left/right, min left index, max right index
<a name="l00572"></a>00572 
<a name="l00573"></a>00573 =head2 _restrict_set
<a name="l00574"></a>00574 Description: Restrict the input set to those needed to build a tree including all members of the leaf set
<a name="l00575"></a>00575 Argument: Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode (leaf set)
<a name="l00576"></a>00576 Argument: Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode (input set)
<a name="l00577"></a>00577 Return: Arrayref of Bio::EnsEMBL::ENA::TaxonomyNode
<a name="l00578"></a>00578 
<a name="l00579"></a>00579 =head1 AUTHOR
<a name="l00580"></a>00580 
<a name="l00581"></a>00581 dstaines
<a name="l00582"></a>00582 
<a name="l00583"></a>00583 =head1 MAINTANER
<a name="l00584"></a>00584 
<a name="l00585"></a>00585 $Author$
<a name="l00586"></a>00586 
<a name="l00587"></a>00587 =head1 VERSION
<a name="l00588"></a>00588 
<a name="l00589"></a>00589 $Revision$
<a name="l00590"></a>00590 =cut
</pre></div></div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="TaxonomyNodeAdaptor_8pm.html">TaxonomyNodeAdaptor.pm</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Files</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
