<!--#set var="decor" value="none"--><!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.4.2_06) on Fri Sep 30 17:02:01 BST 2005 -->
<TITLE>
Example (Ensj API)
</TITLE>

<META NAME="keywords" CONTENT="org.ensembl.Example class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="Example (Ensj API)";
}
</SCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Example.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../org/ensembl/Version.html" title="class in org.ensembl"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Example.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.ensembl</FONT>
<BR>
Class Example</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by"><B>org.ensembl.Example</B>
</PRE>
<HR>
<DL>
<DT>public class <B>Example</B><DT>extends java.lang.Object</DL>

<P>
Example code demonstrating how to retrieve data from ensembl databases
 using ensj.
 
 The examples below retrieve data from the latest human core and variation databases on 
 ensembldb.ensembl.org
<P>

<P>
<DL>
<DT><B>Author:</B></DT>
  <DD>Craig Melsopp</DD>
<DT><B>See Also:</B><DD><a href="Example.java.html">Example.java</a> source <a href="Example.java">(txt)</a></DL>
<HR>

<P>
<!-- ======== NESTED CLASS SUMMARY ======== -->


<!-- =========== FIELD SUMMARY =========== -->


<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#Example()">Example</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates an instance of Example with core and varition drivers configured.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Method Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#convertLocationToCoordinateSystemAndGetTheSeqRegionNames(org.ensembl.datamodel.Location, org.ensembl.datamodel.CoordinateSystem)">convertLocationToCoordinateSystemAndGetTheSeqRegionNames</A></B>(<A HREF="../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A>&nbsp;sourceLoc,
                                                         <A HREF="../../org/ensembl/datamodel/CoordinateSystem.html" title="class in org.ensembl.datamodel">CoordinateSystem</A>&nbsp;targetCS)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Converts location into the target coordinate system and prints the results.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#countGenesAndExonsInEachLocation()">countGenesAndExonsInEachLocation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetch the genes and exons in several locations and print
 how many there are in each one.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A>[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#createLocations()">createLocations</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create and then return some ensj locations.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#displayDriversState()">displayDriversState</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Print details about the drivers' configurations.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#fetchDeletedGeneFromArchive()">fetchDeletedGeneFromArchive</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetches information about a gene from the archive.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#fetchExonByInternalID()">fetchExonByInternalID</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetch an exon by it's internal id and print it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#fetchGeneByStableIDAndViewPeptide()">fetchGeneByStableIDAndViewPeptide</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetch a gene by it's stable ID and then print the peptide corresponding to
 its first transcript.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#fetchGenesByExternalRefs()">fetchGenesByExternalRefs</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetch genes corresponding to various external references/identifier such
 as HUGO ids and Affymetrix probeset names.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#fetchGenesByLocation(org.ensembl.datamodel.Location)">fetchGenesByLocation</A></B>(<A HREF="../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A>&nbsp;location)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This method illustrates several ways to retrieve genes from a specified
 location.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#fetchKaryotypes(org.ensembl.datamodel.CoordinateSystem, java.lang.String)">fetchKaryotypes</A></B>(<A HREF="../../org/ensembl/datamodel/CoordinateSystem.html" title="class in org.ensembl.datamodel">CoordinateSystem</A>&nbsp;coordinateSystem,
                java.lang.String&nbsp;chromosomeName)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetches information about karyotypes (chromosome bands) for the specified
 chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#fetchSequenceRegionsSuchAsChromosomeOrContig(org.ensembl.datamodel.CoordinateSystem)">fetchSequenceRegionsSuchAsChromosomeOrContig</A></B>(<A HREF="../../org/ensembl/datamodel/CoordinateSystem.html" title="class in org.ensembl.datamodel">CoordinateSystem</A>&nbsp;coordinateSystem)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetches information about the sequence regions.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#fetchVariationsOverlapingGenes()">fetchVariationsOverlapingGenes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetches all the variations that overlap a particular gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#fetchVariationsOverlapingLocation()">fetchVariationsOverlapingLocation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetches all the variations in a specified location from the variation
 driver.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#main(java.lang.String[])">main</A></B>(java.lang.String[]&nbsp;args)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Prints some debug information before printing the output from the
 example methods in this class.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/ensembl/Example.html#showExternalRefsForAGene()">showExternalRefsForAGene</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fetches a gene from the database and prints a
 summary of it's external refs if it has any.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from class java.lang.Object</B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->


<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="Example()"><!-- --></A><H3>
Example</H3>
<PRE>
public <B>Example</B>()
        throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Creates an instance of Example with core and varition drivers configured.
<P>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE></DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Method Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="main(java.lang.String[])"><!-- --></A><H3>
main</H3>
<PRE>
public static void <B>main</B>(java.lang.String[]&nbsp;args)
                 throws <A HREF="../../org/ensembl/driver/ConfigurationException.html" title="class in org.ensembl.driver">ConfigurationException</A>,
                        <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A>,
                        java.text.ParseException</PRE>
<DL>
<DD>Prints some debug information before printing the output from the
 example methods in this class.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>args</CODE> - ignored.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE> - if a problem occurs retrieving data
<DD><CODE><A HREF="../../org/ensembl/driver/ConfigurationException.html" title="class in org.ensembl.driver">ConfigurationException</A></CODE> - if a problem occurs initialising the driver(s)
<DD><CODE>java.text.ParseException</CODE> - if a problem occurs parsing the string used to construct a
           location</DL>
</DD>
</DL>
<HR>

<A NAME="fetchVariationsOverlapingGenes()"><!-- --></A><H3>
fetchVariationsOverlapingGenes</H3>
<PRE>
public void <B>fetchVariationsOverlapingGenes</B>()
                                    throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Fetches all the variations that overlap a particular gene.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="fetchVariationsOverlapingLocation()"><!-- --></A><H3>
fetchVariationsOverlapingLocation</H3>
<PRE>
public void <B>fetchVariationsOverlapingLocation</B>()
                                       throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A>,
                                              java.text.ParseException</PRE>
<DL>
<DD>Fetches all the variations in a specified location from the variation
 driver.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE>
<DD><CODE>java.text.ParseException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="fetchGenesByExternalRefs()"><!-- --></A><H3>
fetchGenesByExternalRefs</H3>
<PRE>
public void <B>fetchGenesByExternalRefs</B>()
                              throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Fetch genes corresponding to various external references/identifier such
 as HUGO ids and Affymetrix probeset names.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="convertLocationToCoordinateSystemAndGetTheSeqRegionNames(org.ensembl.datamodel.Location, org.ensembl.datamodel.CoordinateSystem)"><!-- --></A><H3>
convertLocationToCoordinateSystemAndGetTheSeqRegionNames</H3>
<PRE>
public void <B>convertLocationToCoordinateSystemAndGetTheSeqRegionNames</B>(<A HREF="../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A>&nbsp;sourceLoc,
                                                                     <A HREF="../../org/ensembl/datamodel/CoordinateSystem.html" title="class in org.ensembl.datamodel">CoordinateSystem</A>&nbsp;targetCS)
                                                              throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Converts location into the target coordinate system and prints the results.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>sourceLoc</CODE> - source location to be converted<DD><CODE>targetCS</CODE> - target coordinate system to convert location into.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE> - if a problem occured during the conversion</DL>
</DD>
</DL>
<HR>

<A NAME="fetchExonByInternalID()"><!-- --></A><H3>
fetchExonByInternalID</H3>
<PRE>
public void <B>fetchExonByInternalID</B>()
                           throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Fetch an exon by it's internal id and print it.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE> - if a problem occured during the retrieval</DL>
</DD>
</DL>
<HR>

<A NAME="createLocations()"><!-- --></A><H3>
createLocations</H3>
<PRE>
public static <A HREF="../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A>[] <B>createLocations</B>()</PRE>
<DL>
<DD>Create and then return some ensj locations.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>array of locations that can used in queries.</DL>
</DD>
</DL>
<HR>

<A NAME="countGenesAndExonsInEachLocation()"><!-- --></A><H3>
countGenesAndExonsInEachLocation</H3>
<PRE>
public void <B>countGenesAndExonsInEachLocation</B>()
                                      throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Fetch the genes and exons in several locations and print
 how many there are in each one.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE> - if problem occurs during retrieval</DL>
</DD>
</DL>
<HR>

<A NAME="fetchGeneByStableIDAndViewPeptide()"><!-- --></A><H3>
fetchGeneByStableIDAndViewPeptide</H3>
<PRE>
public void <B>fetchGeneByStableIDAndViewPeptide</B>()
                                       throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Fetch a gene by it's stable ID and then print the peptide corresponding to
 its first transcript.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE> - if problem occurs during retrieval</DL>
</DD>
</DL>
<HR>

<A NAME="fetchGenesByLocation(org.ensembl.datamodel.Location)"><!-- --></A><H3>
fetchGenesByLocation</H3>
<PRE>
public void <B>fetchGenesByLocation</B>(<A HREF="../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A>&nbsp;location)
                          throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>This method illustrates several ways to retrieve genes from a specified
 location.
 
 With locations containing few genes the memory and speed differences
 between the methods will be small but for locations containing many genes
 the difference is potentially huge.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>location</CODE> - location to fetch genes from
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE> - if problem occurs during retrieval</DL>
</DD>
</DL>
<HR>

<A NAME="fetchDeletedGeneFromArchive()"><!-- --></A><H3>
fetchDeletedGeneFromArchive</H3>
<PRE>
public void <B>fetchDeletedGeneFromArchive</B>()
                                 throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Fetches information about a gene from the archive.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE> - if problem occurs during retrieval</DL>
</DD>
</DL>
<HR>

<A NAME="fetchKaryotypes(org.ensembl.datamodel.CoordinateSystem, java.lang.String)"><!-- --></A><H3>
fetchKaryotypes</H3>
<PRE>
public void <B>fetchKaryotypes</B>(<A HREF="../../org/ensembl/datamodel/CoordinateSystem.html" title="class in org.ensembl.datamodel">CoordinateSystem</A>&nbsp;coordinateSystem,
                            java.lang.String&nbsp;chromosomeName)
                     throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Fetches information about karyotypes (chromosome bands) for the specified
 chromosome.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>coordinateSystem</CODE> - coordinate system containing the chromosomeName<DD><CODE>chromosomeName</CODE> - name of the chromosome of interest
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE> - if problem occurs during retrieval</DL>
</DD>
</DL>
<HR>

<A NAME="fetchSequenceRegionsSuchAsChromosomeOrContig(org.ensembl.datamodel.CoordinateSystem)"><!-- --></A><H3>
fetchSequenceRegionsSuchAsChromosomeOrContig</H3>
<PRE>
public void <B>fetchSequenceRegionsSuchAsChromosomeOrContig</B>(<A HREF="../../org/ensembl/datamodel/CoordinateSystem.html" title="class in org.ensembl.datamodel">CoordinateSystem</A>&nbsp;coordinateSystem)
                                                  throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Fetches information about the sequence regions.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE> - if problem occurs during retrieval</DL>
</DD>
</DL>
<HR>

<A NAME="showExternalRefsForAGene()"><!-- --></A><H3>
showExternalRefsForAGene</H3>
<PRE>
public void <B>showExternalRefsForAGene</B>()
                              throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Fetches a gene from the database and prints a
 summary of it's external refs if it has any.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="displayDriversState()"><!-- --></A><H3>
displayDriversState</H3>
<PRE>
public void <B>displayDriversState</B>()
                         throws <A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></PRE>
<DL>
<DD>Print details about the drivers' configurations.
<P>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/ensembl/driver/AdaptorException.html" title="class in org.ensembl.driver">AdaptorException</A></CODE></DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/Example.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../org/ensembl/Version.html" title="class in org.ensembl"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Example.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
