<!--#set var="decor" value="none"--><!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.4.2_06) on Tue Jul 12 18:05:24 BST 2005 -->
<TITLE>
GenomicAlignBlock (Ensj API)
</TITLE>

<META NAME="keywords" CONTENT="org.ensembl.compara.datamodel.GenomicAlignBlock interface">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="GenomicAlignBlock (Ensj API)";
}
</SCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/GenomicAlignBlock.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlign.html" title="interface in org.ensembl.compara.datamodel"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/ensembl/compara/datamodel/Homology.html" title="interface in org.ensembl.compara.datamodel"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="GenomicAlignBlock.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.ensembl.compara.datamodel</FONT>
<BR>
Interface GenomicAlignBlock</H2>
<DL>
<DT><B>All Superinterfaces:</B> <DD><A HREF="../../../../org/ensembl/datamodel/Persistent.html" title="interface in org.ensembl.datamodel">Persistent</A>, java.io.Serializable</DD>
</DL>
<HR>
<DL>
<DT>public interface <B>GenomicAlignBlock</B><DT>extends <A HREF="../../../../org/ensembl/datamodel/Persistent.html" title="interface in org.ensembl.datamodel">Persistent</A></DL>

<P>
I am part of the compara-analysis. I keep information about
 groups of dna-dna alignments between species.
<P>

<P>
<HR>

<P>
<!-- ======== NESTED CLASS SUMMARY ======== -->


<!-- =========== FIELD SUMMARY =========== -->


<!-- ======== CONSTRUCTOR SUMMARY ======== -->


<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=2><FONT SIZE="+2">
<B>Method Summary</B></FONT></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#addGenomicAlign(org.ensembl.compara.datamodel.GenomicAlign)">addGenomicAlign</A></B>(<A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlign.html" title="interface in org.ensembl.compara.datamodel">GenomicAlign</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getAlignmentStrings()">getAlignmentStrings</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getAllGenomicAligns()">getAllGenomicAligns</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getAllNonreferenceGenomicAligns()">getAllNonreferenceGenomicAligns</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getAllUngappedGenomicAlignBlocks()">getAllUngappedGenomicAlignBlocks</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getLength()">getLength</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/ensembl/compara/datamodel/MethodLinkSpeciesSet.html" title="interface in org.ensembl.compara.datamodel">MethodLinkSpeciesSet</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getMethodLinkSpeciesSet()">getMethodLinkSpeciesSet</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;long</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getMethodLinkSpeciesSetInternalId()">getMethodLinkSpeciesSetInternalId</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getPercentageID()">getPercentageID</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlign.html" title="interface in org.ensembl.compara.datamodel">GenomicAlign</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getReferenceGenomicAlign()">getReferenceGenomicAlign</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getReferenceSlice()">getReferenceSlice</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getReferenceSliceEnd()">getReferenceSliceEnd</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getReferenceSliceStart()">getReferenceSliceStart</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#getScore()">getScore</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#reverseComplement()">reverseComplement</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#setLength(int)">setLength</A></B>(int&nbsp;length)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#setMethodLinkSpeciesSet(org.ensembl.compara.datamodel.MethodLinkSpeciesSet)">setMethodLinkSpeciesSet</A></B>(<A HREF="../../../../org/ensembl/compara/datamodel/MethodLinkSpeciesSet.html" title="interface in org.ensembl.compara.datamodel">MethodLinkSpeciesSet</A>&nbsp;set)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#setMethodLinkSpeciesSetInternalId(long)">setMethodLinkSpeciesSetInternalId</A></B>(long&nbsp;id)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#setPercentageID(double)">setPercentageID</A></B>(double&nbsp;percId)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#setReferenceGenomicAlign(org.ensembl.compara.datamodel.GenomicAlign)">setReferenceGenomicAlign</A></B>(<A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlign.html" title="interface in org.ensembl.compara.datamodel">GenomicAlign</A>&nbsp;align)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#setReferenceSlice(org.ensembl.datamodel.Location)">setReferenceSlice</A></B>(<A HREF="../../../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A>&nbsp;refSlice)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#setReferenceSliceEnd(int)">setReferenceSliceEnd</A></B>(int&nbsp;end)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#setReferenceSliceStart(int)">setReferenceSliceStart</A></B>(int&nbsp;end)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlignBlock.html#setScore(double)">setScore</A></B>(double&nbsp;score)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.ensembl.datamodel.Persistent"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TD><B>Methods inherited from interface org.ensembl.datamodel.<A HREF="../../../../org/ensembl/datamodel/Persistent.html" title="interface in org.ensembl.datamodel">Persistent</A></B></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../org/ensembl/datamodel/Persistent.html#getDriver()">getDriver</A>, <A HREF="../../../../org/ensembl/datamodel/Persistent.html#getInternalID()">getInternalID</A>, <A HREF="../../../../org/ensembl/datamodel/Persistent.html#sameInternalID(org.ensembl.datamodel.Persistent)">sameInternalID</A>, <A HREF="../../../../org/ensembl/datamodel/Persistent.html#setDriver(org.ensembl.driver.CoreDriver)">setDriver</A>, <A HREF="../../../../org/ensembl/datamodel/Persistent.html#setInternalID(long)">setInternalID</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->


<!-- ========= CONSTRUCTOR DETAIL ======== -->


<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TD COLSPAN=1><FONT SIZE="+2">
<B>Method Detail</B></FONT></TD>
</TR>
</TABLE>

<A NAME="getAlignmentStrings()"><!-- --></A><H3>
getAlignmentStrings</H3>
<PRE>
public java.util.List <B>getAlignmentStrings</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="addGenomicAlign(org.ensembl.compara.datamodel.GenomicAlign)"><!-- --></A><H3>
addGenomicAlign</H3>
<PRE>
public void <B>addGenomicAlign</B>(<A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlign.html" title="interface in org.ensembl.compara.datamodel">GenomicAlign</A>&nbsp;align)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getAllGenomicAligns()"><!-- --></A><H3>
getAllGenomicAligns</H3>
<PRE>
public java.util.List <B>getAllGenomicAligns</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getAllNonreferenceGenomicAligns()"><!-- --></A><H3>
getAllNonreferenceGenomicAligns</H3>
<PRE>
public java.util.List <B>getAllNonreferenceGenomicAligns</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getAllUngappedGenomicAlignBlocks()"><!-- --></A><H3>
getAllUngappedGenomicAlignBlocks</H3>
<PRE>
public java.util.List <B>getAllUngappedGenomicAlignBlocks</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getLength()"><!-- --></A><H3>
getLength</H3>
<PRE>
public int <B>getLength</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setLength(int)"><!-- --></A><H3>
setLength</H3>
<PRE>
public void <B>setLength</B>(int&nbsp;length)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getMethodLinkSpeciesSet()"><!-- --></A><H3>
getMethodLinkSpeciesSet</H3>
<PRE>
public <A HREF="../../../../org/ensembl/compara/datamodel/MethodLinkSpeciesSet.html" title="interface in org.ensembl.compara.datamodel">MethodLinkSpeciesSet</A> <B>getMethodLinkSpeciesSet</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setMethodLinkSpeciesSet(org.ensembl.compara.datamodel.MethodLinkSpeciesSet)"><!-- --></A><H3>
setMethodLinkSpeciesSet</H3>
<PRE>
public void <B>setMethodLinkSpeciesSet</B>(<A HREF="../../../../org/ensembl/compara/datamodel/MethodLinkSpeciesSet.html" title="interface in org.ensembl.compara.datamodel">MethodLinkSpeciesSet</A>&nbsp;set)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getMethodLinkSpeciesSetInternalId()"><!-- --></A><H3>
getMethodLinkSpeciesSetInternalId</H3>
<PRE>
public long <B>getMethodLinkSpeciesSetInternalId</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setMethodLinkSpeciesSetInternalId(long)"><!-- --></A><H3>
setMethodLinkSpeciesSetInternalId</H3>
<PRE>
public void <B>setMethodLinkSpeciesSetInternalId</B>(long&nbsp;id)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getPercentageID()"><!-- --></A><H3>
getPercentageID</H3>
<PRE>
public double <B>getPercentageID</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setPercentageID(double)"><!-- --></A><H3>
setPercentageID</H3>
<PRE>
public void <B>setPercentageID</B>(double&nbsp;percId)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getReferenceGenomicAlign()"><!-- --></A><H3>
getReferenceGenomicAlign</H3>
<PRE>
public <A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlign.html" title="interface in org.ensembl.compara.datamodel">GenomicAlign</A> <B>getReferenceGenomicAlign</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setReferenceGenomicAlign(org.ensembl.compara.datamodel.GenomicAlign)"><!-- --></A><H3>
setReferenceGenomicAlign</H3>
<PRE>
public void <B>setReferenceGenomicAlign</B>(<A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlign.html" title="interface in org.ensembl.compara.datamodel">GenomicAlign</A>&nbsp;align)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getReferenceSlice()"><!-- --></A><H3>
getReferenceSlice</H3>
<PRE>
public <A HREF="../../../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A> <B>getReferenceSlice</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setReferenceSlice(org.ensembl.datamodel.Location)"><!-- --></A><H3>
setReferenceSlice</H3>
<PRE>
public void <B>setReferenceSlice</B>(<A HREF="../../../../org/ensembl/datamodel/Location.html" title="class in org.ensembl.datamodel">Location</A>&nbsp;refSlice)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getReferenceSliceEnd()"><!-- --></A><H3>
getReferenceSliceEnd</H3>
<PRE>
public int <B>getReferenceSliceEnd</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setReferenceSliceEnd(int)"><!-- --></A><H3>
setReferenceSliceEnd</H3>
<PRE>
public void <B>setReferenceSliceEnd</B>(int&nbsp;end)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getReferenceSliceStart()"><!-- --></A><H3>
getReferenceSliceStart</H3>
<PRE>
public int <B>getReferenceSliceStart</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setReferenceSliceStart(int)"><!-- --></A><H3>
setReferenceSliceStart</H3>
<PRE>
public void <B>setReferenceSliceStart</B>(int&nbsp;end)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="reverseComplement()"><!-- --></A><H3>
reverseComplement</H3>
<PRE>
public void <B>reverseComplement</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getScore()"><!-- --></A><H3>
getScore</H3>
<PRE>
public double <B>getScore</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setScore(double)"><!-- --></A><H3>
setScore</H3>
<PRE>
public void <B>setScore</B>(double&nbsp;score)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=3 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/GenomicAlignBlock.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../org/ensembl/compara/datamodel/GenomicAlign.html" title="interface in org.ensembl.compara.datamodel"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../org/ensembl/compara/datamodel/Homology.html" title="interface in org.ensembl.compara.datamodel"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="GenomicAlignBlock.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>

</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;CONSTR&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
