<!--#set var="decor" value="none"--><!--#set var="decor" value="none"--><!--#set var="decor" value="none"--><!--#set var="decor" value="none"--><!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
            "http://www.w3.org/TR/html4/strict.dtd">

<html>

<head>
<title>Overview</title>
</head>

<body>

 <a href="quickstart.html">Quick start</a>  - <a href="download.html">Download</a> - <a href="ensj-overview.html">Overview</a>  - <a href="api/index.html">API</a> - <a href="faq.html">FAQ</a> - <a href="howto.html">How-To</a> - <a href="api/org/ensembl/Example.java.html">Examples</a> - <a href="jython.html">Jython</a>   - <a href="contact.html">Contact</a> - <a href="links.html">Links</a> - <a href="RELEASE_NOTES">Release Notes</a>   - <a href="LICENSE">License</a> 


<h1>Introduction</h1>

Java and Jython API for Ensembl's core, compara and variation (SNP) databases.


<h2>Core databases</h2> 

The ensembl core database types such as Gene and Sequence are defined
in <a
href="org/ensembl/datamodel/package-summary.html">org.ensembl.datamodel</a>
and are retrieved via CoreDriver instances and adaptors defined in <a
href="org/ensembl/driver/package-summary.html">org.ensembl.driver</a>.
This code snippet shows how to retrieve some genes from the latest
human data set published on ensembldb.ensembl.org.

<pre><code>
import org.ensembl.driver.CoreDriver;
import org.ensembl.driver.CoreDriverFactory;
...
CoreDriver humanCoreDriver = CoreDriverFactory.createCoreDriverUsingDatabasePrefix("ensembldb.ensembl.org",3306, "homo_sapiens_core", "anonymous", null);
List genes = humanCoreDriver.getGeneAdaptor().fetch(new Location("chromosome:22:21m-22m"))
</code></pre>



<h2>Variation (SNP) databases</h2> 

Ensembl variation (SNP) databases are accessed via the
org.ensembl.variation package. Variation datatypes such as
VariationFeature and Allele are defined in <a
href="org/ensembl/variation/datamodel/package-summary.html">org.ensembl.variation.datamodel</a>
and are retrieved via VariationDriver instances and adaptors defined
in <a
href="org/ensembl/variation/driver/package-summary.html">org.ensembl.variation.driver</a>.
This code snippet shows how to retrieve some variation features (where
variations appear on the genome) from the latest human data set
published on ensembldb.ensembl.org.

<pre><code>
import org.ensembl.variation.driver.VariationDriver;
import org.ensembl.variation.driver.VariationDriverFactory;
...
VariationDriver humanVariationDriver = VariationDriverFactory.createVariationDriverUsingDatabasePrefix("ensembldb.ensembl.org",3306, "homo_sapiens_variation", "anonymous", null);
List variationFeatures = humanVariationDriver.getVariationFeatureAdaptor().fetch(new Location("chromosome:22:21m-22m"))
</code></pre>



<h2>Compara databases</h2>
Ensembl compara databases are accessed via the org.ensembl.compara
package. Compara datatypes such as Homology and DnaFragment are
defined in <a
href="org/ensembl/compara/datamodel/package-summary.html">org.ensembl.compara.datamodel</a>
and are retrieved via ComparaDriver instances and adaptors defined in
<a
href="org/ensembl/compara/driver/package-summary.html">org.ensembl.compara.driver</a>.


<h2>Examples</h2>

For more examples see <a
href="org/ensembl/Example.java.html">Example.java</a> <a
href="org/ensembl/Example.java">(txt)</a> which demonstrates how to
connect to a ensembl databases with drivers and retrieve data via
adaptors.

</body>
</html>

