<html>
<head>
<title>Ensembl: Data Tools - Assembly converter, ID History converter, Variant Effect Predictor</title>
</head>
<body>

<h1>Ensembl Tools</h1>

<p>We provide a number of ready-made tools for processing your data. At the moment, small datasets can be uploaded to our servers and processed online; for larger datasets, we provide an API script that can be downloaded (you will also need to <a href="/info/docs/api/">install our Perl API</a> to use these).</p>
<p>In the near future we aim to offer an intermediate service, whereby medium-to-large data sets can be submitted to a queue, similar to BLAST.</p>

<table class="ss padded-cell autocenter" style="width:720px">
<tr>
  <th style="width:20%" class="center">Name</th>
  <th style="width:60%" class="center">Description</th>
  <th style="width:10%" class="center">Online tool</th>
  <th style="width:10%" class="center">Download code</th>
</tr>

<tr>
  <td><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadVariations" class="modal_link nodeco"><b>Variant Effect Predictor</b><br /><img src="/i/vep_logo_sm.png" alt="[logo]"/></a></td>
  <td><p>(Formerly SNP Effect Predictor). Analyse your own variants and predict the functional consequences of
known and unknown variants via our Variant Effect Predictor (VEP)
tool.</p>
<p>You can do this online (max 750 variants), or download the script. <a href="/info/docs/variation/vep/vep_formats.html">Accepted formats</a> include a simple tab-delimited file, VCF and other
formats. Uploaded tracks can be viewed on Location pages.</p>
<p>See also <a href="/info/docs/variation/vep/index.html" target="_blank">full documentation</a>.</p>
  <td class="center"><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadVariations" class="modal_link nodeco"><img src="/i/16/tool.png" alt="Tool" title="Go to online tool" /></a></td>
  <td class="center"><a href="http://cvs.sanger.ac.uk/cgi-bin/viewvc.cgi/ensembl-tools/scripts/variant_effect_predictor/variant_effect_predictor.pl?root=ensembl&view=co&pathrev=branch-ensembl-[[SPECIESDEFS::ENSEMBL_VERSION]]" class="nodeco"><img src="/i/16/download.png" alt="Download" title="Download Perl script" /></a></td>
</tr>

<tr class="bg2">
  <td><b>Assembly converter</b></td>
  <td>Map your data to the current assembly. Accepted file formats: 
  <a href="/info/website/upload/gff.html">GFF</a>, 
  <a href="/info/website/upload/gff.html">GTF</a>, 
  <a href="/info/website/upload/bed.html">BED</a>, 
  <a href="/info/website/upload/psl.html">PSL</a>
  <br />N.B. Export is currently in GFF only</td>
  <td class="center"><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/SelectFeatures" class="modal_link nodeco"><img src="/i/16/tool.png" alt="Tool" title="Go to online tool" /></a></td>
  <td class="center"><a href="http://cvs.sanger.ac.uk/cgi-bin/viewvc.cgi/ensembl-tools/scripts/assembly_converter/?root=ensembl" class="nodeco"><img src="/i/16/download.png" alt="Download" title="Download Perl script" /></a></td>
</tr>

<tr>
  <td><b>ID History converter</b></td>
  <td>Convert a set of Ensembl IDs from a previous release into their current equivalents (<b>max 30 ids</b>).</td>
  <td class="center"><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/UploadStableIDs" class="modal_link nodeco"><img src="/i/16/tool.png" alt="Tool" title="Go to online tool" /></a></td>
  <td class="center"><a href="http://cvs.sanger.ac.uk/cgi-bin/viewvc.cgi/ensembl-tools/scripts/id_history_converter/?root=ensembl" class="nodeco"><img src="/i/16/download.png" alt="Download" title="Download Perl script" /></a></td>
</tr>

<!--
<tr>
  <td><b>Region Report</b></td>
  <td>Export standard data sets (genes, sequence, variations, etc) from one or more regions, in either GFF or simple text format (<b>max 5MB total</b>, or 1MB for variation/regulation data)</td>
  <td class="center"><a href="/[[SPECIESDEFS::ENSEMBL_PRIMARY_SPECIES]]/UserData/SelectReportOptions" class="modal_link" style="text-decoration:none"><img src="/i/16/tool.png" alt="Tool" title="Go to online tool" /></a></td>
  <td class="center"><a href="http://cvs.sanger.ac.uk/cgi-bin/viewvc.cgi/ensembl-tools/scripts/region_reporter/?root=ensembl" style="text-decoration:none"><img src="/i/16/download.png" alt="Download" title="Download Perl script" /></a></td>
</tr>
-->

<tr class="bg2">
  <td><b>Ensembl Virtual Machine</b></td>
  <td>VirtualBox virtual Machine with Ubuntu desktop and pre-configured with the latest Ensembl API plus Variant Effect Predictor (VEP). <b>NB: download is >1 GB</b>
  
   <p>See also <a href="/info/data/virtual_machine.html" target="_blank">full documentation</a>.</p>
  </td>
  <td class="center">-</td>
  <td class="center"><a href="ftp://ftp.ensembl.org/pub/current_virtual_machine" class="nodeco"><img src="/i/16/download.png" alt="Download" title="Download Virtual Machine" /></a></td>
</tr>

</table>

</body>
</html>
