<tr class="black" valign="top">
	<td rowspan="5"><img src="/gfx/blank.gif" height="1" alt=""></td>
	<td><img src="/gfx/blank.gif" height="1" alt=""></td>
	<td rowspan="5"><img src="/gfx/blank.gif" height="1" alt=""></td>
</tr>
<tr class="background2"><td class="h5">&nbsp;Submission Guidelines</td></tr>
<tr class="black" valign="top"><td><img src="/gfx/blank.gif" height="1" alt=""></td></tr>
<tr >
	<td class="gs_body">
	<table align="center" border="0" cellpadding="0" cellspacing="0" width="596" >
    	<tr><td colspan="3"><img src="/gfx/blank.gif" height="10" width="596" alt=""></td></tr>
		<tr>
			<td>&nbsp;&nbsp;</td>
			<td><b>Transposon Submission Guidelines</b><p><small>Within any genome, a transposable element (TE) consists of multiple copies generated by transposition events. Depending on the time of transposition and the rate of evolutionary change after transposition, different copies tend to have varied levels of sequence similarities. These related copies are referred as a family of TEs. Some families consist of multiple distinct groups that are subdivided into subfamilies. Obviously, relatedness is a relative concept in evolution and the distinction between family and subfamily may be dictated by specific situation. The task of TE annotation is the combination of identifying and classifying TEs, although both can be rather complicated in some cases. Therefore the annotation process will be evolving as more comprehensive analysis is performed. The primary purpose of this guideline page is to establish a naming convention to help identify TEs in the Anopheles gambiae genome. This document also provides a basic framework to outline the classification of TEs.<br />

It should be noted that the term TE family is also used to refer to a group of related TEs in diverse organisms that usually share conserved amino acid sequences in their transposase or reverse transcriptase. In the context of genome annotation, we emphasize the intra-genomic aspect of the definition of a TE family. The other issue that is not considered here is the evolutionary basis of the similarities between TE copies and TE families. A separate document that presents the canonical sequences of TE families and describes the relationships between TEs will be made available in the future.</small>
	
<ol>
<li><b><small>Name for a Transposable element family<br /></small></b><small>
Names should contain only the English alphabet and the Arabic numerals, which will ensure that they will be easily identified by a wide variety of automated searches. <br />
<br />
<i>Examples:</i> <br />
JockeyFam1 represents Family 1 of the Jockey clade of non-LTR retrotransposons.<br />
JockeyFam2 represents Family 2 of the Jockey clade of non-LTR retrotransposons.<br />
(Instead of simply using Jockey1 as the family name, we recommend using Fam to specify a family within an established clade. This is to avoid confusion with some established clade names. For example, CR1 is an established clade in non-LTR retrotransposons and it is not a family within the CR clade).<br />
<br />
Mariner represents the mariner family in the Tc1-mariner superfamily of "cut and paste" DNA transposons.<br />
<br /></small></li>
<li><b><small>Name for a specific copy of TE family in the A. gambiae genome assembly. <br /></small></b> <small>
An individual copy of a TE family is identified by adding an Arabic numeral to the family name.  An underscore is used between the family name and the Arabic numeral. <br /> 
<br />
<i>Examples:</i> <br />
JockeyFam1_1 represents copy 1 in Family 1 of the Jockey clade.<br />
mariner_1 represents copy 1 of the mariner family.<br />
<br />
Here _1 denotes a given copy (copy 1) of a TE in a given sequencing project (the A. gambiae genome project).<br /><br /></small></li>
<li><b><small>How to classify Transposable elements?<br /></small></b><small>
There is no absolute away to classify. However, we will give two examples to help provide some reference points. <br />
<br />
In the case of non-LTR retrotransposons, a systematic analysis revealed a large number of non-LTR elements in the A. gambiae genome. Phylogenetic analysis showed clear groupings consistent with currently accepted clade structures. We therefore proposed a naming system based on the clades. We further developed a working definition of a family as copies having 80% or higher amino acid sequence identity in the conserved region of the reverse transcriptase domain.  <br />
In the case of mariner, there are about 30 identifiable copies.  Some can be grouped confidently into clades while many others could not. Therefore, no attempts were made to further classify these 30 mariner elements. They are identified as 30 different copies in the mariner family.<br />
<br /></small></li>
<li> <b><small>Previously named TEs and synonyms.<br /></small></b> <small>
If we use new names for TEs that have already been named in the literature, we should provide its "old" name as synonyms.  We recommend keeping the "old" names when there is no confusion with regard to classification. <br /><br /></small></li>
<li> <b><small>General Classification.<br /></small></b> <small>
<u>Class I, RNA-mediated TEs</u><br />
<ul>
<li><small>Non-LTR retrotransposons </small></li>
<li><small>LTR retrotransposons</small></li>
<li><small>Short interspersed nuclear elements (SINEs)</small> </li>	
</ul>	
<u><br />Class II, DNA-mediated TEs</u><br />
<ul>
<li><small>"Cut and paste" DNA transposons (Tc1-mariner superfamily, pogo superfamily , etc.)</small></li>
<li><small>Miniature inverted-repeat transposable elements (MITEs)</small></li> 
<li><small>Helitrons </small></li>
</ul><br /></small></li>
</ol>
</td>
<td>&nbsp;&nbsp;</td></tr>
		<tr><td colspan="3"><img src="/gfx/blank.gif" height="10" width="596" alt=""></td></tr>
	</table>
</td></tr>
<tr class="black" valign="top"><td><img src="/gfx/blank.gif" height="1" alt=""></td></tr>

