<html>
<head>
<meta name="author" content="helpdesk" />
<meta name="navigation" content="Data" />
<meta name="update-period" content="2005-05-27" />
<title>FASTA Database Files</title>
</head>

<body>

<h2>FASTA Database Files</h2>


<h3 class="boxed">Introduction</h3>

<p>
Ensembl provides sequence databases of transcript and translation models predicted by the Ensembl analysis and annotation pipleine, 
as well as by <i>ab initio</i> methods.
The database files in
<a href="http://www.ebi.ac.uk/help/formats.html#fasta" target="external">FASTA format</a> are available from corresponding '<strong>fasta</strong>' directories on the
<a href="ftp://ftp.ensembl.org/pub/" target="external">ftp.ensembl.org</a> site.
This document describes the current naming convention and the sequence header line format used by Ensembl.
Similar descriptions are also available from README files in FTP site directories.
</p>

<p>
While Ensembl provides also a more general description of the FTP site
<a href="./ftp_structure.html">structure</a>, 'fasta' directories contain the following sub-directories:
</p>

<ul>

<li>
<p>
<strong>dna</strong>: assembled genomic DNA sequences
</p>
</li>

<li>
<p>
<strong>cdna</strong>: gene prediction transcripts (cDNAs, mRNAs)
</p>
</li>

<li>
<p>
<strong>pep</strong>: gene prediction translations (peptides, proteins)
</p>
</li>

<li>
<p>
<strong>rna</strong>: non-coding RNA gene predictions
</p>
</li>

</ul>

<p>
To facilitate storage and download all databases are
<a href="http://directory.fsf.org/gzip.html" target="external">GNU Zip</a> (gzip, *.gz) compressed.
</p>


<h3 class="boxed">FASTA Database File Names</h3>

<p>
All files deposited in these directories obey a common naming scheme: 
</p>

<pre>
species.version.month.sequence type.[status].[id type].[id].fa.gz 
</pre>


<ul>

<li>
<p>
<strong>species</strong> -
This is the systematic name of the species.
</p>
</li>

<li>
<p>
<strong>version</strong> -
The version number of the genome sequence assembly build denotes the central sequence version for all Ensembl sequence type exports.
Although, its annotation may change between assembly builds, the underlying genome sequence is only altered upon arrival of a new build.
</p>
</li>

<li>
<p>
<strong>month</strong> -
The month of the corresponding Ensembl release represents the version of genome annotation data.
Although gene-builds are generally performed on new genome sequence assemblies, the corresponding annotation may be updated between gene-builds.
</p>
</li>

<li>
<p>
<strong>sequence type</strong>
</p>

<ul>

<li>
<p>
<strong>dna</strong> -
This stores unmasked genomic DNA sequences.
</p>
</li>

<li>
<p>
<strong>dna_rm</strong> -
This is the masked genomic DNA.
Interspersed repeats and low complexity regions are detected with the
<a href="http://www.repeatmasker.org/" target="external">RepeatMasker</a> tool and masked by replacing repeats with equal numbers of 'N's.
</p>
</li>

<li>
<p>
<strong>cdna</strong> -
Transcript (cDNA) sequences resulting from Ensembl and <i>ab intio</i> gene predictions.
</p>
</li>

<li>
<p>
<strong>pep</strong> -
Translation (peptide) sequences resulting from Ensembl and <i>ab intio</i> gene predictions.
</p>
</li>

<li>
<p>
<strong>rna</strong> -
Transcript (RNA) sequences of non-coding RNA predictions.
</p>
</li>

</ul>

</li>

<li>
<p>
<strong>[status]</strong> -
The status of transcript (cDNA) or translation (peptide) sequence database files indicates the prediction method behind a particular sub-set.
For more information about these classes of Ensembl gene and transcript predictions see the
<a href="../data/docs/genome_annotation.html">Ensembl Gene Predictions</a> section.
</p>

<ul>

<li>
<p>
<strong>known_ccds</strong> -
The Consensus Coding Sequence (CCDS) project is a collaborative effort to identify a core set of human protein coding regions that are consistently annotated and of high quality. 
Initial results from the CCDS project are now available through the appropriate Ensembl gene pages and from the
<a href="http://www.ncbi.nlm.nih.gov/projects/CCDS/" target="external">CCDS</a> project page at NCBI. 
More information is available from the Ensembl 
<a href="/Homo_sapiens/ccds.html" target="external">CCDS</a> page. 
</p>
</li>

<li>
<p>
<strong>known</strong> - 
The sub-set of all transcripts or translations resulting from <strong>Ensembl known</strong> gene predictions. 
If <strong>Ensembl</strong> gene predictions can be mapped to <strong>species-specific</strong> 
<a href="http://www.ebi.ac.uk/swissprot/" target="external">UniProt/Swiss-Prot</a>, 
<a href="http://www.ncbi.nlm.nih.gov/RefSeq/" target="external">RefSeq</a> or 
<a href="http://www.ebi.ac.uk/trembl/" target="external">UniProt/TrEMBL</a> entries then Ensembl refers to them as <strong>known</strong> genes. 
</p>
</li>

<li>
<p>
<strong>novel</strong> - 
The sub-set of all transcripts or translations resulting from <strong>Ensembl novel</strong> gene predictions. 
If <strong>Ensembl</strong> gene predictions cannot be mapped to <strong>species-specific</strong> but 
<a href="http://www.ebi.ac.uk/swissprot/" target="external">UniProt/Swiss-Prot</a>, 
<a href="http://www.ncbi.nlm.nih.gov/RefSeq/" target="external">RefSeq</a> or 
<a href="http://www.ebi.ac.uk/trembl/" target="external">UniProt/TrEMBL</a> entries from <strong>closely related</strong> species then Ensembl refers to them as <strong>novel</strong> genes. 
</p>
</li>

<li>
<p>
<strong>pseudo</strong> - 
All transcripts from <strong>Ensembl pseudogene</strong> predictions. 
</p>
</li>

<li>
<p>
<strong>abinitio</strong> - 
The sub-set of transcripts or translations resulting from <i>ab initio</i> gene prediction algorithms such as the Semi-HMM-based Nucleic Acid Parser 
(<a href="http://homepage.mac.com/iankorf/" target="external">SNAP</a>) or 
<a href="http://genes.mit.edu/GENSCAN.html" target="external">GENSCAN</a>. 
Generally, all <i>ab initio</i> predictions are solely based on the genomic sequence and do not exploit any other experimental evidence.
Since, not all GENSCAN or SNAP transcript or translation predictions may represent biologically <strong>real</strong> molecules, these predictions should be used with care. 
</p>
</li>

</ul>

</li>

<li>
<p>
<strong>[id type]</strong> -
The sequence identifier type specifies the genome coordinate system in use.
Generally, Ensembl provides database files for all <strong>sequence-level</strong> entities that have been used to determine the genome sequences,
as well as for <strong>top-level</strong> entities that have been assembled from sequence-level entities.
</p>

<ul>

<li>
<p>
<strong>Top-level</strong> entities represent the largest contiguous sequences that have been assembled and are annotated by the Ensembl analysis and annotation pipeline.
</p>

<ul>

<li>
<p>
<strong>chromosome</strong> -
The top-level coordinate system for most species in Ensembl.
</p>
</li>

<li>
<p>
<strong>nonchromosomal</strong> -
Contains mitochondrial DNA or DNA sequences (supercontigs) that could not be assigned with confidence to a chromosome based on the current amount of sequence information.
</p>
</li>

</ul>

</li><!-- end top-level -->

<li>
<p>
<strong>Sequence-level</strong> entities result from the genome sequencing (BAC clones, contigs) or result from the assembly build and specify the assembly of top-level entities.
</p>

<ul>

<li>
<p>
<strong>scaffold</strong> -
Larger sequence contigs that result from the assembly of shorter (often whole genome shotgun, WGS) sequencing reads which could not yet be assembled into chromosomes.
Generally, more genome sequencing, narrowing of gaps and establishment of a tiling path is neccessary before a chromosome assembly can be established.
</p>
</li>

<li>
<p>
<strong>chunk</strong> -
While contig sequences can be assembled into larger entities, they sometimes need to be artificially broken down into the so-called 'chunks'.
This is due to limitations in the annotation pipeline and the finite record size imposed by the relational database management system (RDBMS) MySQL, which stores the genome sequence and annotation information.
</p>
</li>

<li>
<p>
<strong>clone</strong> - 
In general, this is the smallest sequence entity.
It is often identical to the sequence of one BAC clone, or the part of its sequence in the tiling path.
</p>
</li>

</ul>

</li><!-- end sequence-level -->

</ul>

</li><!-- end id type -->

<li>
<p>
<strong>[id]</strong> -
The actual sequence identifier.
Depending on the [id type] the [id] could represent the name of a chromosome, a scaffold, a contig, a clone, ...
</p>
</li>

<li>
<p>
<strong>fa</strong> - 
All files in these directories represent sequence database files in 
<a href="http://www.ebi.ac.uk/help/formats.html#fasta" target="external">FASTA format</a>. 
</p>
</li>

<li>
<p>
<strong>gz</strong> - 
All files are compressed with 
<a href="http://directory.fsf.org/gzip.html" target="external">GNU Zip</a> facilitating storage and network transfer efficiency. 
</p>
</li>

</ul>


<h3 class="boxed">FASTA Sequence Header Lines</h3>

<p>
The 
<a href="http://www.ebi.ac.uk/help/formats.html#fasta" target="external">FASTA format</a> sequence header lines are designed to be consistent across all types of Ensembl sequences, giving enough info for the sequence to be identified outside the context of the sequence database file. 
</p>

<pre>
>ID SEQTYPE:IDTYPE LOCATION 
</pre>

<p>
The following sequence header line is an example for the simple Ensembl cDNA header format: 
</p>

<pre>
>ENST00000289823 cdna:known chromosome:NCBI34:8:21922367:21927699:1 
 ^               ^    ^     ^
 ID              |    |     LOCATION 
                 |    IDTYPE 
                 SEQTYPE 
</pre>

<p>
There is obviously a great deal more transcript-specific meta data that could be added to the header. 
</p>

<pre>
>ID SEQTYPE:IDTYPE LOCATION META 
</pre>

<p>
An example for the extended sequence header format is the following line: 
</p>

<pre>
>ENST00000289823 cdna:known chromosome:NCBI34:8:21922367:21927699:1 gene:ENSG00000158815:HUGO:FGF17
 ^               ^    ^     ^                                       ^ 
 ID              |    |     LOCATION                                META
                 |    IDTYPE 
                 SEQTYPE 
</pre>


<h3 class="boxed">DNA directory</h3>


<h4>Top Level</h4>

<p>
These files contain the full sequence of the assembly in 
<a href="http://www.ebi.ac.uk/help/formats.html#fasta" target="external">FASTA format</a>. 
They contain one chromosome per file. 
</p>

<pre>
species.version.month.sequence type.id type.id.fa.gz 
</pre>

<p>
<strong>Examples</strong>
</p>

<p>
The genomic sequence of human chromosome 1: 
</p>

<pre>
Homo_sapiens.NCBI34.may.dna.chromosome.1.fa.gz 
</pre>

<p>
The masked version of the genome sequence on human chromosome 1 contains '_rm' in the name: 
</p>

<pre>
Homo_sapiens.NCBI34.may.dna_rm.chromosome.1.fa.gz 
</pre>

<p>
Non-chromosomal assembly sequences (e.g. mitochondrial genome, sequence contigs not yet mapped on chromosomes): 
</p>

<pre>
Homo_sapiens.NCBI34.may.dna.nonchromosomal.fa.gz 
Homo_sapiens.NCBI34.may.dna_rm.nonchromosomal.fa.gz 
</pre>


<h4>Sequence Level</h4>

<p>
These files represent dumps of the assembly at the sequence level in 
<a href="http://www.ebi.ac.uk/help/formats.html#fasta" target="external">FASTA format</a>. 
</p>

<pre>
species.version.month.sequence type.id type.fa.gz 
</pre>

<p>
<strong>Examples</strong>
</p>

<p>
Unmasked sequence file name examples: 
</p>

<pre>
Homo_sapiens.NCBI34.may.dna.contig.fa.gz 
Anopheles_gambiae.MOZ2a.may.dna.chunk.fa.gz 
Fugu_rubripes.FUGU2.may.dna.scaffold.fa.gz 
</pre>

<p>
Repeat masked files contain '_rm' in the file name: 
</p>

<pre>
Homo_sapiens.NCBI34.may.dna_rm.contig.fa.gz 
Anopheles_gambiae.MOZ2a.may.dna_rm.chunk.fa.gz 
Fugu_rubripes.FUGU2.may.dna_rm.scaffold.fa.gz 
</pre>

<p>
Note that the sequence 'id type' varies in different species: contigs in human, chunks in <i>Anopheles gambiae</i>, scaffolds in <i>Takifugu rubripes</i>.
</p>


<h3 class="boxed">README</h3>

<p>
Each directory on 
<a href="ftp://ftp.ensembl.org/pub/" target="external">ftp.ensembl.org</a> contains an auto-generated 
<a href="ftp://ftp.ensembl.org/pub/README">README</a> file, explaining the filenames and
<a href="http://www.ebi.ac.uk/help/formats.html#fasta" target="external">FASTA format</a> header line conventions in use. 
</p>

</body>
</html>
