<html>
<head>
<meta name="author" content="webteam" />
<meta name="navigation" content="Data" />
<title>Data</title>
</head>

<body>

<h2>Data</h2>


<h3 class="boxed"  id="documentation">General Documentation and Information</h3>

<ul>

<li>
<p>
The Ensembl project provides data sets resulting from an automated
<a href="./docs/genome_annotation.html">genome analysis and annotation process</a>.
In addition to Ensembl and EST gene sets, the project also provides other data sets on a genome-wide scale.
</p>

<div align="right"><a href="./docs/index.html">More Documentation...</a></div>
</li>

</ul>


<h3 class="boxed"  id="downloads">Data Downloads</h3>

<ul>

<li>
<p>
All data sets generated by the Ensembl project are freely available to download from the
<a href="ftp://ftp.ensembl.org/pub/" target="external">ftp.ensembl.org</a> site.
Please see also the <a href="../about/disclaimer.html">disclaimer</a>.
</p>
</li>

<li>
<p>
Ensembl provides sequence databases of gene, transcript and protein predictions suitable for local installations of sequence similarity search systems,
as well as database table dumps of <strong>all</strong> databases underlying the Ensembl system.
</p>

<div align="right"><a href="./download.html">More on FTP Downloads...</a></div>
</li>
</ul>

<h3 class="boxed">Comparative genomics and proteomics data</h3>
<ul>
<li>
<p>
Ensembl produces comparative genomics data using BlastZ-net, translated blat, MLAGAN and gene homology analysis.  
</p>
<div align="right"><a href="./compara/index.html">More on comparative data...</a></div>
</li>
</ul>


<h3 class="boxed"  id="mining">Data Mining</h3>

<ul>
<li>
<p>
Use
<a href="/multi/martview">BioMart</a> our large-scale data mining tool for more advanced data queries.
The tool goes way beyond the simple data sets we provide via the FTP site and you can export the data in 
HTML, Microsoft Excel and various text formats suitable for import into other database or analysis systems.
</p>

<div align="right"><a href="http://www.ebi.ac.uk/biomart/" target="external">More on BioMart...</a></div>
</li>

</ul>


<h3 class="boxed" id="export">Data Export</h3>

<ul>

<li>
<p>
For each species Ensembl provides a dedicated
<a href="/Homo_sapiens/exportview">ExportView</a> page.
This data display is especially designed for small-scale data exports in HTML, text or zipped format.
</p>
</li>

</ul>


<h3 class="boxed" id="import">Data Import</h3>

<ul>

<li>
<p>
View small
<a href="./external_data/index.html">custom data sets</a> in Ensembl.
</p>
</li>

<li>
<p>
Set up a
<a href="./external_data/das/index.html">DAS</a> server to display larger data sets in a more stable and flexible manner.
</p>
</li>

</ul>


<h3 class="boxed"  id="searching">Searching Ensembl</h3>

<p>
Ensembl data sets can be searched by various criteria.
</p>

<div align="right"><a href="./searching.html">More on Searching...</a></div>

<h3 class="boxed">Software Downloads</h3>

<ul>

<li>
<p>
Ensembl software is generally <strong>not</strong> released via the Ensembl FTP site, but can be obtained via anonymous access to our
<a href="http://directory.fsf.org/cvs.html" target="external">Concurrent Versions System</a> (CVS) repository.
Please see the Ensembl
<a href="../software/index.html">Software</a> section for details on
<a href="../software/website/installation/index.html">website</a> and
<a href="../software/api_installation.html">Perl API</a>, as well as
<a href="../software/java/index.html">Java API</a> installation.
</p>

<div align="right"><a href="../software/index.html">More on Ensembl Software...</a></div>
</li>

</ul>


</body>
</html>
