<html>
<head>
<meta name="author" content="webteam" />
<meta name="navigation" content="Data" />
<title>Data</title>
</head>

<body>

<h2>Data</h2>


<h3 class="boxed"  id="documentation">Genome Analysis and Annotation</h3>

<p>
The Ensembl project provides data sets resulting from an automated
<a href="./docs/genome_annotation.html">genome analysis and annotation process</a>.
In addition to Ensembl and EST gene sets, the project also provides other data
sets on a genome-wide scale.
<a href="./docs/index.html">More...</a>
</p>


<h3 class="boxed" id="import">Custom Annotation</h3>

<p>
Ensembl offers you the ability to display <strong>your own</strong> data sets
in the genome browser framework.
You may also share your data source with others if you whish to do so.
</p>

<ul>

<li>
<p>
<strong>Data Upload</strong> -
Upload data sets directly into Ensembl.
This procedure is entirely based on web forms and recommended for smaller,
more transient data sets.
<a href="./external_data/index.html">More...</a>
</p>
</li>

<li>
<p>
<strong>Distributed Annotation System</strong> -
The distributed Annotation System (DAS) is a more stable way of integrating
large data sets with the Ensembl genome browser.
It requires installation of a DAS server but provides additional flexibility
for data display.
<a href="./external_data/das/index.html">More...</a>
</p>
</li>

</ul>


<h3 class="boxed">Comparative Genomics and Proteomics</h3>

<p>
Ensembl produces comparative genomics data using BlastZ-net, translated blat,
MLAGAN and gene homology analysis.
<a href="./compara/index.html">More...</a>
</p>


<h3 class="boxed"  id="mining">Data Mining</h3>

<p>
The Ensembl project provides huge data sets.
We aim however to also provide several routes of access to these data sets.
</p>

<ul>

<li>
<p>
<strong>Entry Points</strong> -
Ensembl data sets can be searched by various criteria.
<a href="./searching.html">More...</a>
</p>
</li>

<li>
<p>
<strong>Large-Scale Exports</strong> -
Use
<a href="/multi/martview">BioMart</a> our large-scale data mining tool for more
advanced data queries.
The tool goes way beyond the simple data sets we provide via the FTP site and
you can export the data in HTML, Microsoft Excel and various text formats
suitable for import into other database or analysis systems.
<a href="http://www.ebi.ac.uk/biomart/" target="external">More...</a>
</p>
</li>

<li>
<p>
<strong>Small-Scale Exports</strong> -
For each species Ensembl provides a dedicated
<a href="/default/exportview">ExportView</a> page.
This data display is especially designed for small-scale data exports in HTML,
text or zipped format.
</p>
</li>

</ul>


<h3 class="boxed"  id="downloads">Data Downloads</h3>

<p>
<strong>All</strong> data sets generated by the Ensembl project are freely
available to download from the <strong>Ensembl FTP</strong> site.
<a href="./download.html">More...</a>
</p>


<h3 class="boxed">Software Downloads</h3>

<p>
Ensembl software is generally <strong>not</strong> released via the Ensembl FTP
site, but can be obtained via anonymous access to our
<a href="http://directory.fsf.org/cvs.html" target="external">Concurrent Versions System</a> (CVS) repository.
Please see the Ensembl
<a href="../software/index.html">Software</a> section for details on
<a href="../software/website/installation/index.html">website</a> and
<a href="../software/api_installation.html">Perl API</a>, as well as
<a href="../software/java/index.html">Java API</a> installation.
<a href="../software/index.html">More...</a>
</p>

</body>
</html>
