<html>
<head>
<title>Filtering results</title>
<meta name="order" content="5" />
</head>

<body>
    
<style>
tr:nth-child(odd) {background-color: #f0f0f0;}
</style>

<div>
	
	<div style="float:right"><img src="/img/vep_logo.png"/></div>
	
	<h1 id="top"><span style="color:#006">Variant Effect Predictor</span> <img src="/i/16/search.png"> <span style="color:#666">Filtering results</span></h1>
	<hr/>
    
		
	
    <p> The filter_vep.pl script is included along side the main VEP script. It
    can be used to filter VEP output files down to find important or interesting
    results. </p>
    
    <p> It operates on either standard or VCF formatted output (NB only VCF
    output produced by the VEP or in the same format can be used). </p>
    
    <h2 id="filter_run">Running the script</h2>
    
    <p> Run the script as follows: </p>
    
    <pre class="code">perl variant_effect_predictor.pl -i in.vcf -o out.txt -cache -everything
perl filter_vep.pl -i out.txt -o out_filtered.txt -filter "[filter_text]"</pre>
    
    <p> The script can also read from STDIN and write to STDOUT, and so may be
    used in a UNIX pipe: </p>
    
    <pre class="code">perl variant_effect_predictor.pl -i in.vcf -o stdout -cache -check_existing | perl filter_vep.pl -filter "not Existing_variation" -o out.txt</pre>
    
    <p> The above command removes known variants from your output </p>
    
    <hr/>
    <h2 id="filter_opt">Options</h2>
    
	<table class="ss" style="width:75%;">
		<tr>
			<th>Flag</th>
			<th>Alternate</th>
			<th>Description</th>
		</tr>
		<tr>
			<td><pre>--help</pre></td>
			<td><pre>-h</pre></td>
			<td>Print usage message and exit</td>
		</tr>
		<tr>
			<td><pre>--input_file [file]</pre></td>
			<td><pre>-i</pre></td>
			<td>
                Specify the input file (i.e. the VEP results file). If no input
                file is specified, the script will attempt to read from STDIN.
                Input may be gzipped - to force the script to read a file as
                gzipped, use --gz
			</td>
		</tr>
		<tr>
			<td><pre>--format [format]</pre></td>
			<td>&nbsp;</td>
			<td>
                Specify input file format (vep or vcf)
			</td>
		</tr>
		<tr>
			<td><pre>--output_file [file]</pre></td>
			<td><pre>-o</pre></td>
			<td>
                Specify the output file to write to. If no output file is
                specified, the script will write to STDOUT
			</td>
		</tr>
		<tr>
			<td><pre>--force_overwrite</pre></td>
			<td>&nbsp;</td>
			<td>
                Force the script to overwrite the output file if it already
                exists
			</td>
		</tr>
		<tr>
			<td><pre>--filter [filters]</pre></td>
			<td><pre>-f</pre></td>
			<td>
                Add filter (see below). Multiple --filter flags may be used, and
                are treated as logical ANDs, i.e. all filters must pass for a
                line to be printed
			</td>
		</tr>
		<tr>
			<td><pre>--list</pre></td>
			<td><pre>-l</pre></td>
			<td>
                List allowed fields from the input file
			</td>
		</tr>
		<tr>
			<td><pre>--count</pre></td>
			<td><pre>-c</pre></td>
			<td>
                Print only a count of matched lines
			</td>
		</tr>
		<tr>
			<td><pre>--only_matched</pre></td>
			<td>&nbsp;</td>
			<td>
                In VCF files, the CSQ field that contains the consequence data
                will often contain more than one "block" of consequence data,
                where each block corresponds to a variant/feature overlap. Using
                --only_matched will remove blocks that do not pass the filters.
                By default, the script prints out the entire VCF line if any of
                the blocks pass the filters.
			</td>
		</tr>
    </table>
    
    <hr/>
    <h2 id="filter_write">Writing filters</h2>
    
    <p> Filter strings consist of three components: </p>
    
    <ol>
        <li><b>Field</b> : A field name from the VEP results file. This can be
        any field in the "main" columns of the output, or any in the "Extra"
        final column. For VCF files, this is any field defined in the
        "##INFO=&lt;ID=CSQ" header. You can list available fields using
        --list</li>
        <li><b>Operator</b> : The operator defines the comparison carried out.
        </li>
        <li><b>Value</b> : The value to which the content of the field is
        compared.</li>
    </ol>
    
    <p> Examples: </p>
    
    <pre class="code"># match entries where Feature (Transcript) is "ENST00000307301"
--filter "Feature is ENST00000307301"

# match entries where Protein_position is less than 10
--filter "Protein_position < 10"

# match entries where Consequence contains "stream" (this will match upstream and downstream)
--filter "Consequence matches stream"</pre>
        
    <p> For certain fields you may only be interested in whether it is defined;
    in this case the operator and value can be left out: </p>
    
    <pre class="code"># match entries where HGNC gene symbol is defined
--filter "HGNC"</pre>
    
    <p> Filter strings can also be linked together by the logical operators "or"
    and "and", and inverted by prefixing with "not": </p>
    
    <pre class="code"># filter for missense variants in CCDS transcripts where the variant falls in a protein domain
--filter "Consequence is missense_variant and CCDS and DOMAINS"

# find variants where the MAF is greater than 10% in either AFR or ASN populations
--filter "AFR_MAF > 0.1 or ASN_MAF > 0.1"

# filter out known variants
--filter "not Existing_variation"</pre>
    
    <p> For fields that contain string and number components, the script will
    try and match the relevant part based on the operator in use. For example,
    using <a href="vep_options.html#opt_sift">--sift b</a> in the VEP gives strings that look like
    "tolerated(0.46)". This will give a match to either of the following
    filters: </p>
    
    <pre class="code"># match string part
--filter "SIFT is tolerated"

# match number part
--filter "SIFT < 0.5"</pre>
    
    
    <hr/>
    <h2 id="filter_op">Operators</h2>
    
    <ul>
        <li><b>is</b> (synonyms: = , eq) : Match exactly
        <pre class="code"># get only transcript consequences
--filter "Feature_type is Transcript"</pre></li>
        <li><b>!=</b> (synonym: ne) : Does not match exactly
        <pre class="code"># filter out tolerated SIFT predictions
--filter "SIFT != tolerated"</pre></li>
        <li><b>match</b> (synonyms: matches , re , regex) : Match string using
        regular expression. You may include any regular expression notation,
        e.g. "\d" for any numerical character
        <pre class="code"># match stop_gained, stop_lost and stop_retained
--filter "Consequence match stop"</pre></li>
        <li><b>&lt;</b> (synonym: lt) : Less than
        <pre class="code"># find SIFT scores less than 0.1
--filter "SIFT < 0.1"</pre></li>
        <li><b>&gt;</b> (synonym: gt) : Greater than
        <pre class="code"># find variants not in the first exon
--filter "Exon > 1"</pre></li>
        <li><b>&lt;=</b> (synonym: lte) : Less than or equal to</li>
        <li><b>&gt;=</b> (synonym: gte) : Greater than or equal to</li>
        <li><b>exists</b> (synonyms: ex , defined) : Field is defined - equivalent to using no operator and value</li>
        <li><b>in</b> : Find in list or file. Value may be either a
        comma-separated list or a file containing values on separate lines. Each
        list item is compared using the "is" operator.
        <pre class="code"># find variants in a list of gene names
--filter "HGNC in BRCA1,BRCA2"

# filter using a file of MotifFeatures
--filter "Feature in /data/files/motifs_list.txt"</pre></li>
    </ul>
    
    
</div>

</body>
</html>
