<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>

  <meta content="text/html; charset=ISO-8859-1" http-equiv="content-type">
  <title>Comparative Genomics</title>


  <meta name="author" content="web">

  <meta name="navigation" content="Comparative Genomics">

</head>
<body>

<h2>Gene Orthology/Paralogy prediction method</h2>
<div style="text-align: justify;">
The gene orthology and paralogy predictions are generated by a pipeline
where maximum likelihood phylogenetic gene trees (generated by NJTREE)
play a central role. They aim to represent the evolutionary history of
gene families, i.e. genes that diverged from a common ancestor. These
gene trees reconciled with their species tree (using RAP) have their
internal nodes annotated to distinguish duplication or speciation events.<br>
</div>

<br>
<div style="text-align: justify;">
There is a clear concordance with reciprocal best approaches in the
simple case of unique orthologous genes. However, the gene tree
pipeline is able to find more complex one-to-many and many-to-many
relations. This for instance significantly raises the number of Teleost
(bony fish) to Mammal orthologs and has even more dramatic effects on
Fly/Mammal or Worm/Mammal orthologous gene predictions. Using this
approach, we are also able to "time" duplication events which produce
paralogs by
identifying the most recent common ancestor (=taxonomy level) for a
given
internal node of the tree.<br>
</div>

<br>

The gene orthology and paralogy prediction pipeline has 7 basic steps:
<br>

<br>
<span style="font-weight: bold;">
1.</span> Load the longest translation of each gene from all species used in
Ensembl<br>

<br>
<span style="font-weight: bold;">
2.</span> run WUBlastp+SmithWaterman of every gene against every other (both
self and non-self species) in a genome-wise manner<br>

<br>
<span style="font-weight: bold;">
3.</span> Build a graph of gene relations based on Best Reciprocal Hits
(BRH<sup>1</sup>) and
Best Score Ratio (BSR)<sup>2</sup> values<br>

<br>
<span style="font-weight: bold;">
4.</span> Extract the connected components (=single linkage clusters), each
cluster
representing a gene family.<br>

<br>
<span style="font-weight: bold;">
5.</span> For each cluster, build a multiple alignment based on the
protein sequences using MUSCLE<sup>3</sup><br>

<br>
<span style="font-weight: bold;">
6.</span> For each aligned cluster, build a phylogenetic tree using
NJTREE<sup>4</sup> using the CDS sequences of the multiple alignment.
An rooted tree with internal duplication tags is obtained at this
stage, reconciling it with the species tree in
'ensembl-compara/scripts/pipeline/species_tree_njtree.taxon_id.nh'.<br>

<br>
<span style="font-weight: bold;">
7.</span> From each gene tree, infer gene pairwise relations of orthology and
paralogy types.<br>

<br>
<span style="font-weight: bold;">
Orthologues</span> : any gene pairwise relation where the ancestor node is a
speciation event.<br>
We predict several descriptions of orthologues.<br>

&nbsp;<span style="font-style: italic;">ortholog_one2one</span><br style="font-style: italic;">
<span style="font-style: italic;">
&nbsp;ortholog_one2many</span><br style="font-style: italic;">
<span style="font-style: italic;">
&nbsp;ortholog_many2many</span><br style="font-style: italic;">
<span style="font-style: italic;">
&nbsp;apparent_ortholog_one2one</span> (is a special case, see below)<br>

<br>

<span style="font-weight: bold;">Paralogues</span> : any gene pairwise relation where the ancestor node is a
duplication event.<br>
We predict several descriptions of paralogues.<br>

&nbsp;<span style="font-style: italic;">within_species_paralog</span><br style="font-style: italic;">
<span style="font-style: italic;">
&nbsp;between_species_paralog</span><br>

<img style="width: 602px; height: 414px;" alt="tree_example1" src="tree_example1.jpg"><br>

<br>
<div style="text-align: justify;">
A within_species_paralog corresponds to a relation between 2
genes of the SAME species where the ancestor node
has been labelled as a duplication node e.g. H2:H2', M2:M2' but does
not necessarily mean that the duplication event
has occurred in this species only. For example, M2':M3' are
also within_species_paralog but the duplication event has occurred
in the common ancestor between species H and species M. If H is human
and M Mouse, the taxonomy level "times" the duplication
event to the ancestor of "Euarchontoglires".
<br>
</div>

<br>
<div style="text-align: justify;">
A between_species_paralog corresponds to a relation between 2
genes of DIFFERENT species where the ancestor node
has been labelled as a duplication node e.g. M1:H2 or M1:H3.
<br>

Special cases of&nbsp;between_species_paralog can be singled out where they
can be characterised as one2one relations. Such cases are then relabelled
apparent_ortholog_one2one. Apparent orthologs can be the results of real
duplication followed
by gene loss (as shown in the picture below), but can also be the results of a wrong gene tree topology
and wrong duplication node labelling.
<br>
</div>

<br>

<img style="width: 618px; height: 470px;" alt="tree_example2" src="tree_example2.jpg"><br>

<br>
<span style="font-style: italic; text-decoration: underline;">Notes and References</span><br>
<br>
<sup>
1</sup> BRH: Best Reciprocal Hit. When a query gene translation has an
unambiguous 'best' hit to a target translation, and that particular
target translation has an unambiguous 'best' hit back to the starting
query translation. The discrimination function for 'best' is a simple
sort of blastp &gt;score, <evalue ,="">%identity,
&gt;%positivity.
<br>
<br>
<sup>2</sup> BSR: Blast Score Ratio. When 2 proteins P1 and P2 are compared,
BSR=scoreP1P2/max(self-scoreP1 or self-scoreP2). The default threshold
used in the initial clustering step is 0.33.
<br>
<br>
<sup>3</sup>&nbsp;Edgar RC "MUSCLE: a multiple sequence alignment method with reduced time and space complexity" <a href="http://www.biomedcentral.com/1471-2105/5/113">BMC Bioinformatics 2004 Aug 19;5:113.</a><br>
</evalue><evalue ,=""><br>
<sup>4</sup>&nbsp;Li H et al. NJTREE: <a href="http://treesoft.sourceforge.net/njtree.shtml">http://treesoft.sourceforge.net/njtree.shtml</a>. Manuscript in preparation.<br> 
</evalue><font size="+1"><b></b></font><br>
<evalue ,="">

</evalue>
<br>
</body>
</html>
